<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>PgR6 class with Methods and Sequences. — PgR6MS • pagoo</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="PgR6 class with Methods and Sequences. — PgR6MS" />
<meta property="og:description" content="PgR6 with Methods and Sequences. Final users should use pagoo
instead of this, since is more easy to understand.
 Inherits: PgR6M" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-164668935-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-164668935-1');
</script>


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/pagoo.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Input.html">1 - Input</a>
    </li>
    <li>
      <a href="../articles/Querying_Data.html">2 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/Subseting.html">3 - Subseting</a>
    </li>
    <li>
      <a href="../articles/Methods_Plots.html">4 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/Recipes.html">5 - Recipes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/iferres/pagoo/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>PgR6 class with Methods and Sequences.</h1>
    <small class="dont-index">Source: <a href='https://github.com/iferres/pagoo/blob/master/R/PgR6MS.R'><code>R/PgR6MS.R</code></a></small>
    <div class="hidden name"><code>PgR6MS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>PgR6 with Methods and Sequences. Final users should use <code><a href='pagoo.html'>pagoo</a></code>
instead of this, since is more easy to understand.
 Inherits: <code><a href='PgR6M.html'>PgR6M</a></code></p>
    </div>



    <h2 class="hasAnchor" id="super-classes"><a class="anchor" href="#super-classes"></a>Super classes</h2>

    <p><code><a href='PgR6.html'>pagoo::PgR6</a></code> -&gt; <code><a href='PgR6M.html'>pagoo::PgR6M</a></code> -&gt; <code>PgR6MS</code></p>
    <h2 class="hasAnchor" id="active-bindings"><a class="anchor" href="#active-bindings"></a>Active bindings</h2>

    <p><div class="r6-active-bindings"></p><dl>
<dt><code>sequences</code></dt><dd><p>A <code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSetList-class.html'>DNAStringSetList</a></code> with the
set of sequences grouped by cluster. Each group is accessible as were a list. All
<code>Biostrings</code> methods are available.</p></dd>

<dt><code>core_sequences</code></dt><dd><p>Like <code>$sequences</code>, but only showing core
sequences.</p></dd>

<dt><code>cloud_sequences</code></dt><dd><p>Like <code>$sequences</code>, but only showing cloud
sequences as defined above.</p></dd>

<dt><code>shell_sequences</code></dt><dd><p>Like <code>$sequences</code>, but only showing shell
sequences, as defined above.</p></dd>

</dl><p></div></p>
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    
<h3 class='hasAnchor' id='public-methods'><a class='anchor' aria-hidden='true' href='#public-methods'></a>Public methods</h3>

<ul>
<li><p><a href='#method-new'><code>PgR6MS$new()</code></a></p></li>
<li><p><a href='#method-core_seqs_4_phylo'><code>PgR6MS$core_seqs_4_phylo()</code></a></p></li>
<li><p><a href='#method-clone'><code>PgR6MS$clone()</code></a></p></li>
</ul>
<p><details ><summary>Inherited methods</summary>
<ul>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6" data-id="add_metadata"><a href='../../pagoo/html/PgR6.html#method-add_metadata'><code>pagoo::PgR6$add_metadata()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6" data-id="drop"><a href='../../pagoo/html/PgR6.html#method-drop'><code>pagoo::PgR6$drop()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6" data-id="recover"><a href='../../pagoo/html/PgR6.html#method-recover'><code>pagoo::PgR6$recover()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6" data-id="save_pangenomeRDS"><a href='../../pagoo/html/PgR6.html#method-save_pangenomeRDS'><code>pagoo::PgR6$save_pangenomeRDS()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6" data-id="write_pangenome"><a href='../../pagoo/html/PgR6.html#method-write_pangenome'><code>pagoo::PgR6$write_pangenome()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="cg_exp_decay_fit"><a href='../../pagoo/html/PgR6M.html#method-cg_exp_decay_fit'><code>pagoo::PgR6M$cg_exp_decay_fit()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="dist"><a href='../../pagoo/html/PgR6M.html#method-dist'><code>pagoo::PgR6M$dist()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="gg_barplot"><a href='../../pagoo/html/PgR6M.html#method-gg_barplot'><code>pagoo::PgR6M$gg_barplot()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="gg_binmap"><a href='../../pagoo/html/PgR6M.html#method-gg_binmap'><code>pagoo::PgR6M$gg_binmap()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="gg_curves"><a href='../../pagoo/html/PgR6M.html#method-gg_curves'><code>pagoo::PgR6M$gg_curves()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="gg_dist"><a href='../../pagoo/html/PgR6M.html#method-gg_dist'><code>pagoo::PgR6M$gg_dist()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="gg_pca"><a href='../../pagoo/html/PgR6M.html#method-gg_pca'><code>pagoo::PgR6M$gg_pca()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="gg_pie"><a href='../../pagoo/html/PgR6M.html#method-gg_pie'><code>pagoo::PgR6M$gg_pie()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="pan_pca"><a href='../../pagoo/html/PgR6M.html#method-pan_pca'><code>pagoo::PgR6M$pan_pca()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="pg_power_law_fit"><a href='../../pagoo/html/PgR6M.html#method-pg_power_law_fit'><code>pagoo::PgR6M$pg_power_law_fit()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="rarefact"><a href='../../pagoo/html/PgR6M.html#method-rarefact'><code>pagoo::PgR6M$rarefact()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="pagoo" data-topic="PgR6M" data-id="runShinyApp"><a href='../../pagoo/html/PgR6M.html#method-runShinyApp'><code>pagoo::PgR6M$runShinyApp()</code></a></span></p></li>
</ul>
</details>

<hr>
<a id="method-new"></a></p><h3 class='hasAnchor' id='method-'><a class='anchor' aria-hidden='true' href='#method-'></a>Method <code>new()</code></h3>
<p>Create a <code>PgR6MS</code> object.</p><h4 class='hasAnchor' id='usage'><a class='anchor' aria-hidden='true' href='#usage'></a>Usage</h4>
<p><div class="r"></p><pre class='sourceCode r'><code><span class='va'>PgR6MS</span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>
  <span class='va'>data</span>,
  <span class='va'>org_meta</span>,
  <span class='va'>cluster_meta</span>,
  core_level <span class='op'>=</span> <span class='fl'>95</span>,
  sep <span class='op'>=</span> <span class='st'>"__"</span>,
  <span class='va'>DF</span>,
  <span class='va'>group_meta</span>,
  <span class='va'>sequences</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></code></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' aria-hidden='true' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>data</code></dt><dd><p>A <code>data.frame</code> or <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code> containing at least the
following columns: <code>gene</code> (gene name), <code>org</code> (organism name to which the gene belongs to),
and <code>cluster</code> (group of orthologous to which the gene belongs to). More columns can be added as metadata
for each gene.</p></dd>

<dt><code>org_meta</code></dt><dd><p>(optional) A <code>data.frame</code> or <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code>
containing additional metadata for organisms. This <code>data.frame</code> must have a column named "org" with
valid organisms names (that is, they should match with those provided in <code>data</code>, column <code>org</code>), and
additional columns will be used as metadata. Each row should correspond to each organism.</p></dd>

<dt><code>cluster_meta</code></dt><dd><p>(optional) A <code>data.frame</code> or <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code>
containing additional metadata for clusters. This <code>data.frame</code> must have a column named "cluster" with
valid organisms names (that is, they should match with those provided in <code>data</code>, column <code>cluster</code>), and
additional columns will be used as metadata. Each row should correspond to each cluster.</p></dd>

<dt><code>core_level</code></dt><dd><p>The initial core_level (that's the percentage of organisms a core cluster must be in to be
considered as part of the core genome). Must be a number between 100 and 85, (default: 95). You can change it
later by using the <code>$core_level</code> field once the object was created.</p></dd>

<dt><code>sep</code></dt><dd><p>A separator. By default is '__'(two underscores). It will be used to
create a unique <code>gid</code> (gene identifier) for each gene. <code>gid</code>s are created by pasting
<code>org</code> to <code>gene</code>, separated by <code>sep</code>.</p></dd>

<dt><code>DF</code></dt><dd><p>Deprecated. Use <code>data</code> instead.</p></dd>

<dt><code>group_meta</code></dt><dd><p>Deprecated. Use <code>cluster_meta</code> instead.</p></dd>

<dt><code>sequences</code></dt><dd><p>Can accept: 1) a named <code>list</code> of named <code>character</code> vector. Name of list
are names of organisms, names of character vector are gene names; or 2) a named <code>list</code> of <code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSetList</a></code>
objects (same requirements as (1), but with BStringSet names as gene names); or 3) a <code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSetList-class.html'>DNAStringSetList</a></code>
(same requirements as (2) but <code>DNAStringSetList</code> names are organisms names).</p></dd>

<dt><code>verbose</code></dt><dd><p><code>logical</code>. Whether to display progress messages when loading class.</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='returns'><a class='anchor' aria-hidden='true' href='#returns'></a>Returns</h4>
<p>An R6 object of class PgR6MS. It contains basic fields and methods for analyzing a pangenome. It also
contains additional statistical methods for analyze it, methods to make basic exploratory plots, and methods
for sequence manipulation.</p>
<h3 class='hasAnchor' id='list-new-'><a class='anchor' aria-hidden='true' href='#list-new-'></a>Method <code>new()</code></h3>
<p>Create a <code>PgR6MS</code> object.</p><h4 class='hasAnchor' id='usage'><a class='anchor' aria-hidden='true' href='#usage'></a>Usage</h4>
<p><div class="r"></p><pre class='sourceCode r'><code><span class='va'>PgR6MS</span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>
  <span class='va'>data</span>,
  <span class='va'>org_meta</span>,
  <span class='va'>cluster_meta</span>,
  core_level <span class='op'>=</span> <span class='fl'>95</span>,
  sep <span class='op'>=</span> <span class='st'>"__"</span>,
  <span class='va'>DF</span>,
  <span class='va'>group_meta</span>,
  <span class='va'>sequences</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></code></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' aria-hidden='true' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>data</code></dt><dd><p>A <code>data.frame</code> or <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code> containing at least the
following columns: <code>gene</code> (gene name), <code>org</code> (organism name to which the gene belongs to),
and <code>cluster</code> (group of orthologous to which the gene belongs to). More columns can be added as metadata
for each gene.</p></dd>

<dt><code>org_meta</code></dt><dd><p>(optional) A <code>data.frame</code> or <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code>
containing additional metadata for organisms. This <code>data.frame</code> must have a column named "org" with
valid organisms names (that is, they should match with those provided in <code>data</code>, column <code>org</code>), and
additional columns will be used as metadata. Each row should correspond to each organism.</p></dd>

<dt><code>cluster_meta</code></dt><dd><p>(optional) A <code>data.frame</code> or <code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code>
containing additional metadata for clusters. This <code>data.frame</code> must have a column named "cluster" with
valid organisms names (that is, they should match with those provided in <code>data</code>, column <code>cluster</code>), and
additional columns will be used as metadata. Each row should correspond to each cluster.</p></dd>

<dt><code>core_level</code></dt><dd><p>The initial core_level (that's the percentage of organisms a core cluster must be in to be
considered as part of the core genome). Must be a number between 100 and 85, (default: 95). You can change it
later by using the <code>$core_level</code> field once the object was created.</p></dd>

<dt><code>sep</code></dt><dd><p>A separator. By default is '__'(two underscores). It will be used to
create a unique <code>gid</code> (gene identifier) for each gene. <code>gid</code>s are created by pasting
<code>org</code> to <code>gene</code>, separated by <code>sep</code>.</p></dd>

<dt><code>DF</code></dt><dd><p>Deprecated. Use <code>data</code> instead.</p></dd>

<dt><code>group_meta</code></dt><dd><p>Deprecated. Use <code>cluster_meta</code> instead.</p></dd>

<dt><code>sequences</code></dt><dd><p>Can accept: 1) a named <code>list</code> of named <code>character</code> vector. Name of list
are names of organisms, names of character vector are gene names; or 2) a named <code>list</code> of <code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSetList</a></code>
objects (same requirements as (1), but with BStringSet names as gene names); or 3) a <code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSetList-class.html'>DNAStringSetList</a></code>
(same requirements as (2) but <code>DNAStringSetList</code> names are organisms names).</p></dd>

<dt><code>verbose</code></dt><dd><p><code>logical</code>. Whether to display progress messages when loading class.</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='returns'><a class='anchor' aria-hidden='true' href='#returns'></a>Returns</h4>
<p>An R6 object of class PgR6MS. It contains basic fields and methods for analyzing a pangenome. It also
contains additional statistical methods for analyze it, methods to make basic exploratory plots, and methods
for sequence manipulation.</p>
<p><hr>
<a id="method-core_seqs_4_phylo"></a></p><h3 class='hasAnchor' id='method-'><a class='anchor' aria-hidden='true' href='#method-'></a>Method <code>core_seqs_4_phylo()</code></h3>
<p>A field for obtaining core gene sequences is available (see below), but for creating a phylogeny with this
sets is useful to: 1) have the possibility of extracting just one sequence of each organism on each cluster, in
case paralogues are present, and 2) filling gaps with empty sequences in case the core_level was set below 100%,
allowing more genes (some not in 100% of organisms) to be incorporated to the phylogeny. That is the purpose of
this special function.</p><h4 class='hasAnchor' id='usage'><a class='anchor' aria-hidden='true' href='#usage'></a>Usage</h4>
<p><div class="r"></p><pre class='sourceCode r'><code><span class='va'>PgR6MS</span><span class='op'>$</span><span class='fu'>core_seqs_4_phylo</span><span class='op'>(</span>max_per_org <span class='op'>=</span> <span class='fl'>1</span>, fill <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></code></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' aria-hidden='true' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>max_per_org</code></dt><dd><p>Maximum number of sequences of each organism to be taken from each cluster.</p></dd>

<dt><code>fill</code></dt><dd><p><code>logical</code>. If fill <code>DNAStringSet</code> with empty <code>DNAString</code> in cases where
<code>core_level</code> is set below 100%, and some clusters with missing organisms are also considered.</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='returns'><a class='anchor' aria-hidden='true' href='#returns'></a>Returns</h4>
<p>A <code>DNAStringSetList</code> with core genes. Order of organisms on each cluster is conserved, so it is easier
to concatenate them into a super-gene suitable for phylogenetic inference.</p>
<h3 class='hasAnchor' id='list-core-seqs-phylo-'><a class='anchor' aria-hidden='true' href='#list-core-seqs-phylo-'></a>Method <code>core_seqs_4_phylo()</code></h3>
<p>A field for obtaining core gene sequences is available (see below), but for creating a phylogeny with this
sets is useful to: 1) have the possibility of extracting just one sequence of each organism on each cluster, in
case paralogues are present, and 2) filling gaps with empty sequences in case the core_level was set below 100%,
allowing more genes (some not in 100% of organisms) to be incorporated to the phylogeny. That is the purpose of
this special function.</p><h4 class='hasAnchor' id='usage'><a class='anchor' aria-hidden='true' href='#usage'></a>Usage</h4>
<p><div class="r"></p><pre class='sourceCode r'><code><span class='va'>PgR6MS</span><span class='op'>$</span><span class='fu'>core_seqs_4_phylo</span><span class='op'>(</span>max_per_org <span class='op'>=</span> <span class='fl'>1</span>, fill <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></code></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' aria-hidden='true' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>max_per_org</code></dt><dd><p>Maximum number of sequences of each organism to be taken from each cluster.</p></dd>

<dt><code>fill</code></dt><dd><p><code>logical</code>. If fill <code>DNAStringSet</code> with empty <code>DNAString</code> in cases where
<code>core_level</code> is set below 100%, and some clusters with missing organisms are also considered.</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='returns'><a class='anchor' aria-hidden='true' href='#returns'></a>Returns</h4>
<p>A <code>DNAStringSetList</code> with core genes. Order of organisms on each cluster is conserved, so it is easier
to concatenate them into a super-gene suitable for phylogenetic inference.</p>
<p><hr>
<a id="method-clone"></a></p><h3 class='hasAnchor' id='method-'><a class='anchor' aria-hidden='true' href='#method-'></a>Method <code>clone()</code></h3>
<p>The objects of this class are cloneable with this method.</p><h4 class='hasAnchor' id='usage'><a class='anchor' aria-hidden='true' href='#usage'></a>Usage</h4>
<p><div class="r"></p><pre class='sourceCode r'><code><span class='va'>PgR6MS</span><span class='op'>$</span><span class='fu'>clone</span><span class='op'>(</span>deep <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></code></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' aria-hidden='true' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>deep</code></dt><dd><p>Whether to make a deep clone.</p></dd>

</dl><p></div></p>
<h3 class='hasAnchor' id='list-clone-'><a class='anchor' aria-hidden='true' href='#list-clone-'></a>Method <code>clone()</code></h3>
<p>The objects of this class are cloneable with this method.</p><h4 class='hasAnchor' id='usage'><a class='anchor' aria-hidden='true' href='#usage'></a>Usage</h4>
<p><div class="r"></p><pre class='sourceCode r'><code><span class='va'>PgR6MS</span><span class='op'>$</span><span class='fu'>clone</span><span class='op'>(</span>deep <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></code></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' aria-hidden='true' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>deep</code></dt><dd><p>Whether to make a deep clone.</p></dd>

</dl><p></div></p>



  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Ignacio Ferres, Gregorio Iraola, Institut Pasteur de Montevideo.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


