<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>1 - Quick Start • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="1 - Quick Start">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1-Quick_Start.html">1 - Quick Start</a>
    </li>
    <li>
      <a href="../articles/2-Input.html">2 - Input</a>
    </li>
    <li>
      <a href="../articles/3-Querying_Data.html">3 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/4-Subseting.html">4 - Subseting</a>
    </li>
    <li>
      <a href="../articles/5-Methods_Plots.html">5 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/6-Recipes.html">6 - Recipes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>1 - Quick Start</h1>
            
      
      
      <div class="hidden name"><code>1-Quick_Start.Rmd</code></div>

    </div>

    
    
<p>If you want a very quick look at <code>pagoo</code> and start playing with pangenome objects, this is a short tutorial to show the concept. Let’s start by loading a <em>Campylobacter spp.</em> dataset, included in the package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pagoo, <span class="dt">quietly =</span> <span class="ot">TRUE</span>, <span class="dt">warn.conflicts =</span> <span class="ot">FALSE</span>) <span class="co"># Load package</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">rds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'campylobacter.RDS'</span>, <span class="dt">package =</span> <span class="st">'pagoo'</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">campy &lt;-<span class="st"> </span><span class="kw">load_pangenomeRDS</span>(rds) <span class="co"># Load pangenome</span></a></code></pre></div>
<p>Now that the object (<code>campy</code>) is loaded, we can start by querying it. <code>pagoo</code> was developed considering that in a pangenome each individual <em>gene</em> belongs to a given <em>organism</em>, and is assigned to a <em>cluster</em> of orthologous. So those variables are interconnected, but each of them can have metadata asociated that is specific to each of them, i.e.: an individual gene can have coordinates inside a genome, but this doesn’t apply to a whole cluster, and a given organism has, for instance, a host where it was isolated from, but this information doesn’t apply to an individual gene.</p>
<div id="basic-fields" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-fields" class="anchor"></a>Basic Fields</h2>
<p>So this 3 variables are 3 separate tables that can be queried:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">campy<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 69 rows and 8 columns
##            org       id    strain      year  country     host   source
##       &lt;factor&gt; &lt;factor&gt;  &lt;factor&gt; &lt;integer&gt; &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1    16244_6_1     FR10 2006/367h      2006   France    Human      CSF
## 2    16244_6_2     FR11 2006/479h      2006   France    Human    Feces
## 3    16244_6_3     FR12 2006/588h      2006   France    Human      CSF
## 4    16244_6_4     FR13 2006/621h      2006   France    Human    Blood
## 5    16244_6_5     FR14 2006/649h      2006   France    Human    Feces
## ...        ...      ...       ...       ...      ...      ...      ...
## 65  16244_6_32      TR1     CF156      2013   Turkey    Human    Blood
## 66  17059_2_39      TW3      7035      2004   Taiwan    Human    Blood
## 67  17059_2_40      TW4    My5726      2006   Taiwan    Human    Blood
## 68  17059_2_41      TW5      1592      2006   Taiwan    Human    Blood
## 69  17059_2_42      TW6      1830      2008   Taiwan    Human    Blood
##     accession
##      &lt;factor&gt;
## 1   ERS672242
## 2   ERS672243
## 3   ERS672244
## 4   ERS672245
## 5   ERS672246
## ...       ...
## 65  ERS672273
## 66  ERS739258
## 67  ERS739259
## 68  ERS739260
## 69  ERS739261</code></pre>
<p><em>(Tip: To see all fields and methods, in any R console type <code>campy$</code> and press the [TAB] key two times.)</em></p>
<p>This dataset consist in 69 <em>Campylobacter spp</em> genomes. For each organism, you have a row with associated metadata. The first column, <code>org</code>, indicates the organism.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">campy<span class="op">$</span>clusters</a></code></pre></div>
<pre><code>## DataFrame with 3179 rows and 2 columns
##          cluster                   Pfam_Arch
##      &lt;character&gt;                 &lt;character&gt;
## 1      group0001                2-Hacid_dh_C
## 2      group0002     2-Hacid_dh_C;2-Hacid_dh
## 3      group0003 2-Hacid_dh_C;ACT;2-Hacid_dh
## 4      group0004             2Fe-2S_thioredx
## 5      group0005                   4HB_MCP_1
## ...          ...                         ...
## 3175   group3175                   zf-RING_7
## 3176   group3176                    zf-TFIIB
## 3177   group3177                        ZinT
## 3178   group3178                        ZnuA
## 3179   group3179                    ZT_dimer</code></pre>
<p>The <code>$clusters</code> field returns a table with metadata associated to each group of orthologous, in this case is the Pfam architecture domain (second column).</p>
<p>The last, and most important field is <code>$genes</code>, which returns a list of <code>DataFrame</code> with information given for each individual gene, grouped by cluster. We let the user to inspect this field by him/herself.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">campy<span class="op">$</span>genes</a></code></pre></div>
<p>The first 3 columns (<code>cluster</code>, <code>org</code>, and <code>gene</code>) are the glue that interconects each of 3 “variables”.</p>
<p>Another useful field is <code>$pan_matrix</code>, which returns a <code>matrix</code> with gene abundance for each cluster (columns), and each organism (rows).</p>
</div>
<div id="basic-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-methods" class="anchor"></a>Basic Methods</h2>
<p><code>pagoo</code> objects contain basic methods to analyse the pangenome, from general statistics to some basic plotting capabilities. Some of these methods can also take arguments.</p>
<p>For example:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">campy<span class="op">$</span><span class="kw">fluidity</span>(<span class="dt">n.sim =</span> <span class="dv">10</span>)</a></code></pre></div>
<pre><code>##       Mean        Std 
## 0.09937901 0.02612196</code></pre>
<p>Or:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">campy<span class="op">$</span><span class="kw">gg_barplot</span>()</a></code></pre></div>
<p><img src="1-Quick_Start_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="sequence-manipulation" class="section level2">
<h2 class="hasAnchor">
<a href="#sequence-manipulation" class="anchor"></a>Sequence Manipulation</h2>
<p>One of the main advantages of using <code>pagoo</code> is the ability to very easily manipulate sequences. This tutorial dataset doesn’t contain sequences, and they are an optional requirement. We strongly recommend to also giving a look to other <code>pagoo</code> tutorials in which sequences are used to understand the potential of using this package to work with pangenomes.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#basic-fields">Basic Fields</a></li>
      <li><a href="#basic-methods">Basic Methods</a></li>
      <li><a href="#sequence-manipulation">Sequence Manipulation</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
