<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2 - Input • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2 - Input">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1-Quick_Start.html">1 - Quick Start</a>
    </li>
    <li>
      <a href="../articles/2-Input.html">2 - Input</a>
    </li>
    <li>
      <a href="../articles/3-Querying_Data.html">3 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/4-Subseting.html">4 - Subseting</a>
    </li>
    <li>
      <a href="../articles/5-Methods_Plots.html">5 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/6-Recipes.html">6 - Recipes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>2 - Input</h1>
            
      
      
      <div class="hidden name"><code>2-Input.Rmd</code></div>

    </div>

    
    
<div id="input-from-scratch" class="section level1">
<h1 class="hasAnchor">
<a href="#input-from-scratch" class="anchor"></a>Input From Scratch</h1>
<p>In order to generate a <code>pagoo</code> object, the only mandatory data structure is a <code>data.frame</code> which has to contain 3 basic columns. Let’s ilustrate it with an toy dataset.</p>
<p>First load <code>pagoo</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pagoo) <span class="co"># Load package</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">tgz &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'toy_data.tar.gz'</span>, <span class="dt">package =</span> <span class="st">'pagoo'</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/untar.html">untar</a></span>(<span class="dt">tarfile =</span> tgz, <span class="dt">exdir =</span> <span class="kw"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>()) <span class="co"># Decompress example dataset</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(), <span class="dt">full.names =</span> <span class="ot">TRUE</span>, <span class="dt">pattern =</span> <span class="st">'tsv$|fasta$'</span>) <span class="co"># List files</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">files</a></code></pre></div>
<pre><code>## [1] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/case_clusters_meta.tsv"
## [2] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/case_df.tsv"           
## [3] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/case_orgs_meta.tsv"    
## [4] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/organismA.fasta"       
## [5] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/organismB.fasta"       
## [6] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/organismC.fasta"       
## [7] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/organismD.fasta"       
## [8] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//Rtmpsdmipe/organismE.fasta"</code></pre>
<p>The file we need now is <code>case_df.tsv</code>. Lets load it and see what’s its structure:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">data_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"case_df.tsv"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">data &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(data_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">quote =</span> <span class="st">''</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(data)</a></code></pre></div>
<pre><code>##      gene       org cluster                             annot
## 1 gene081 organismA   OG001  Thioesterase superfamily protein
## 2 gene122 organismB   OG001          Thioesterase superfamily
## 3 gene299 organismC   OG001  Thioesterase superfamily protein
## 4 gene186 organismD   OG001  Thioesterase superfamily protein
## 5 gene076 organismE   OG001          Thioesterase superfamily
## 6 gene352 organismA   OG002 Inherit from proNOG: Thioesterase</code></pre>
<p>So it is a <code>data.frame</code> with 4 columns. The first one with the name of each gene, the second one with the organism to which each gene belongs, the third one with the cluster to which each gene was assigned in the pangenome reconstruction, and the last one with annotation metadata for each gene. Of the 4 columns, the former 3 are required, and <code>pagoo</code> will look for columns named “gene”, “org”, and “cluster”. More columns are optional, and you can add as many as you want (or none) to add metadata of each gene.</p>
<p>With only this data (even ignoring the fourth column, which is metadata), you can start working with pagoo:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">pg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pagoo.html">pagoo</a></span>(<span class="dt">data =</span> data)</a></code></pre></div>
<p>but let’s continue adding more.</p>
<div id="including-organism-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#including-organism-metadata" class="anchor"></a>Including organism metadata</h2>
<p>The next 2 <code>.tsv</code> files contains metadata for each cluster and for each organism, respectively, and are optional arguments.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># Organism metadata</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">orgs_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"case_orgs_meta.tsv"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">orgs_meta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(orgs_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">quote =</span> <span class="st">''</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(orgs_meta)</a></code></pre></div>
<pre><code>##         org sero  country
## 1 organismA    a Westeros
## 2 organismB    b Westeros
## 3 organismC    c Westeros
## 4 organismD    a    Essos
## 5 organismE    b    Essos</code></pre>
<p>On this <code>data.frame</code> we have a column named <code>org</code> which is mandatory in case you provide this argument. Other columns are metadata asociated to each organism. Beware that organisms provided in this table (<code>orgs_meta$org</code>) must concide with the names provided in the <code>data$org</code> field, in order to correctly map each variables.</p>
</div>
<div id="including-cluster-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#including-cluster-metadata" class="anchor"></a>Including cluster metadata</h2>
<p>Last file contains metadata asociated with each cluster of orthologous:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># Cluster metadata</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">clust_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"case_clusters_meta.tsv"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">clust_meta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(clust_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">quote =</span> <span class="st">''</span>)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(clust_meta)</a></code></pre></div>
<pre><code>##   cluster   kegg  cog
## 1   OG001   &lt;NA&gt;    S
## 2   OG002   &lt;NA&gt;    S
## 3   OG003   &lt;NA&gt; &lt;NA&gt;
## 4   OG004   &lt;NA&gt;    D
## 5   OG005 K01990    V
## 6   OG006   &lt;NA&gt;    V</code></pre>
<p>Again, the column <code>clust_meta$cluster</code> must contain the same elements as <code>data$cluster</code> column to be able to map one into the other.</p>
<p>With all this data the <code>pagoo</code> object will look much more complete. But you can still add sequence information to the pangenome, which makes it much more useful and interesting to work with.</p>
</div>
<div id="including-sequences" class="section level2">
<h2 class="hasAnchor">
<a href="#including-sequences" class="anchor"></a>Including sequences</h2>
<p>In this made up dataset we have 5 organisms, so if you decide to add sequences to the pangenome you must provide them for all 5 organisms. The type of data needed is a DNA multifasta file for each organism, in which each sequence is a gene whose name can be mapped to the <code>data$gene</code> column. You must first load the sequences into a <code>list</code>, and name each list element as the organism provided in <code>data$org</code> (as well as <code>org_meta$org</code>). The <code>list</code> would look something like:</p>
<ol style="list-style-type: decimal">
<li>organism1
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneN</li>
</ul>
</li>
<li>organism2
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneM</li>
</ul>
</li>
<li>organism3
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneP</li>
</ul>
</li>
<li>…</li>
</ol>
<p>In the case of the example we are working on:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># Sequences</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">fasta_files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"[.]fasta"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>) <span class="co"># List fasta files</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(fasta_files) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">'[.]fasta'</span>, <span class="st">''</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(fasta_files)) <span class="co"># Name them </span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co"># Read fasta files with Biostrings:</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Biostrings)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">sq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(fasta_files, readDNAStringSet)</a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(sq) <span class="co"># Is list?</span></a></code></pre></div>
<pre><code>## [1] "list"</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(sq) <span class="co"># One list element per organism</span></a></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(sq) <span class="co"># Names are the same as in data$org</span></a></code></pre></div>
<pre><code>## [1] "organismA" "organismB" "organismC" "organismD" "organismE"</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(sq[[<span class="dv">1</span>]]) <span class="co"># Class of each element of the list</span></a></code></pre></div>
<pre><code>## [1] "DNAStringSet"
## attr(,"package")
## [1] "Biostrings"</code></pre>
<p>And we have a <code>list</code> of <code>DNAStringSet</code> (Biostrings package). Now we can load a quite complete <code>pagoo</code> object (you could still add more metadata to genes, clusters, or organisms):</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pagoo.html">pagoo</a></span>(<span class="dt">data =</span> data, <span class="co"># Required data</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2">           <span class="dt">org_meta =</span> orgs_meta, <span class="co"># Organism's metadata</span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3">           <span class="dt">cluster_meta =</span> clust_meta, <span class="co"># Cluster's metadata</span></a>
<a class="sourceLine" id="cb18-4" data-line-number="4">           <span class="dt">sequences =</span> sq) <span class="co"># Sequences</span></a></code></pre></div>
<p>All this stuff with preparing data and loading classes seems difficult and time-consuming, but in real life working datasets this will be rarely needed. We are explaning it here to provide full details about how the software works, but this package also provides functions to automatically read-in output files from pangenome reconstruction softwares into <code>pagoo</code>, avoiding any formating or manipulation of data.</p>
<p>For pangenome reconstruction you can use the software of your choice. For example, since last years &lt;a <a href="https://sanger-pathogens.github.io/Roary/">roary</a> (Page et al., 2015) has been the standard and most cited software for pangenome reconstruction. More recently, other good software has emerged like <a href="http://pangenome.tuebingen.mpg.de">PanX</a> (Ding et al., 2018), <a href="https://github.com/SionBayliss/PIRATE">PRIATE</a> (Bayliss et al., 2019), <a href="https://github.com/cran/micropan">micropan</a> (Snipen &amp; Liland, 2015), <a href="https://github.com/zheminzhou/PEPPA">PEPPA</a> (Zhou &amp; Achtman, 2020), <a href="https://github.com/gtonkinhill/panaroo">panaroo</a> (Tonkin-Hill et al., 2020), among others. Also, we have created our own pangenome reconstruction software called <a href="https://github.com/iferres/pewit">pewit</a> (Ferrés et al., still unpublished), which automatically generates a <code>pagoo</code> object to perform downstream analyses. To use <code>pagoo</code> from the outputs produced by most of the other software tools mentioned above, we provide specific functions to read-in pangenome data.</p>
</div>
</div>
<div id="adding-metadata-after-object-creation" class="section level1">
<h1 class="hasAnchor">
<a href="#adding-metadata-after-object-creation" class="anchor"></a>Adding Metadata After Object Creation</h1>
<p>After object creation, you may want to add new metadata given new information or as result of posterior analyses. <code>pagoo</code> objects include a function to add columns of metadata to each gene, each cluster, or each organism. To illustrate it, we will add a new column to the <code>$organisms</code> field named <code>host</code> to add made up information about the host where each genome was isolated from.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">host_df &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">org =</span> p<span class="op">$</span>organisms<span class="op">$</span>org, <span class="dt">host =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Cow"</span>, <span class="st">"Dog"</span>, <span class="st">"Cat"</span>, <span class="st">"Cow"</span>, <span class="st">"Sheep"</span>))</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">p<span class="op">$</span><span class="kw">add_metadata</span>(<span class="dt">map =</span> <span class="st">"org"</span>, host_df)</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">p<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 4 columns
##         org     sero  country     host
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros      Cow
## 2 organismB        b Westeros      Dog
## 3 organismC        c Westeros      Cat
## 4 organismD        a    Essos      Cow
## 5 organismE        b    Essos    Sheep</code></pre>
<p>In order to allow <code>pagoo</code> correct data mapping, the values in the first column of the metadata table should be the same as <code>p$organisms$org</code>, and its column header must also be named <code>org</code>.</p>
<p>As said, you can add <code>gene</code> or <code>cluster</code> metadata following the same idea.</p>
</div>
<div id="saving-and-loading-pagoo-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#saving-and-loading-pagoo-objects" class="anchor"></a>Saving and Loading <code>pagoo</code> objects</h1>
<p>Once loaded, it has two methods for saving and reloading to a new R session. The first one is by saving them as flat (text) files:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">p<span class="op">$</span><span class="kw">write_pangenome</span>(<span class="dt">dir =</span> <span class="st">"pangenome"</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="st">"pangenome"</span>)</a></code></pre></div>
<pre><code>## [1] "clusters.tsv"  "data.tsv"      "organisms.tsv"</code></pre>
<p>This creates a directory with 3 text files. The advantage of this approach is that you can analyze it outside R, the disadvantage is that from a reproducibility point of view reading text could be less stable since class or number precision can be lost, and also you can’t save the state of the object in any given time. Only available organisms/genes/clusters are saved, and if you reload the class using the tsv files, any previously dropped organisms/gene/cluster won’t be available any more. (For information about dropping/recovering organisms, see “Subset” tutorial).</p>
<p>If you want to save the object and continue working with it in other R session, we recommend to save them as R objects with the RDS methods provided:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">p<span class="op">$</span><span class="kw">save_pangenomeRDS</span>(<span class="dt">file =</span> <span class="st">"pangenome.RDS"</span>)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2">p2 &lt;-<span class="st"> </span><span class="kw">load_pangenomeRDS</span>(<span class="st">"pangenome.RDS"</span>)</a></code></pre></div>
<p>This method is more stable (compatible between <code>pagoo</code> versions), secure (uses the same metadata classes, and precision isn’t lost), and convinient (the exact state of the object saved is restored, keeping dropped organisms/genes/clusters hidden, available to be recovered, and other object state configuration, e.g. <code>core_level</code>, is also saved).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#input-from-scratch">Input From Scratch</a><ul class="nav nav-pills nav-stacked">
<li><a href="#including-organism-metadata">Including organism metadata</a></li>
      <li><a href="#including-cluster-metadata">Including cluster metadata</a></li>
      <li><a href="#including-sequences">Including sequences</a></li>
      </ul>
</li>
      <li><a href="#adding-metadata-after-object-creation">Adding Metadata After Object Creation</a></li>
      <li><a href="#saving-and-loading-pagoo-objects">Saving and Loading <code>pagoo</code> objects</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
