<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3 - Querying Data • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3 - Querying Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1-Quick_Start.html">1 - Quick Start</a>
    </li>
    <li>
      <a href="../articles/2-Input.html">2 - Input</a>
    </li>
    <li>
      <a href="../articles/3-Querying_Data.html">3 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/4-Subseting.html">4 - Subseting</a>
    </li>
    <li>
      <a href="../articles/5-Methods_Plots.html">5 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/6-Recipes.html">6 - Recipes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>3 - Querying Data</h1>
            
      
      
      <div class="hidden name"><code>3-Querying_Data.Rmd</code></div>

    </div>

    
    
<p>In this section we will start exploring what is stored inside the <code>pagoo</code> object and how we can access this information. Keep in mind that this object has its own associated data and methods that can be easily queried with the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> operator. These methods allow for the rapid subsetting, extraction and visualization of pangenome data.</p>
<p>First of all, we will load a pangenome using a toy dataset included in the package. This is the same pangenome used in the “Input” tutorial, just loaded in a more convinient way.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pagoo) <span class="co"># Load package</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">toy_rds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'toy_data.RDS'</span>, <span class="dt">package =</span> <span class="st">'pagoo'</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">p &lt;-<span class="st"> </span><span class="kw">load_pangenomeRDS</span>(toy_rds)</a></code></pre></div>
<div id="summary-statistics" class="section level1">
<h1 class="hasAnchor">
<a href="#summary-statistics" class="anchor"></a>Summary statistics</h1>
<p>A pangenome can be stratified in different gene subsets according to their frequency in the dataset. The <code>core genes</code> can be defined as those present in all or almost every genome (typically 95-100%). The remaining genes are defined as the accessory genome, that can be subdivided in <code>cloud genes</code> or singletons (present in one genome or in genomes that are identical) and <code>shell genes</code> which are those in the middle. Let’s see this using <code>pagoo</code>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">p<span class="op">$</span>summary_stats</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total       600
## 2        Core       254
## 3       Shell       118
## 4       Cloud       228</code></pre>
<div id="core-level" class="section level2">
<h2 class="hasAnchor">
<a href="#core-level" class="anchor"></a>Core level</h2>
<p>The <code>core level</code> defines the minimum number of genomes (as a percentage) in which a certain gene should be present to be considered a core gene. By default, <code>pagoo</code> considers as core all genes present in at least 95% of organisms. The core level can be modified to be more or less stringent defining the core genome. This feature exemplifies R6’s reference semantics, since modifying the core level will affect the pangenome object state resulting in different core, shell and cloud sets. Have a look:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">p<span class="op">$</span>core_level       <span class="co"># Default</span></a></code></pre></div>
<pre><code>## [1] 95</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">p<span class="op">$</span>core_level &lt;-<span class="st"> </span><span class="dv">99</span> <span class="co"># Change value</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">p<span class="op">$</span>summary_stats    <span class="co"># Updated object</span></a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total       600
## 2        Core       254
## 3       Shell       118
## 4       Cloud       228</code></pre>
<p>As you can see, changing the core level from 95% to a more stringent 90% cause an increase in the number of core genes from 1557 to 1613, and a concomitant decrease in shell genes from 1341 to 1285. This means that 56 genes migrated from shell to core when relaxing the core level to 90%. Now this changes remain in the object for subsequent analysis, or can be reverted by setting the core level again at the original value.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">p<span class="op">$</span>core_level &lt;-<span class="st"> </span><span class="dv">95</span></a></code></pre></div>
</div>
</div>
<div id="pangenome-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#pangenome-matrix" class="anchor"></a>Pangenome matrix</h1>
<p>The pangenome matrix is one of the most useful things when analyzing pangenomes. Typically, it represents organisms in rows and clusters of orthologous in columns informing about gene abundance (considering paralogues). The pangenome matrix looks like this (printing only first 5 columns):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">p<span class="op">$</span>pan_matrix[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##           OG001 OG002 OG003 OG004 OG005
## organismA     1     1     1     1     1
## organismB     1     1     0     1     1
## organismC     1     1     1     1     1
## organismD     1     1     1     1     1
## organismE     1     0     0     1     1</code></pre>
</div>
<div id="gene-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-metadata" class="anchor"></a>Gene metadata</h1>
<p>Individual gene metadata can be accessed by using the <code>$genes</code> suffix. It always contains the gene name, the organism to which it belongs, the cluster to where it was assigned, and a gene identifier (<code>gid</code>) that is mainly used internally to organize the data. Also, it may typically include annotation data, genomic coordinates, etc, but this other metadata is optional. Gene metadata is splited by cluster, so it consist in a <code>List</code> of <code>DataFrame</code>s.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">p<span class="op">$</span>genes</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 600
## $OG001
## DataFrame with 5 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG001 organismA  gene081 organismA__gene081
## 2    OG001 organismB  gene122 organismB__gene122
## 3    OG001 organismC  gene299 organismC__gene299
## 4    OG001 organismD  gene186 organismD__gene186
## 5    OG001 organismE  gene076 organismE__gene076
##                              annot
##                           &lt;factor&gt;
## 1 Thioesterase superfamily protein
## 2         Thioesterase superfamily
## 3 Thioesterase superfamily protein
## 4 Thioesterase superfamily protein
## 5         Thioesterase superfamily
## 
## $OG002
## DataFrame with 4 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG002 organismA  gene352 organismA__gene352
## 2    OG002 organismB  gene145 organismB__gene145
## 3    OG002 organismC  gene251 organismC__gene251
## 4    OG002 organismD  gene299 organismD__gene299
##                               annot
##                            &lt;factor&gt;
## 1 Inherit from proNOG: Thioesterase
## 2 Inherit from proNOG: Thioesterase
## 3 Inherit from proNOG: Thioesterase
## 4 Inherit from proNOG: Thioesterase
## 
## $OG003
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG003 organismA  gene359 organismA__gene359       NA
## 2    OG003 organismC  gene237 organismC__gene237       NA
## 3    OG003 organismD  gene069 organismD__gene069       NA
## 
## ...
## &lt;597 more elements&gt;</code></pre>
<p>If you want to work with this data as a single <code>DataFrame</code>, just <code>unlist</code> it:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(p<span class="op">$</span>genes, <span class="dt">use.names =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## DataFrame with 1796 rows and 5 columns
##       cluster       org     gene                gid
##      &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1       OG001 organismA  gene081 organismA__gene081
## 2       OG001 organismB  gene122 organismB__gene122
## 3       OG001 organismC  gene299 organismC__gene299
## 4       OG001 organismD  gene186 organismD__gene186
## 5       OG001 organismE  gene076 organismE__gene076
## ...       ...       ...      ...                ...
## 1792    OG596 organismD  gene205 organismD__gene205
## 1793    OG597 organismD  gene344 organismD__gene344
## 1794    OG598 organismE  gene123 organismE__gene123
## 1795    OG599 organismE  gene216 organismE__gene216
## 1796    OG600 organismE  gene104 organismE__gene104
##                                 annot
##                              &lt;factor&gt;
## 1    Thioesterase superfamily protein
## 2            Thioesterase superfamily
## 3    Thioesterase superfamily protein
## 4    Thioesterase superfamily protein
## 5            Thioesterase superfamily
## ...                               ...
## 1792      Integrase catalytic subunit
## 1793                Methyltransferase
## 1794     von Willebrand factor type A
## 1795                               NA
## 1796                               NA</code></pre>
<p><code>pagoo</code> also includes predefined subsets fields to list only certain pangenome category, these are queried by adding a prefix with the desired category followed by an underscore: <code>$core_genes</code>, <code>$shell_genes</code>, and <code>$cloud_genes</code>. These kind of subsets are better explaied in the ‘4 - Subets’ tutorial, and also apply to other pangenome data described below.</p>
</div>
<div id="clusters-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#clusters-metadata" class="anchor"></a>Clusters metadata</h1>
<p>Groups of orthologues (clusters) are also stored in <code>pagoo</code> objects as a table with a cluster identifier per row, and optional metadata associated as additional columns.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">p<span class="op">$</span>clusters</a></code></pre></div>
<pre><code>## DataFrame with 600 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG001       NA        S
## 2         OG002       NA        S
## 3         OG003       NA       NA
## 4         OG004       NA        D
## 5         OG005   K01990        V
## ...         ...      ...      ...
## 596       OG596       NA        L
## 597       OG597       NA        H
## 598       OG598       NA        S
## 599       OG599       NA       NA
## 600       OG600       NA       NA</code></pre>
<p>Subsets also exists for this field: <code>$core_clusters</code>, <code>$shell_clusters</code>, and <code>$cloud_clusters</code>.</p>
</div>
<div id="sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#sequences" class="anchor"></a>Sequences</h1>
<p>Although is an optional field (it exists only if user provide this data as an argument when object is created), <code>$sequences</code> gives access to sequence data. Sequences are stored as a <code>List</code> of <code>DNAStringSet</code> (a.k.a <code>DNAStringSetList</code>, Biostrings package), grouped by cluster.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">p<span class="op">$</span>sequences             <span class="co"># List all sequences grouped by cluster</span></a></code></pre></div>
<pre><code>## DNAStringSetList of length 600
## [["OG001"]] organismA__gene081=ATGGAATCGGAAACAATCTTTCGGGAAATCAAAGCAAGCCAAAACG...
## [["OG002"]] organismA__gene352=ATGACAGATATACAGATTGAATTTCCGGAGAACTATCATTTTTCAA...
## [["OG003"]] organismA__gene359=TTGTTTTTTTATTTCACATTATGGATCACAAAAAGGATCGGTTTTC...
## [["OG004"]] organismA__gene258=ATGAAACAGATCCTCTGTATTGCAAACCAGAAAGGCGGAGTCGGTA...
## [["OG005"]] organismA__gene118=ATGATCAAAGTTAAAAACCTATCTAAATTCTATGGCAAAAAACTCG...
## [["OG006"]] organismA__gene348=GTGAGTCTGATCAAAGTAAGAAATTTGGTAAAGAATTATCATATTC...
## [["OG007"]] organismA__gene140=ATGAAGTTTTTTTTCAGACTTCTATCCTATTCCGTACATTACAAAT...
## [["OG008"]] organismB__gene075=ATGATTCAAAAAGCGAATTATAAACAGATTTTATTAGAATTTTTCT...
## [["OG009"]] organismB__gene328=ATGAATGATTATCTCACGTACGTATCTGACAATCGTTCCGTTCGAA...
## [["OG010"]] organismA__gene013=ATGAGCGAAATTGAAAGCGGTTTTTTTCAATCCGGCGGGTACAATC...
## ...
## &lt;590 more elements&gt;</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">p<span class="op">$</span>sequences[[<span class="st">"OG001"</span>]]  <span class="co"># List first cluster</span></a></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 5
##     width seq                                               names               
## [1]   576 ATGGAATCGGAAACAATCTTTCG...AACTTCAAAAGACGGATTCCTGA organismA__gene081
## [2]   576 ATGGAACCAGAGGCAATCTACCG...AACTTCAAAACATAGAACCCTAA organismB__gene122
## [3]   576 ATGGATACGGAAACGATCTACCG...AACTTCAAAAGGTAAATTCTTAA organismC__gene299
## [4]   576 ATGGAAACAGAAACAATCTACCG...AACTTCAAAAGGTAAATTCCTAG organismD__gene186
## [5]   576 ATGGAATCAGAGGCAATCTACCG...AACTTCAAAAGATAATTTCCTAG organismE__gene076</code></pre>
<p>Note that sequence names are created by pasteing organism names and gene names, separated by a string that by default is <code>sep = '__'</code> (two underscores). This are the same as the <code>gid</code> column in the <code>$genes</code> field, and are initially set when <code>pagoo</code> object is created. If you think your dataset contain names with this separator, then you should set this parameter to other string to avoid conflicts. <code>$sequences</code> field also has predefied subsets: <code>$core_sequences</code>, <code>$shell_sequences</code>, and <code>$cloud_sequences</code>.</p>
</div>
<div id="organism-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#organism-metadata" class="anchor"></a>Organism metadata</h1>
<p>The <code>$organisms</code> field contain a table with organisms and metadata as additional columns if provided.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">p<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#summary-statistics">Summary statistics</a><ul class="nav nav-pills nav-stacked">
<li><a href="#core-level">Core level</a></li>
      </ul>
</li>
      <li><a href="#pangenome-matrix">Pangenome matrix</a></li>
      <li><a href="#gene-metadata">Gene metadata</a></li>
      <li><a href="#clusters-metadata">Clusters metadata</a></li>
      <li><a href="#sequences">Sequences</a></li>
      <li><a href="#organism-metadata">Organism metadata</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
