<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>4 - Subseting • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="4 - Subseting">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1-Quick_Start.html">1 - Quick Start</a>
    </li>
    <li>
      <a href="../articles/2-Input.html">2 - Input</a>
    </li>
    <li>
      <a href="../articles/3-Querying_Data.html">3 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/4-Subseting.html">4 - Subseting</a>
    </li>
    <li>
      <a href="../articles/5-Methods_Plots.html">5 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/6-Recipes.html">6 - Recipes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>4 - Subseting</h1>
            
      
      
      <div class="hidden name"><code>4-Subseting.Rmd</code></div>

    </div>

    
    
<p>One fundamental operation when working with pangenomic data is to subset it in order to better analyze it. <code>pagoo</code> provides three convinent ways that can help in different scenarios. First we will see predefined subsets, then we will understand how to use the classic R’s <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> operator with <code>pagoo</code> objects, and finally we will see how to temporarily remove one or more organisms from the dataset, and then to recover them.</p>
<p>Start the tutorial by loading the included pangenome.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pagoo) <span class="co"># Load package</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">toy_rds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'toy_data.RDS'</span>, <span class="dt">package =</span> <span class="st">'pagoo'</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">p &lt;-<span class="st"> </span><span class="kw">load_pangenomeRDS</span>(toy_rds)</a></code></pre></div>
<div id="predefined-subsets" class="section level1">
<h1 class="hasAnchor">
<a href="#predefined-subsets" class="anchor"></a>Predefined Subsets</h1>
<p>Predefined subsets were already superficially seen in <strong>3 - Querying Data</strong> tutorial. Here we will review them and understand its notation.</p>
<table class="table">
<thead><tr class="header">
<th>-</th>
<th><code>$core_*</code></th>
<th><code>$shell_*</code></th>
<th><code>$cloud_*</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong><code>$*_genes</code></strong></td>
<td><code>$core_genes</code></td>
<td><code>$shell_genes</code></td>
<td><code>$cloud_genes</code></td>
</tr>
<tr class="even">
<td><strong><code>$*_clusters</code></strong></td>
<td><code>$core_clusters</code></td>
<td><code>$shell_clusters</code></td>
<td><code>$cloud_clusters</code></td>
</tr>
<tr class="odd">
<td><strong><code>$*_sequences</code></strong></td>
<td><code>$core_sequences</code></td>
<td><code>$shell_sequences</code></td>
<td><code>$cloud_sequences</code></td>
</tr>
</tbody>
</table>
<p>As seen in the above table, the notation is quite straightforward. Lets take <code>clusters</code> for instance to illustrate it better.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">p<span class="op">$</span>clusters</a></code></pre></div>
<pre><code>## DataFrame with 600 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG001       NA        S
## 2         OG002       NA        S
## 3         OG003       NA       NA
## 4         OG004       NA        D
## 5         OG005   K01990        V
## ...         ...      ...      ...
## 596       OG596       NA        L
## 597       OG597       NA        H
## 598       OG598       NA        S
## 599       OG599       NA       NA
## 600       OG600       NA       NA</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">p<span class="op">$</span>core_clusters</a></code></pre></div>
<pre><code>## DataFrame with 254 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG001       NA        S
## 2         OG002       NA        S
## 3         OG004       NA        D
## 4         OG005   K01990        V
## 5         OG006       NA        V
## ...         ...      ...      ...
## 250       OG437       NA       NA
## 251       OG439       NA        S
## 252       OG444       NA       NA
## 253       OG447       NA       NA
## 254       OG454       NA        L</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">p<span class="op">$</span>shell_clusters</a></code></pre></div>
<pre><code>## DataFrame with 118 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG003       NA       NA
## 2         OG009       NA       NA
## 3         OG012       NA        S
## 4         OG029       NA       NA
## 5         OG031       NA        I
## ...         ...      ...      ...
## 114       OG490       NA       NA
## 115       OG492       NA       NA
## 116       OG493       NA       NA
## 117       OG496       NA       NA
## 118       OG497   K07098        S</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">p<span class="op">$</span>cloud_clusters</a></code></pre></div>
<pre><code>## DataFrame with 228 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG008       NA        V
## 2         OG024       NA        S
## 3         OG028       NA        S
## 4         OG030       NA       N,
## 5         OG033       NA        O
## ...         ...      ...      ...
## 224       OG596       NA        L
## 225       OG597       NA        H
## 226       OG598       NA        S
## 227       OG599       NA       NA
## 228       OG600       NA       NA</code></pre>
</div>
<div id="the-operator" class="section level1">
<h1 class="hasAnchor">
<a href="#the-operator" class="anchor"></a>The <code>[</code> operator</h1>
<p>Using <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> for subset vectors, lists, or matrices in R is one of the most common operations we, as R users, do every day. Methods using <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> for subsetting can be divided into 2 types:</p>
<ol style="list-style-type: decimal">
<li>Subsetting vector or list like objects.</li>
<li>Subsetting matrix-like objects.</li>
</ol>
<p>In the following subsections we will see how to use them to subset <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, or<code>$organisms</code> fields.</p>
<div id="vectorlist-notation" class="section level2">
<h2 class="hasAnchor">
<a href="#vectorlist-notation" class="anchor"></a>Vector/List notation</h2>
<p>When we subset a vector or list in R we use a single number or a vector of numbers, lets say from <code>i</code> to <code>j</code>, to pick those elements by indexes: <code>x[i:j]</code>. In <code>pagoo</code> we implement a method for this generic function to subset data fields, where indexes represent <strong>clusters</strong>. So instead of subseting directly the final object (which you can of course) you apply this method directly to the object, and then select any of the data fields.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># From clusters 1 to 3</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>pan_matrix</a></code></pre></div>
<pre><code>##           OG001 OG002 OG003
## organismA     1     1     1
## organismB     1     1     0
## organismC     1     1     1
## organismD     1     1     1
## organismE     1     0     0</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>genes</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 3
## $OG001
## DataFrame with 5 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG001 organismA  gene081 organismA__gene081
## 2    OG001 organismB  gene122 organismB__gene122
## 3    OG001 organismC  gene299 organismC__gene299
## 4    OG001 organismD  gene186 organismD__gene186
## 5    OG001 organismE  gene076 organismE__gene076
##                              annot
##                           &lt;factor&gt;
## 1 Thioesterase superfamily protein
## 2         Thioesterase superfamily
## 3 Thioesterase superfamily protein
## 4 Thioesterase superfamily protein
## 5         Thioesterase superfamily
## 
## $OG002
## DataFrame with 4 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG002 organismA  gene352 organismA__gene352
## 2    OG002 organismB  gene145 organismB__gene145
## 3    OG002 organismC  gene251 organismC__gene251
## 4    OG002 organismD  gene299 organismD__gene299
##                               annot
##                            &lt;factor&gt;
## 1 Inherit from proNOG: Thioesterase
## 2 Inherit from proNOG: Thioesterase
## 3 Inherit from proNOG: Thioesterase
## 4 Inherit from proNOG: Thioesterase
## 
## $OG003
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG003 organismA  gene359 organismA__gene359       NA
## 2    OG003 organismC  gene237 organismC__gene237       NA
## 3    OG003 organismD  gene069 organismD__gene069       NA</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>clusters</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##       cluster     kegg      cog
##   &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1       OG001       NA        S
## 2       OG002       NA        S
## 3       OG003       NA       NA</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>sequences</a></code></pre></div>
<pre><code>## DNAStringSetList of length 3
## [["OG001"]] organismA__gene081=ATGGAATCGGAAACAATCTTTCGGGAAATCAAAGCAAGCCAAAACG...
## [["OG002"]] organismA__gene352=ATGACAGATATACAGATTGAATTTCCGGAGAACTATCATTTTTCAA...
## [["OG003"]] organismA__gene359=TTGTTTTTTTATTTCACATTATGGATCACAAAAAGGATCGGTTTTC...</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
<p>Note that in each case, indexes can be interpreted as clusters. In <code>pan_matrix</code> they refer to columns, in <code>genes</code> and <code>sequences</code> to elements in a list, in <code>clusters</code> rows in the dataframe, and in the case of <code>$organisms</code> it returns the organisms (rows) where those clusters are present. See what happen for this last case when we query for organisms present for a shell cluster:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">shell_clust &lt;-<span class="st"> </span>p<span class="op">$</span>shell_clusters<span class="op">$</span>cluster[<span class="dv">1</span>] <span class="co"># [1] "OG008"</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">p[shell_clust]<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos</code></pre>
<p>Only (3) organisms which contain that shell cluster will be listed.</p>
</div>
<div id="matrix-notation" class="section level2">
<h2 class="hasAnchor">
<a href="#matrix-notation" class="anchor"></a>Matrix notation</h2>
<p>The other use of <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> notation is when we subset a matrix-like object. In this case we provide 2 sets of indexes: the first for rows, and the second for columns, separated by a coma. <code>pagoo</code> interprets these indexes as organisms (rows) and clusters (columns), so you are referencing a set of genes (cell value). Basically you are referencing cells in the pan_matrix.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">p<span class="op">$</span>pan_matrix[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">125</span><span class="op">:</span><span class="dv">127</span>]</a></code></pre></div>
<pre><code>##           OG125 OG126 OG127
## organismA     1     1     1
## organismB     0     0     1
## organismC     1     0     1</code></pre>
<p>Here we are selecting the first 3 organisms, and the clusters indexed from the 125 to the 127. Note that we see (sum columns) 2 genes in the first cluster, 1 in the second, and 3 in the third. Now we flip the notation and subset directly the <code>pagoo</code> object, and then ask for a data field:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">125</span><span class="op">:</span><span class="dv">127</span>]<span class="op">$</span>pan_matrix <span class="co"># The same as above</span></a></code></pre></div>
<pre><code>##           OG125 OG126 OG127
## organismA     1     1     1
## organismB     0     0     1
## organismC     1     0     1</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">125</span><span class="op">:</span><span class="dv">127</span>]<span class="op">$</span>genes</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 3
## $OG125
## DataFrame with 2 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG125 organismA  gene072 organismA__gene072
## 2    OG125 organismC  gene042 organismC__gene042
##                                   annot
##                                &lt;factor&gt;
## 1 Tetr family transcriptional regulator
## 2 Tetr family transcriptional regulator
## 
## $OG126
## DataFrame with 1 row and 5 columns
##    cluster       org     gene                gid                     annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;                  &lt;factor&gt;
## 1    OG126 organismA  gene341 organismA__gene341 Transcriptional regulator
## 
## $OG127
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid                        annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;                     &lt;factor&gt;
## 1    OG127 organismA  gene087 organismA__gene087 acetyl-coa acetyltransferase
## 2    OG127 organismB  gene234 organismB__gene234 acetyl-coa acetyltransferase
## 3    OG127 organismC  gene178 organismC__gene178 acetyl-coa acetyltransferase</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">125</span><span class="op">:</span><span class="dv">127</span>]<span class="op">$</span>clusters</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##       cluster     kegg      cog
##   &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1       OG125       NA        K
## 2       OG126       NA        K
## 3       OG127       NA        I</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">125</span><span class="op">:</span><span class="dv">127</span>]<span class="op">$</span>sequences</a></code></pre></div>
<pre><code>## DNAStringSetList of length 3
## [["OG125"]] organismA__gene072=ATGAAAATCTCCGCAACAGAATCGGCAACAAAACGAGAGAAGATCC...
## [["OG126"]] organismA__gene341=ATGGCTCAAAATTTCGACAAACAAAGTAATCCGTACGATCGCCTGA...
## [["OG127"]] organismA__gene087=ATGTCAAACGCCTATGTAATCGACGCCGTTAGAACTCCGAGAGGAA...</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">p[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">125</span><span class="op">:</span><span class="dv">127</span>]<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros</code></pre>
<p>So this selection returns a list of length 3, with 2, 1, and 3 elements respectively for <code>$genes</code> and <code>$sequences</code>, a dataframe showing only the selected clusters in the <code>$clusters</code> field, and a dataframe showing only the selected organisms in the <code>$organisms</code> field.</p>
<p>We found this implemetation quite useful for both data exploration and fine grained analysis.</p>
</div>
</div>
<div id="dropping-and-recovering-organisms" class="section level1">
<h1 class="hasAnchor">
<a href="#dropping-and-recovering-organisms" class="anchor"></a>Dropping and Recovering Organisms</h1>
<p>One useful feature implemented in <code>pagoo</code> is the possibility of easily removing (hiding) organisms from the dataset. This is useful if during the analysis we identify some genome with weird charactersitics (i.e. potentially contaminated), or if we want to focus just in a subset of genomes of interest given any metadata value, or if we included an outgroup for phylogenetic purposes but we want to remove it from downstream analyses. Let’s see how it works..</p>
<div id="dropping-organisms" class="section level2">
<h2 class="hasAnchor">
<a href="#dropping-organisms" class="anchor"></a>Dropping organisms</h2>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">p<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1">p<span class="op">$</span>summary_stats</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total       600
## 2        Core       254
## 3       Shell       118
## 4       Cloud       228</code></pre>
<p>We can see that we have 5 organisms, and we have the pangenome summary statistics. Let’s say, for instance, that we want to remove organismC (row 3) because we only want to analyze organisms with <code>sero</code> <code>a</code> and <code>b</code>.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">p<span class="op">$</span><span class="kw"><a href="https://rdrr.io/r/base/drop.html">drop</a></span>(<span class="dv">3</span>) <span class="co"># We can use p$drop("organismC") as well</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2">p<span class="op">$</span>organisms <span class="co"># Updated organims !!</span></a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">p<span class="op">$</span>summary_stats <span class="co"># Updated stats !!</span></a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total       564
## 2        Core       265
## 3       Shell        76
## 4       Cloud       223</code></pre>
<p>As you see, not only the organismC has been removed from the dataset, but also summary statistics have been updated! This is also true for any other data field: <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, and<code>$organisms</code> fields are automatically updated. Also any embebbed statistical or visualization methods will now only consider available organisms/clusters/genes. When we drop an organims, we are hiding all features associated to that genomes including genes, sequences, gene clusters and metadata.</p>
<p><strong>It’s important to note that you don’t have to reassign the object to a new one, it is self modified (in place modification). That’s R6 reference semantics (see R6 documentation for details), use with caution.</strong></p>
</div>
<div id="recovering-dropped-organisms" class="section level2">
<h2 class="hasAnchor">
<a href="#recovering-dropped-organisms" class="anchor"></a>Recovering dropped organisms</h2>
<p>You can recover dropped organisms. To see if we have any hidden organism, we use the <code>$dropped</code> field, and then recover it using its index or its name.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">p<span class="op">$</span>dropped</a></code></pre></div>
<pre><code>##           3 
## "organismC"</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1">p<span class="op">$</span><span class="kw"><a href="https://rdrr.io/r/utils/recover.html">recover</a></span>(<span class="dv">3</span>) <span class="co"># We could also use p$recover("organismC")</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2">p<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
<p>Now organismC and all its features will be available again.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#predefined-subsets">Predefined Subsets</a></li>
      <li>
<a href="#the-operator">The <code>[</code> operator</a><ul class="nav nav-pills nav-stacked">
<li><a href="#vectorlist-notation">Vector/List notation</a></li>
      <li><a href="#matrix-notation">Matrix notation</a></li>
      </ul>
</li>
      <li>
<a href="#dropping-and-recovering-organisms">Dropping and Recovering Organisms</a><ul class="nav nav-pills nav-stacked">
<li><a href="#dropping-organisms">Dropping organisms</a></li>
      <li><a href="#recovering-dropped-organisms">Recovering dropped organisms</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
