<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pagoo: An encapsulated OO class system for analyzing bacterial pangenomes • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/darkly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pagoo: An encapsulated OO class system for analyzing bacterial pangenomes">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Input.html">Input</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Pagoo: An encapsulated OO class system for analyzing bacterial pangenomes</a>
    </li>
    <li>
      <a href="../articles/Quick_Start.html">Quick Start</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Pagoo: An encapsulated OO class system for analyzing bacterial pangenomes</h1>
                        <h4 class="author">Ignacio Ferrés</h4>
            
            <h4 class="date">April 26, 2019</h4>
      
      
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><code>pagoo</code> is an encapsulated, object-oriented class system for analyzing bacterial pangenomes. It uses the <a href="https://r6.r-lib.org/">R6</a> package as backend. It was designed in order to facilitate and speed-up the comparative analysis of multiple bacterial genomes, standardizing and optimizing routine tasks performed everyday. There are a handful of things done everyday when working with bacterial pangenomes: subset, summarize, extract, visualize and store data. So, <code>pagoo</code> is intended to facilitate these tasks as much as possible. <code>pagoo</code> is composed by three <code>R6</code> classes, each one more complex than the other, and that the more basic are inherited by the more complex ones. The most basic one, <code>PgR6</code>, contains basic subset methods and data manipulation functions. The second one, <code>PgR6M</code> inherits all the methods and fields from the previous one, and incorporates statistical and visualization methods. The last one, <code>PgR6MS</code>, inherits from <code>PgR6M</code> all its capabilities, and adds methods to manipulate DNA sequences.</p>
<p>Pagoo works over the pangenome after it has been built with any pangenome reconstruction software. So, you can do it with the software of your preference. Despite we recommend <code>Pewit</code>, our own pangenome reconstruction software, <code>pagoo</code> can read-in the output of most popular and state-of-the-art pangenome softwares like <code>Roary</code>, <code>Panaroo</code>, <code>panX</code> and <code>PIRATE</code>. <code>pagoo</code> also runs a <code>Shinny</code> application that provides reactive interaction with the data and facillitates handling and visualization.</p>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<p>Currently, <code>pagoo</code> is only available at GitHub. The easiest way to install this package is by using <code>devtools</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">rq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(devtools)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span>rq) <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">devtools<span class="op">::</span><span class="kw"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">'iferres/pagoo'</span>)</a></code></pre></div>
<p>We have tested <code>pagoo</code> in UNIX operating systems. However it should, we cannot warrant it fully works on Windows. Also, in the future we plan to put the code in CRAN and/or Bioconductor repositries.</p>
</div>
<div id="help" class="section level3">
<h3 class="hasAnchor">
<a href="#help" class="anchor"></a>Help</h3>
<p>All three classes are documented. You can access R help pages as with any other function:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pagoo)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">?PgR6</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/help.html">help</a></span>(<span class="st">'PgR6MS'</span>)</a></code></pre></div>
<p>but as R6 classes documentation is still not standarized, we recomend you to use the <code><a href="../reference/pagoo.html">pagoo::pagoo()</a></code> function to read the documentation from, and to to use it also instead of the raw classes.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">?pagoo</a></code></pre></div>
</div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>To show the full capabilities of <code>pagoo</code> I will use the most complex class, <code>PgR6MS</code> so you can see all the fancy things you can do. The example uses toy data that comes with the package.</p>
<div id="input" class="section level3">
<h3 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h3>
<p>First of all, lets extract and decompress the example data attached to this package:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pagoo)</a></code></pre></div>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Warning: package 'S4Vectors' was built under R version 3.6.1</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Warning: package 'BiocGenerics' was built under R version 3.6.1</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Warning: package 'IRanges' was built under R version 3.6.1</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## Warning: multiple methods tables found for 'type'</code></pre>
<pre><code>## 
## Attaching package: 'Biostrings'</code></pre>
<pre><code>## The following object is masked from 'package:BiocGenerics':
## 
##     type</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     strsplit</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<pre><code>## Loading required package: shiny</code></pre>
<pre><code>## Warning in .recacheSubclasses(def@className, def, env): undefined subclass
## "UnstitchedIPos" of class "IRanges_OR_IPos"; definition not updated</code></pre>
<pre><code>## Warning in .recacheSubclasses(def@className, def, env): undefined subclass
## "StitchedIPos" of class "IRanges_OR_IPos"; definition not updated</code></pre>
<pre><code>## Warning: no function found corresponding to methods exports from 'GenomicRanges'
## for: 'FactorToClass'</code></pre>
<pre><code>## Registered S3 method overwritten by 'seriation':
##   method         from 
##   reorder.hclust gclus</code></pre>
<pre><code>## Registered S3 methods overwritten by 'vegan':
##   method         from      
##   reorder.hclust seriation 
##   rev.hclust     dendextend</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">tgz &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'toy_data.tar.gz'</span>, <span class="dt">package =</span> <span class="st">'pagoo'</span>)</a>
<a class="sourceLine" id="cb32-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/untar.html">untar</a></span>(<span class="dt">tarfile =</span> tgz, <span class="dt">exdir =</span> <span class="kw"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>())</a>
<a class="sourceLine" id="cb32-3" data-line-number="3">files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(), <span class="dt">full.names =</span> <span class="ot">TRUE</span>, <span class="dt">pattern =</span> <span class="st">'tsv$|fasta$'</span>)</a>
<a class="sourceLine" id="cb32-4" data-line-number="4">files</a></code></pre></div>
<pre><code>## [1] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/case_clusters_meta.tsv"
## [2] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/case_df.tsv"           
## [3] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/case_orgs_meta.tsv"    
## [4] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/organismA.fasta"       
## [5] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/organismB.fasta"       
## [6] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/organismC.fasta"       
## [7] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/organismD.fasta"       
## [8] "/var/folders/cq/v60_yqxs0qjg01c2n2f6yc2w0000gn/T//RtmpMihatM/organismE.fasta"</code></pre>
<p>There are 3 <code>.tsv</code> files, and 5 <code>.fasta</code> files. The <code>case_df.tsv</code> file is the main one, having the information about the membership of each gene from each organism, to each group of orthologous (clusters). <code>case_*_meta.tsv</code> files have metadata for the clusters and for the organisms. <code>organism*.fasta</code> files have the sequences for each organism:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">data_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"case_df.tsv"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb34-2" data-line-number="2">df &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(data_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">quote =</span> <span class="st">''</span>)</a>
<a class="sourceLine" id="cb34-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(df)</a></code></pre></div>
<pre><code>##      gene       org cluster                             annot
## 1 gene081 organismA   OG001  Thioesterase superfamily protein
## 2 gene122 organismB   OG001          Thioesterase superfamily
## 3 gene299 organismC   OG001  Thioesterase superfamily protein
## 4 gene186 organismD   OG001  Thioesterase superfamily protein
## 5 gene076 organismE   OG001          Thioesterase superfamily
## 6 gene352 organismA   OG002 Inherit from proNOG: Thioesterase</code></pre>
<p>As you can see, <code>df</code> is a <code>data.frame</code> with 4 columns: the first one with the name of each gene, the second one with the organism to which each gene belongs, the third one with the cluster to which each gene was assigned in the pangenome reconstruction, and the last one with annotation metadata for each gene. Of the 4 columns, the former 3 are required, and <code>pagoo</code> will look for columns named “gene”, “org”, and “group”. More columns are optional, and you can add as many as you want (or none) to add metadata of each gene.</p>
<p>The next 2 <code>.tsv</code> files contains metadata for each cluster and for each organism, respectively, and are optional arguments.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="co">#</span></a>
<a class="sourceLine" id="cb36-2" data-line-number="2">clust_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"case_clusters_meta.tsv"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb36-3" data-line-number="3">group_meta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(clust_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">quote =</span> <span class="st">''</span>, <span class="dt">na.strings =</span> <span class="st">'&lt;NA&gt;'</span>)</a>
<a class="sourceLine" id="cb36-4" data-line-number="4">orgs_file &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"case_orgs_meta.tsv"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb36-5" data-line-number="5">org_meta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(orgs_file, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">quote =</span> <span class="st">''</span>, <span class="dt">na.strings =</span> <span class="st">'&lt;NA&gt;'</span>)</a>
<a class="sourceLine" id="cb36-6" data-line-number="6"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(group_meta)</a></code></pre></div>
<pre><code>##   cluster   kegg cog
## 1   OG001     NA   S
## 2   OG002     NA   S
## 3   OG003     NA  NA
## 4   OG004     NA   D
## 5   OG005 K01990   V
## 6   OG006     NA   V</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(org_meta)</a></code></pre></div>
<pre><code>##         org sero  country
## 1 organismA    a Westeros
## 2 organismB    b Westeros
## 3 organismC    c Westeros
## 4 organismD    a    Essos
## 5 organismE    b    Essos</code></pre>
<p>In both cases, the columns identifying each cluster and each organism must coincide with the names provided in the first data frame, so <code>pagoo</code> can map the metadata correctly.</p>
<p>As we will be working with the class <code>PgR6MS</code> in this example, DNA sequences are also required. The other 2 classes are lightweight, and don’t use all this data, so keep it in mind you can work without sequences if you have limited computational resources.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Biostrings))</a>
<a class="sourceLine" id="cb40-2" data-line-number="2">fasta_files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"[.]fasta"</span>, files, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb40-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(fasta_files) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">'[.]fasta'</span>, <span class="st">''</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(fasta_files))</a>
<a class="sourceLine" id="cb40-4" data-line-number="4"><span class="co"># A list of DNAStringSet, with organism names as list names, and DNAString names as gene names:</span></a>
<a class="sourceLine" id="cb40-5" data-line-number="5">seqs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(fasta_files, readDNAStringSet)</a></code></pre></div>
<p>Keep in mind that names should match with those provided in the data frames before.</p>
</div>
<div id="load-class" class="section level3">
<h3 class="hasAnchor">
<a href="#load-class" class="anchor"></a>Load class</h3>
<p>Once we have all the input data, lets load the example into a <code>PgR6MS</code> class:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1">pg &lt;-<span class="st"> </span>PgR6MS<span class="op">$</span><span class="kw">new</span>(<span class="dt">data =</span> df, <span class="dt">org_meta =</span> org_meta, <span class="dt">cluster_meta =</span> group_meta, <span class="dt">sequences =</span> seqs)</a></code></pre></div>
<pre><code>## Checking class.</code></pre>
<pre><code>## Checking dimnames.</code></pre>
<pre><code>## Creating gid (gene ids).</code></pre>
<pre><code>## Checking provided organism metadata.</code></pre>
<pre><code>## Checking provided cluster metadata.</code></pre>
<pre><code>## Creating panmatrix.</code></pre>
<pre><code>## Populating class.</code></pre>
<pre><code>## Checking input sequences.</code></pre>
<pre><code>## Checking that sequence names matches with DataFrame.</code></pre>
<pre><code>## Adding metadata to sequences.</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1">pg</a></code></pre></div>
<pre><code>## &lt;PgR6MS&gt;
##   Inherits from: &lt;PgR6M&gt;
##   Public:
##     add_metadata: function (map = "org", data) 
##     binomix_estimate: function (K.range = 3:5, core.detect.prob = 1, verbose = TRUE) 
##     cg_exp_decay_fit: function (raref, pcounts = 10, ...) 
##     clone: function (deep = FALSE) 
##     cloud_clusters: active binding
##     cloud_genes: active binding
##     cloud_sequences: active binding
##     clusters: active binding
##     core_clusters: active binding
##     core_genes: active binding
##     core_level: active binding
##     core_seqs_4_phylo: function (max_per_org = 1, fill = TRUE) 
##     core_sequences: active binding
##     dist: function (method = "bray", binary = FALSE, diag = FALSE, upper = FALSE, 
##     drop: function (x) 
##     dropped: active binding
##     fluidity: function (n.sim = 10) 
##     genes: active binding
##     gg_barplot: function () 
##     gg_binmap: function () 
##     gg_curves: function (what = c("pangenome", "coregenome"), ...) 
##     gg_dist: function (method = "bray", ...) 
##     gg_pca: function (colour = NULL, ...) 
##     gg_pie: function () 
##     initialize: function (data, org_meta, cluster_meta, core_level = 95, sep = "__", 
##     organisms: active binding
##     pan_matrix: active binding
##     pan_pca: function (center = TRUE, scale. = FALSE, ...) 
##     pg_power_law_fit: function (raref, ...) 
##     random_seed: active binding
##     rarefact: function (what = "pangenome", n.perm = 10) 
##     recover: function (x) 
##     runShinyApp: function () 
##     save_pangenomeRDS: function (file = "pangenome.rds", seqs.if.avail = TRUE) 
##     sequences: active binding
##     shell_clusters: active binding
##     shell_genes: active binding
##     shell_sequences: active binding
##     summary_stats: active binding
##     write_pangenome: function (dir = "pangenome", force = FALSE) 
##   Private:
##     .clusters: DataFrame
##     .data: DataFrame
##     .dropped: NULL
##     .level: 95
##     .organisms: DataFrame
##     .panmatrix: 1 1 1 1 1 1 1 1 1 0 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1  ...
##     .sep: __
##     .sequences: DNAStringSet
##     version: package_version, numeric_version</code></pre>
<p>Wow. I know. A better print method will be written in the future, this is the default print method for <code>R6</code> classes. But you can see that basically there are public and private fields/methods. In private, all raw data is stored. You will not have easy access to it. Instead, you will be able to access public functions and active bindings. Active bindings are functions that behave as it were variables, you will understand it better with some examples below.</p>
</div>
<div id="basic-fields-and-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#basic-fields-and-functions" class="anchor"></a>0. Basic fields and functions</h3>
<p>What do you want from a pangenome? Well, let see some examples:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="co"># Get a panmatrix</span></a>
<a class="sourceLine" id="cb54-2" data-line-number="2">pg<span class="op">$</span>pan_matrix[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##           OG001 OG002 OG003 OG004 OG005
## organismA     1     1     1     1     1
## organismB     1     1     0     1     1
## organismC     1     1     1     1     1
## organismD     1     1     1     1     1
## organismE     1     0     0     1     1</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="co"># Get organism information</span></a>
<a class="sourceLine" id="cb56-2" data-line-number="2">pg<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="co"># Get cluster information</span></a>
<a class="sourceLine" id="cb58-2" data-line-number="2">pg<span class="op">$</span>clusters</a></code></pre></div>
<pre><code>## DataFrame with 600 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG001       NA        S
## 2         OG002       NA        S
## 3         OG003       NA       NA
## 4         OG004       NA        D
## 5         OG005   K01990        V
## ...         ...      ...      ...
## 596       OG596       NA        L
## 597       OG597       NA        H
## 598       OG598       NA        S
## 599       OG599       NA       NA
## 600       OG600       NA       NA</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb60-1" data-line-number="1"><span class="co"># Get gene information</span></a>
<a class="sourceLine" id="cb60-2" data-line-number="2">pg<span class="op">$</span>genes</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 600
## $OG001
## DataFrame with 5 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG001 organismA  gene081 organismA__gene081
## 2    OG001 organismB  gene122 organismB__gene122
## 3    OG001 organismC  gene299 organismC__gene299
## 4    OG001 organismD  gene186 organismD__gene186
## 5    OG001 organismE  gene076 organismE__gene076
##                              annot
##                           &lt;factor&gt;
## 1 Thioesterase superfamily protein
## 2         Thioesterase superfamily
## 3 Thioesterase superfamily protein
## 4 Thioesterase superfamily protein
## 5         Thioesterase superfamily
## 
## $OG002
## DataFrame with 4 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG002 organismA  gene352 organismA__gene352
## 2    OG002 organismB  gene145 organismB__gene145
## 3    OG002 organismC  gene251 organismC__gene251
## 4    OG002 organismD  gene299 organismD__gene299
##                               annot
##                            &lt;factor&gt;
## 1 Inherit from proNOG: Thioesterase
## 2 Inherit from proNOG: Thioesterase
## 3 Inherit from proNOG: Thioesterase
## 4 Inherit from proNOG: Thioesterase
## 
## $OG003
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG003 organismA  gene359 organismA__gene359       NA
## 2    OG003 organismC  gene237 organismC__gene237       NA
## 3    OG003 organismD  gene069 organismD__gene069       NA
## 
## ...
## &lt;597 more elements&gt;</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" data-line-number="1"><span class="co"># Get sequences</span></a>
<a class="sourceLine" id="cb62-2" data-line-number="2">pg<span class="op">$</span>sequences</a></code></pre></div>
<pre><code>## DNAStringSetList of length 600
## [["OG001"]] organismA__gene081=ATGGAATCGGAAACAATCTTTCGGGAAATCAAAGCAAGCCAAAACG...
## [["OG002"]] organismA__gene352=ATGACAGATATACAGATTGAATTTCCGGAGAACTATCATTTTTCAA...
## [["OG003"]] organismA__gene359=TTGTTTTTTTATTTCACATTATGGATCACAAAAAGGATCGGTTTTC...
## [["OG004"]] organismA__gene258=ATGAAACAGATCCTCTGTATTGCAAACCAGAAAGGCGGAGTCGGTA...
## [["OG005"]] organismA__gene118=ATGATCAAAGTTAAAAACCTATCTAAATTCTATGGCAAAAAACTCG...
## [["OG006"]] organismA__gene348=GTGAGTCTGATCAAAGTAAGAAATTTGGTAAAGAATTATCATATTC...
## [["OG007"]] organismA__gene140=ATGAAGTTTTTTTTCAGACTTCTATCCTATTCCGTACATTACAAAT...
## [["OG008"]] organismB__gene075=ATGATTCAAAAAGCGAATTATAAACAGATTTTATTAGAATTTTTCT...
## [["OG009"]] organismB__gene328=ATGAATGATTATCTCACGTACGTATCTGACAATCGTTCCGTTCGAA...
## [["OG010"]] organismA__gene013=ATGAGCGAAATTGAAAGCGGTTTTTTTCAATCCGGCGGGTACAATC...
## ...
## &lt;590 more elements&gt;</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" data-line-number="1"><span class="co"># Access each cluster</span></a>
<a class="sourceLine" id="cb64-2" data-line-number="2">pg<span class="op">$</span>sequences[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 5
##     width seq                                               names               
## [1]   576 ATGGAATCGGAAACAATCTTTCG...AACTTCAAAAGACGGATTCCTGA organismA__gene081
## [2]   576 ATGGAACCAGAGGCAATCTACCG...AACTTCAAAACATAGAACCCTAA organismB__gene122
## [3]   576 ATGGATACGGAAACGATCTACCG...AACTTCAAAAGGTAAATTCTTAA organismC__gene299
## [4]   576 ATGGAAACAGAAACAATCTACCG...AACTTCAAAAGGTAAATTCCTAG organismD__gene186
## [5]   576 ATGGAATCAGAGGCAATCTACCG...AACTTCAAAAGATAATTTCCTAG organismE__gene076</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" data-line-number="1"><span class="co"># What is the default core level?</span></a>
<a class="sourceLine" id="cb66-2" data-line-number="2">pg<span class="op">$</span>core_level <span class="co">#You can change it by assiging another value (&lt;=100)</span></a></code></pre></div>
<pre><code>## [1] 95</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb68-1" data-line-number="1"><span class="co"># Get core clusters</span></a>
<a class="sourceLine" id="cb68-2" data-line-number="2">pg<span class="op">$</span>core_clusters</a></code></pre></div>
<pre><code>## DataFrame with 254 rows and 3 columns
##         cluster     kegg      cog
##     &lt;character&gt; &lt;factor&gt; &lt;factor&gt;
## 1         OG001       NA        S
## 2         OG002       NA        S
## 3         OG004       NA        D
## 4         OG005   K01990        V
## 5         OG006       NA        V
## ...         ...      ...      ...
## 250       OG437       NA       NA
## 251       OG439       NA        S
## 252       OG444       NA       NA
## 253       OG447       NA       NA
## 254       OG454       NA        L</code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" data-line-number="1"><span class="co"># Get shell genes</span></a>
<a class="sourceLine" id="cb70-2" data-line-number="2">pg<span class="op">$</span>shell_genes</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 118
## $OG003
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG003 organismA  gene359 organismA__gene359       NA
## 2    OG003 organismC  gene237 organismC__gene237       NA
## 3    OG003 organismD  gene069 organismD__gene069       NA
## 
## $OG009
## DataFrame with 2 rows and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG009 organismB  gene328 organismB__gene328       NA
## 2    OG009 organismE  gene181 organismE__gene181       NA
## 
## $OG012
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid                annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;             &lt;factor&gt;
## 1    OG012 organismA  gene138 organismA__gene138 Alpha beta hydrolase
## 2    OG012 organismB  gene025 organismB__gene025 Alpha beta hydrolase
## 3    OG012 organismE  gene031 organismE__gene031 Alpha beta hydrolase
## 
## ...
## &lt;115 more elements&gt;</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" data-line-number="1"><span class="co"># Get cloud sequences</span></a>
<a class="sourceLine" id="cb72-2" data-line-number="2">pg<span class="op">$</span>cloud_sequences</a></code></pre></div>
<pre><code>## DNAStringSetList of length 228
## [["OG008"]] organismB__gene075=ATGATTCAAAAAGCGAATTATAAACAGATTTTATTAGAATTTTTCT...
## [["OG024"]] organismD__gene092=TTGAAAAATTATTTGAAAAAAGAACCGAGCCATTTTGTTCCTAAAA...
## [["OG028"]] organismC__gene108=ATGAAGAAGCTCAACTTTCCATTTCATTATTTTTCTAAAACGATCT...
## [["OG030"]] organismE__gene222=ATGAAAATATCCGTTAAAACTCTTTTTTCTTTCTCCATTTTTATAC...
## [["OG033"]] organismE__gene115=TTGTTTAAAACGACTCATCCAAACGGATTGAAGAGAACTTTAACGG...
## [["OG045"]] organismE__gene241=ATGAAGAAAATTGTTAGTGTAAAACCGGAATTTGAATCTTTATCCG...
## [["OG047"]] organismE__gene001=ATGAAAGAAAAATTACCAGGTCTATTTGATTTAGTATTCCATTTTG...
## [["OG054"]] organismA__gene331=ATGTCTATTTTGATCACTGGATCTTCGGGTTTGATCGGTTCGGCTT...
## [["OG062"]] organismC__gene348=ATGGCACTTATTAATCCTCATATTAACTTCAACGGAAATGCCGAAG...
## [["OG063"]] organismA__gene257=ATGAAAATAGGACTCACGAGCGTAATCGTTCAAGATCCGATTGCGG...
## ...
## &lt;218 more elements&gt;</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" data-line-number="1"><span class="co"># Get summary statitics</span></a>
<a class="sourceLine" id="cb74-2" data-line-number="2">pg<span class="op">$</span>summary_stats</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total       600
## 2        Core       254
## 3       Shell       118
## 4       Cloud       228</code></pre>
</div>
<div id="subset" class="section level3">
<h3 class="hasAnchor">
<a href="#subset" class="anchor"></a>1. Subset</h3>
<p>Of course, you can subset each of the above fields as you would subset any other vector or list in R, but I wanted something better. Lets see a little example:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" data-line-number="1">pg<span class="op">$</span>pan_matrix[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>##           OG001 OG002 OG003
## organismA     1     1     1
## organismB     1     1     0</code></pre>
<p>I want to extract those precise genes. Wouldn’t be nice to have the possibility of using the same subset notation in lists as it were a matrix? Well, you can:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" data-line-number="1">pg[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>genes</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 3
## $OG001
## DataFrame with 2 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG001 organismA  gene081 organismA__gene081
## 2    OG001 organismB  gene122 organismB__gene122
##                              annot
##                           &lt;factor&gt;
## 1 Thioesterase superfamily protein
## 2         Thioesterase superfamily
## 
## $OG002
## DataFrame with 2 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG002 organismA  gene352 organismA__gene352
## 2    OG002 organismB  gene145 organismB__gene145
##                               annot
##                            &lt;factor&gt;
## 1 Inherit from proNOG: Thioesterase
## 2 Inherit from proNOG: Thioesterase
## 
## $OG003
## DataFrame with 1 row and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG003 organismA  gene359 organismA__gene359       NA</code></pre>
<p>I’m referencing the same genes in both the panmatrix and in the list above, using the same notation. You can use it with sequences also:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" data-line-number="1">pg[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">$</span>sequences</a></code></pre></div>
<pre><code>## DNAStringSetList of length 3
## [["OG001"]] organismA__gene081=ATGGAATCGGAAACAATCTTTCGGGAAATCAAAGCAAGCCAAAACG...
## [["OG002"]] organismA__gene352=ATGACAGATATACAGATTGAATTTCCGGAGAACTATCATTTTTCAA...
## [["OG003"]] organismA__gene359=TTGTTTTTTTATTTCACATTATGGATCACAAAAAGGATCGGTTTTC...</code></pre>
<p>The default print method for <code>DNAStringSetList</code> (<a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> package) is not showing what’s inside each list element, but you will check it has the precise sequences we asked for if you look inside each element.</p>
<p>Another tedious task I found working with microbial genomics is when you run a pangenome tool over a data set, and when you canalize it you realize that one or some of the input genomes have very bad quality, or you want to remove them from downstream analysis for any other cause. I implemented <code><a href="https://rdrr.io/r/base/drop.html">drop()</a></code> and <code><a href="https://rdrr.io/r/utils/recover.html">recover()</a></code> functions to remove or reload particular organisms from the data set:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb82-1" data-line-number="1"><span class="co"># Original dataset</span></a>
<a class="sourceLine" id="cb82-2" data-line-number="2">pg<span class="op">$</span>pan_matrix[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##           OG001 OG002 OG003 OG004 OG005
## organismA     1     1     1     1     1
## organismB     1     1     0     1     1
## organismC     1     1     1     1     1
## organismD     1     1     1     1     1
## organismE     1     0     0     1     1</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb84-1" data-line-number="1"><span class="co"># I want to drop organismD because blah blah</span></a>
<a class="sourceLine" id="cb84-2" data-line-number="2">pg<span class="op">$</span><span class="kw"><a href="https://rdrr.io/r/base/drop.html">drop</a></span>(<span class="st">'organismD'</span>)</a>
<a class="sourceLine" id="cb84-3" data-line-number="3"><span class="co"># Now is was removed from the dataset</span></a>
<a class="sourceLine" id="cb84-4" data-line-number="4">pg<span class="op">$</span>pan_matrix[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##           OG001 OG002 OG003 OG004 OG005
## organismA     1     1     1     1     1
## organismB     1     1     0     1     1
## organismC     1     1     1     1     1
## organismE     1     0     0     1     1</code></pre>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" data-line-number="1">pg<span class="op">$</span>genes[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## SplitDataFrameList of length 3
## $OG001
## DataFrame with 4 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG001 organismA  gene081 organismA__gene081
## 2    OG001 organismB  gene122 organismB__gene122
## 3    OG001 organismC  gene299 organismC__gene299
## 4    OG001 organismE  gene076 organismE__gene076
##                              annot
##                           &lt;factor&gt;
## 1 Thioesterase superfamily protein
## 2         Thioesterase superfamily
## 3 Thioesterase superfamily protein
## 4         Thioesterase superfamily
## 
## $OG002
## DataFrame with 3 rows and 5 columns
##    cluster       org     gene                gid
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt;
## 1    OG002 organismA  gene352 organismA__gene352
## 2    OG002 organismB  gene145 organismB__gene145
## 3    OG002 organismC  gene251 organismC__gene251
##                               annot
##                            &lt;factor&gt;
## 1 Inherit from proNOG: Thioesterase
## 2 Inherit from proNOG: Thioesterase
## 3 Inherit from proNOG: Thioesterase
## 
## $OG003
## DataFrame with 2 rows and 5 columns
##    cluster       org     gene                gid    annot
##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;        &lt;character&gt; &lt;factor&gt;
## 1    OG003 organismA  gene359 organismA__gene359       NA
## 2    OG003 organismC  gene237 organismC__gene237       NA</code></pre>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" data-line-number="1">pg<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 5 organismE        b    Essos</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" data-line-number="1"><span class="co"># And summary statistics are also updated (compare with pg$summary_stats called before)</span></a>
<a class="sourceLine" id="cb90-2" data-line-number="2">pg<span class="op">$</span>summary_stats</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total       563
## 2        Core       261
## 3       Shell        74
## 4       Cloud       228</code></pre>
<p>Now you want to get it back:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" data-line-number="1"><span class="co"># Any dropped organism?</span></a>
<a class="sourceLine" id="cb92-2" data-line-number="2">pg<span class="op">$</span>dropped</a></code></pre></div>
<pre><code>##           4 
## "organismD"</code></pre>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="co"># You can drop/recover them by name or by index</span></a>
<a class="sourceLine" id="cb94-2" data-line-number="2">pg<span class="op">$</span><span class="kw"><a href="https://rdrr.io/r/utils/recover.html">recover</a></span>(<span class="dv">4</span>)</a>
<a class="sourceLine" id="cb94-3" data-line-number="3">pg<span class="op">$</span>organisms</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##         org     sero  country
##    &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;
## 1 organismA        a Westeros
## 2 organismB        b Westeros
## 3 organismC        c Westeros
## 4 organismD        a    Essos
## 5 organismE        b    Essos</code></pre>
<p>Note that you don’t have to reassign the result of the function to a new variable, that’s <a href="https://r6.r-lib.org/articles/Introduction.html#r6-classes">R6</a> reference semantics magic. Use with <a href="https://r6.r-lib.org/articles/Introduction.html#cloning-objects">caution</a>.</p>
<p>An everyday application of using <code><a href="https://rdrr.io/r/base/drop.html">drop()</a></code> and <code><a href="https://rdrr.io/r/utils/recover.html">recover()</a></code> is to reconstruct a pangenome including an outgroup organism for phylogenetic purposes, but discarding it for specific clade analysis which don’t require it.</p>
</div>
<div id="applying-statistical-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#applying-statistical-methods" class="anchor"></a>2. Applying statistical methods</h3>
<p>Once you have a pangenome, you may want to get some statistics about it. I implemented some methods in <code>pagoo</code>, and take some others from widely used packages as <a href="https://github.com/vegandevs/vegan">vegan</a> and <a href="https://github.com/larssnip/micropan">micropan</a>. You can contribute by adding more.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" data-line-number="1"><span class="co"># Default: vegan::vegdist(method='bray')</span></a>
<a class="sourceLine" id="cb96-2" data-line-number="2">pg<span class="op">$</span><span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>()</a></code></pre></div>
<pre><code>##           organismA organismB organismC organismD
## organismB 0.3015428                              
## organismC 0.2537313 0.2927536                    
## organismD 0.2755906 0.3034965 0.2151556          
## organismE 0.2762431 0.2821270 0.2582026 0.2617080</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" data-line-number="1"><span class="co"># micropan::fluidity</span></a>
<a class="sourceLine" id="cb98-2" data-line-number="2">pg<span class="op">$</span><span class="kw">fluidity</span>()</a></code></pre></div>
<pre><code>##       Mean        Std 
## 0.25044734 0.02812111</code></pre>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb100-1" data-line-number="1"><span class="co"># Fitting a power law to pangenome rarefaction curves</span></a>
<a class="sourceLine" id="cb100-2" data-line-number="2">pg<span class="op">$</span><span class="kw">pg_power_law_fit</span>(<span class="dt">n.perm=</span><span class="dv">100</span>)</a></code></pre></div>
<pre><code>## $formula
## function (x) 
## K * x^delta
## &lt;environment: 0x7fb723ca4c58&gt;
## 
## $params
##          K      delta 
## 343.231117   0.344505 
## 
## attr(,"alpha")
## [1] 0.655495</code></pre>
<p>…and some others. It is worth noting that each statistic will be computed for available organisms, not taking into consideration previously dropped ones.</p>
</div>
<div id="visualizing-results" class="section level3">
<h3 class="hasAnchor">
<a href="#visualizing-results" class="anchor"></a>3. Visualizing results</h3>
<p>I used <a href="https://ggplot2.tidyverse.org/">ggplot2</a> as backend for visualization methods. Here I’m showing some of the available.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb102-1" data-line-number="1"><span class="co"># Summary pie</span></a>
<a class="sourceLine" id="cb102-2" data-line-number="2">pg<span class="op">$</span><span class="kw">gg_pie</span>()</a></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb103-1" data-line-number="1"><span class="co"># Classic barplot</span></a>
<a class="sourceLine" id="cb103-2" data-line-number="2">pg<span class="op">$</span><span class="kw">gg_barplot</span>()</a></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-16-2.png" width="700"></p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb104-1" data-line-number="1"><span class="co"># Binmap</span></a>
<a class="sourceLine" id="cb104-2" data-line-number="2">pg<span class="op">$</span><span class="kw">gg_binmap</span>()</a></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-16-3.png" width="700"></p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb105-1" data-line-number="1"><span class="co"># Pangenome and coregenome curves</span></a>
<a class="sourceLine" id="cb105-2" data-line-number="2">pg<span class="op">$</span><span class="kw">gg_curves</span>()</a></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-16-4.png" width="700"></p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb106-1" data-line-number="1"><span class="co"># Customize with ggplot2</span></a>
<a class="sourceLine" id="cb106-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb106-3" data-line-number="3">pg<span class="op">$</span><span class="kw">gg_curves</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb106-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb106-5" data-line-number="5"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>category, <span class="dt">scales =</span> <span class="st">'free_y'</span>)</a></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-16-5.png" width="700"></p>
</div>
<div id="recipes" class="section level3">
<h3 class="hasAnchor">
<a href="#recipes" class="anchor"></a>Recipes</h3>
<p>I call recipes to short and easy scripts that solves a specific problem. Here I show a few to demonstrate the potentialities of <code>pagoo</code>, and to show how to use it with popular genomics R packages. These examples uses sequence data because I think it’s the most interesting case to work with, so assumes that you are working with a <code>PgR6MS</code> class, like the one in the tutorial.</p>
<p>You can help by expanding the following with more examples:</p>
<div id="plot-coregenome-alignment-nj-tree" class="section level4">
<h4 class="hasAnchor">
<a href="#plot-coregenome-alignment-nj-tree" class="anchor"></a>Plot Coregenome alignment NJ tree</h4>
<p>One standard analysis when dealing with a pangenome is to align each core cluster, concatenate them into a super alignment, and infer a phylogeny. Since core clusters might contain in-paralogs, you need to select no more than one of them on each cluster from each organism. Also, if <code>core_level</code> (that is, the percentage of organisms a gene has to be in, to be considered core) is set below 100%, some clusters might have few organisms missing, so you would have to fill with gaps and also not miss the sequence order before concatenating. For this purpose I wrote the <code>core_seqs_4_phylo()</code> function, which return something similar to <code>core_sequences</code> field, but with above considerations. In the following example I use <a href="https://github.com/KlausVigo/phangorn">phangorn</a> package to compute and visualize a phylogenetic tree.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb107-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Biostrings) <span class="co"># DNAString xscat</span></a>
<a class="sourceLine" id="cb107-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)   <span class="co"># %&gt;% (pipe operator)</span></a>
<a class="sourceLine" id="cb107-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(DECIPHER)   <span class="co"># AlignSeqs() </span></a>
<a class="sourceLine" id="cb107-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(phangorn)   <span class="co"># phyDat(), dist.ml(), NJ()</span></a>
<a class="sourceLine" id="cb107-5" data-line-number="5"></a>
<a class="sourceLine" id="cb107-6" data-line-number="6"><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(pg<span class="op">$</span><span class="kw">core_seqs_4_phylo</span>(), AlignSeqs) <span class="op">%&gt;%</span><span class="st">   </span><span class="co"># Align each core cluster</span></a>
<a class="sourceLine" id="cb107-7" data-line-number="7"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(xscat, .) <span class="op">%&gt;%</span><span class="st">                         </span><span class="co"># Concatenate alignments</span></a>
<a class="sourceLine" id="cb107-8" data-line-number="8"><span class="st">  </span><span class="kw">as</span>(<span class="st">'matrix'</span>) <span class="op">%&gt;%</span><span class="st">                              </span><span class="co"># Transform to matrix</span></a>
<a class="sourceLine" id="cb107-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/phangorn/man/phyDat.html">phyDat</a></span>(<span class="dt">type =</span> <span class="st">'DNA'</span>) <span class="op">%&gt;%</span><span class="st">                      </span><span class="co"># Transform to phangorn's phyDat</span></a>
<a class="sourceLine" id="cb107-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/phangorn/man/dist.hamming.html">dist.ml</a></span>() <span class="op">%&gt;%</span><span class="st">                                 </span><span class="co"># Compute distance</span></a>
<a class="sourceLine" id="cb107-11" data-line-number="11"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/phangorn/man/NJ.html">NJ</a></span>() <span class="op">%&gt;%</span><span class="st">                                      </span><span class="co"># Compute NJ</span></a>
<a class="sourceLine" id="cb107-12" data-line-number="12"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>()                                        <span class="co"># Plot it</span></a></code></pre></div>
</div>
<div id="blastp-cluster-1-against-custom-blast-db" class="section level4">
<h4 class="hasAnchor">
<a href="#blastp-cluster-1-against-custom-blast-db" class="anchor"></a>Blastp cluster 1 against custom blast db</h4>
<p>You can use the <a href="https://github.com/mhahsler/rBLAST">rBLAST</a> package to search a specific set of sequences on a custom blast database. In the example below, sequences are first translated, and then searched against a protein blast database. The result is returned as a <code>data.frame</code> for parsing and better curation.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb108-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">'mhahsler/rBLAST'</span>)</a>
<a class="sourceLine" id="cb108-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Biostrings)</a>
<a class="sourceLine" id="cb108-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(rBLAST)</a>
<a class="sourceLine" id="cb108-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</a>
<a class="sourceLine" id="cb108-5" data-line-number="5"></a>
<a class="sourceLine" id="cb108-6" data-line-number="6">db_path &lt;-<span class="st"> 'path/to/custom/blastpdb'</span></a>
<a class="sourceLine" id="cb108-7" data-line-number="7">db &lt;-<span class="st"> </span><span class="kw">blast</span>(<span class="dt">db =</span> db_path, <span class="dt">type =</span> <span class="st">'blastp'</span>)</a>
<a class="sourceLine" id="cb108-8" data-line-number="8"></a>
<a class="sourceLine" id="cb108-9" data-line-number="9">pg<span class="op">$</span>sequences[[<span class="dv">1</span>]] <span class="op">%&gt;%</span><span class="st">   </span><span class="co"># Get DNAStringSet of cluster 1</span></a>
<a class="sourceLine" id="cb108-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/translate.html">translate</a></span>() <span class="op">%&gt;%</span><span class="st">       </span><span class="co"># Translate DNAStringSet</span></a>
<a class="sourceLine" id="cb108-11" data-line-number="11"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(db, .)        <span class="co"># Run Blastp</span></a></code></pre></div>
</div>
<div id="compute-dnds-over-alignment-of-core-clusters" class="section level4">
<h4 class="hasAnchor">
<a href="#compute-dnds-over-alignment-of-core-clusters" class="anchor"></a>Compute dn/ds over alignment of core clusters</h4>
<p>A standard approach for detecting selection over core genes is to compute dn/ds: the number of substitutions per synonymous site divided by the number of non-synonymous substitutions between two protein-coding genes. <a href="http://ape-package.ird.fr/">ape</a> package provides a function for fast dnds computation, and <a href="http://www2.decipher.codes/">DECIPHER</a> provides the <code>AlignTranslation</code> function which aligns DNA sequences by aligning their translated amino acid sequences.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb109-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(DECIPHER)</a>
<a class="sourceLine" id="cb109-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ape)</a>
<a class="sourceLine" id="cb109-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</a>
<a class="sourceLine" id="cb109-4" data-line-number="4"></a>
<a class="sourceLine" id="cb109-5" data-line-number="5"><span class="co"># This function returns the mean dnds of all pairwise dnds computation</span></a>
<a class="sourceLine" id="cb109-6" data-line-number="6"><span class="co"># for a given DNAStringSet.</span></a>
<a class="sourceLine" id="cb109-7" data-line-number="7">compute_dnds &lt;-<span class="st"> </span><span class="cf">function</span>(dnastringset){</a>
<a class="sourceLine" id="cb109-8" data-line-number="8">  DECIPHER<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignTranslation.html">AlignTranslation</a></span>(dnastringset) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb109-9" data-line-number="9"><span class="st">    </span>ape<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html">as.DNAbin</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb109-10" data-line-number="10"><span class="st">    </span>ape<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/ape/man/dnds.html">dnds</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb109-11" data-line-number="11"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>()</a>
<a class="sourceLine" id="cb109-12" data-line-number="12">}</a>
<a class="sourceLine" id="cb109-13" data-line-number="13"></a>
<a class="sourceLine" id="cb109-14" data-line-number="14"><span class="co"># Set core_level to 100%</span></a>
<a class="sourceLine" id="cb109-15" data-line-number="15">pg<span class="op">$</span>core_level &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb109-16" data-line-number="16"><span class="co"># Compute dnds for all core sequences:</span></a>
<a class="sourceLine" id="cb109-17" data-line-number="17"><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(pg<span class="op">$</span>core_sequences, compute_dnds)</a></code></pre></div>
</div>
<div id="heatmap-associated-to-phylogenetic-tree" class="section level4">
<h4 class="hasAnchor">
<a href="#heatmap-associated-to-phylogenetic-tree" class="anchor"></a>Heatmap associated to phylogenetic tree</h4>
<p>Is often useful to map organism metadata (or even the <code>pan_matrix</code>, like <code>$gg_binmap()</code> function) next to a phylogenetic tree. The below example shows how to do that using <a href="https://github.com/GuangchuangYu/ggtree">ggtree</a>. First computes the tree, then formats the metadata, and last creates the plot. Assumes you have provided organisms metadata when building the class, just like in the tutorial example.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb110-1" data-line-number="1">tree &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(pg<span class="op">$</span><span class="kw">core_seqs_4_phylo</span>(), AlignSeqs) <span class="op">%&gt;%</span><span class="st">   </span><span class="co"># Align each core cluster</span></a>
<a class="sourceLine" id="cb110-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(xscat, .) <span class="op">%&gt;%</span><span class="st">                                 </span><span class="co"># Concatenate alignments</span></a>
<a class="sourceLine" id="cb110-3" data-line-number="3"><span class="st">  </span><span class="kw">as</span>(<span class="st">'matrix'</span>) <span class="op">%&gt;%</span><span class="st">                                      </span><span class="co"># Transform to matrix</span></a>
<a class="sourceLine" id="cb110-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/phangorn/man/phyDat.html">phyDat</a></span>(<span class="dt">type =</span> <span class="st">'DNA'</span>) <span class="op">%&gt;%</span><span class="st">                              </span><span class="co"># Transform to phangorn's phyDat</span></a>
<a class="sourceLine" id="cb110-5" data-line-number="5"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/phangorn/man/dist.hamming.html">dist.ml</a></span>() <span class="op">%&gt;%</span><span class="st">                                         </span><span class="co"># Compute distance</span></a>
<a class="sourceLine" id="cb110-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/phangorn/man/NJ.html">NJ</a></span>()                                                  <span class="co"># Compute NJ tree</span></a>
<a class="sourceLine" id="cb110-7" data-line-number="7"></a>
<a class="sourceLine" id="cb110-8" data-line-number="8"><span class="co"># Formats labels and converts 'phylo' object to 'gg' object</span></a>
<a class="sourceLine" id="cb110-9" data-line-number="9">tree<span class="op">$</span>tip.label &lt;-<span class="st"> </span>pg<span class="op">$</span>organisms<span class="op">$</span>org</a>
<a class="sourceLine" id="cb110-10" data-line-number="10">gtree &lt;-<span class="st"> </span><span class="kw">ggtree</span>(tree) <span class="op">+</span><span class="st"> </span><span class="kw">geom_tiplab</span>()</a>
<a class="sourceLine" id="cb110-11" data-line-number="11"></a>
<a class="sourceLine" id="cb110-12" data-line-number="12"><span class="co"># Formats organisms metadata</span></a>
<a class="sourceLine" id="cb110-13" data-line-number="13">dat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(pg<span class="op">$</span>organisms[<span class="op">-</span><span class="dv">1</span>])</a>
<a class="sourceLine" id="cb110-14" data-line-number="14">dat[] &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(dat, as.character)</a>
<a class="sourceLine" id="cb110-15" data-line-number="15"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(dat) &lt;-<span class="st"> </span>pg<span class="op">$</span>organisms[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb110-16" data-line-number="16"></a>
<a class="sourceLine" id="cb110-17" data-line-number="17"><span class="co"># Plot heatmap associated to tree</span></a>
<a class="sourceLine" id="cb110-18" data-line-number="18"><span class="kw">gheatmap</span>(gtree, <span class="dt">data =</span> dat)</a></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#example">Example</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
