<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3 - Querying Data • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3 - Querying Data">
<meta property="og:description" content="pagoo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-164668935-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-164668935-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.3.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1-Quick_Start.html">1 - Quick Start</a>
    </li>
    <li>
      <a href="../articles/2-Input.html">2 - Input</a>
    </li>
    <li>
      <a href="../articles/3-Querying_Data.html">3 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/4-Subseting.html">4 - Subseting</a>
    </li>
    <li>
      <a href="../articles/5-Methods_Plots.html">5 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/6-Recipes.html">6 - Recipes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iferres/pagoo/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>3 - Querying Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/iferres/pagoo/blob/master/vignettes/3-Querying_Data.Rmd"><code>vignettes/3-Querying_Data.Rmd</code></a></small>
      <div class="hidden name"><code>3-Querying_Data.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre, code {white-space:pre !important; overflow-x:auto;}
</style>
<p>In this section we will start exploring what is stored inside the <code>pagoo</code> object and how we can access this information. Keep in mind that this object has its own associated data and methods that can be easily queried with the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> operator. These methods allow for the rapid subsetting, extraction and visualization of pangenome data.</p>
<p>First of all, we will load a pangenome using a toy dataset included in the package. This is a preloaded set of 10 <em>Campylobacter spp.</em> genomes, with metadata associated.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">pagoo</span>) <span class="co"># Load package</span>
<span class="no">rds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'campylobacter.RDS'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'pagoo'</span>)
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/load_pangenomeRDS.html">load_pangenomeRDS</a></span>(<span class="no">rds</span>)</pre></body></html></div>
<div id="summary-statistics" class="section level1">
<h1 class="hasAnchor">
<a href="#summary-statistics" class="anchor"></a>Summary statistics</h1>
<p>A pangenome can be stratified in different gene subsets according to their frequency in the dataset. The <code>core genes</code> can be defined as those present in all or almost every genome (typically 95-100%). The remaining genes are defined as the accessory genome, that can be subdivided in <code>cloud genes</code> or singletons (present in one genome or in genomes that are identical) and <code>shell genes</code> which are those in the middle. Let’s see this using <code>pagoo</code>:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">p</span>$<span class="no">summary_stats</span></pre></body></html></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total      2599
## 2        Core      1623
## 3       Shell       441
## 4       Cloud       535</code></pre>
<div id="core-level" class="section level2">
<h2 class="hasAnchor">
<a href="#core-level" class="anchor"></a>Core level</h2>
<p>The <code>core level</code> defines the minimum number of genomes (as a percentage) in which a certain gene should be present to be considered a core gene. By default, <code>pagoo</code> considers as core all genes present in at least 95% of organisms. The core level can be modified to be more or less stringent defining the core genome. This feature exemplifies R6’s reference semantics, since modifying the core level will affect the pangenome object state resulting in different core, shell and cloud sets. Have a look:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">p</span>$<span class="no">core_level</span></pre></body></html></div>
<pre><code>## [1] 95</code></pre>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">p</span>$<span class="no">core_level</span> <span class="kw">&lt;-</span> <span class="fl">100</span> <span class="co"># Change value</span>
<span class="no">p</span>$<span class="no">summary_stats</span>     <span class="co"># Updated object</span></pre></body></html></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total      2599
## 2        Core      1553
## 3       Shell       511
## 4       Cloud       535</code></pre>
<p>As you can see, changing the core level from 95% to a more stringent 100% cause in decrease in the number of core genes from 1623 to 1553, and a concomitant increase in shell genes from 441 to 511. This means that 70 genes migrated from core to shell when increasing the threshold to consider a cluster as “core” to 100%. Now this changes remain in the object for subsequent analysis, or can be reverted by setting the core level again at the original value.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">p</span>$<span class="no">core_level</span> <span class="kw">&lt;-</span> <span class="fl">95</span></pre></body></html></div>
</div>
</div>
<div id="pangenome-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#pangenome-matrix" class="anchor"></a>Pangenome matrix</h1>
<p>The pangenome matrix is one of the most useful things when analyzing pangenomes. Typically, it represents organisms in rows and clusters of orthologous in columns informing about gene abundance (considering paralogues). The pangenome matrix looks like this (printing only first 5 columns):</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">p</span>$<span class="no">pan_matrix</span>[, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##            OG0001 OG0002 OG0003 OG0004 OG0005
## 16244_6_6       1      1      1      1      1
## 16244_6_18      1      1      1      1      1
## 16244_6_24      1      1      1      1      1
## 16244_6_31      1      1      1      1      1
## 17059_2_16      1      1      1      1      1
## 17059_2_23      1      1      1      1      0
## 17059_2_27      1      1      1      1      1
## 17150_1_73      1      1      1      1      1
## 17059_2_11      1      1      1      1      0
## 17059_2_42      1      1      1      1      0</code></pre>
</div>
<div id="gene-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-metadata" class="anchor"></a>Gene metadata</h1>
<p>Individual gene metadata can be accessed by using the <code>$genes</code> suffix. It always contains the gene name, the organism to which it belongs, the cluster to where it was assigned, and a gene identifier (<code>gid</code>) that is mainly used internally to organize the data. Also, it may typically include annotation data, genomic coordinates, etc, but this other metadata is optional. Gene metadata is splited by cluster, so it consist in a <code>List</code> of <code>DataFrame</code>s.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">p</span>$<span class="no">genes</span></pre></body></html></div>
<pre><code>## SplitDataFrameList of length 2599
## $OG0001
## DataFrame with 10 rows and 10 columns
##     cluster        org             gene                          gid
##    &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt;
## 1    OG0001  16244_6_6  16244_6_6_00150   16244_6_6__16244_6_6_00150
## 2    OG0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_18_00172
## 3    OG0001 16244_6_24 16244_6_24_01512 16244_6_24__16244_6_24_01512
## 4    OG0001 16244_6_31 16244_6_31_00173 16244_6_31__16244_6_31_00173
## 5    OG0001 17059_2_16 17059_2_16_01012 17059_2_16__17059_2_16_01012
## 6    OG0001 17059_2_23 17059_2_23_01040 17059_2_23__17059_2_23_01040
## 7    OG0001 17059_2_27 17059_2_27_00492 17059_2_27__17059_2_27_00492
## 8    OG0001 17150_1_73 17150_1_73_00221 17150_1_73__17150_1_73_00221
## 9    OG0001 17059_2_11 17059_2_11_00174 17059_2_11__17059_2_11_00174
## 10   OG0001 17059_2_42 17059_2_42_00176 17059_2_42__17059_2_42_00176
##       geneName
##    &lt;character&gt;
## 1         ilvC
## 2         ilvC
## 3         ilvC
## 4         ilvC
## 5         ilvC
## 6         ilvC
## 7         ilvC
## 8         ilvC
## 9         ilvC
## 10        ilvC
##                                                                                                                                                                   product
##                                                                                                                                                               &lt;character&gt;
## 1  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 2  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 3  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 4  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 5  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 6  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 7  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 8  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 9  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 10 ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
##                       contig      from        to      strand
##                  &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1  ERS672247|SC|contig000001    137333    138355           +
## 2  ERS672259|SC|contig000001    173541    174563           +
## 3  ERS672265|SC|contig000005    173654    174676           +
## 4  ERS672272|SC|contig000001    173541    174563           +
## 5  ERS739235|SC|contig000005     43645     44667           -
## 6  ERS739242|SC|contig000004    173578    174600           +
## 7  ERS739246|SC|contig000002    184035    185057           -
## 8  ERS686652|SC|contig000001    207206    208228           +
## 9  ERS739230|SC|contig000001    173577    174599           +
## 10 ERS739261|SC|contig000001    173569    174591           +
## 
## $OG0002
## DataFrame with 10 rows and 10 columns
##     cluster        org             gene                          gid
##    &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt;
## 1    OG0002  16244_6_6  16244_6_6_01290   16244_6_6__16244_6_6_01290
## 2    OG0002 16244_6_18 16244_6_18_01307 16244_6_18__16244_6_18_01307
## 3    OG0002 16244_6_24 16244_6_24_01054 16244_6_24__16244_6_24_01054
## 4    OG0002 16244_6_31 16244_6_31_01312 16244_6_31__16244_6_31_01312
## 5    OG0002 17059_2_16 17059_2_16_01627 17059_2_16__17059_2_16_01627
## 6    OG0002 17059_2_23 17059_2_23_00348 17059_2_23__17059_2_23_00348
## 7    OG0002 17059_2_27 17059_2_27_01208 17059_2_27__17059_2_27_01208
## 8    OG0002 17150_1_73 17150_1_73_01398 17150_1_73__17150_1_73_01398
## 9    OG0002 17059_2_11 17059_2_11_01048 17059_2_11__17059_2_11_01048
## 10   OG0002 17059_2_42 17059_2_42_00751 17059_2_42__17059_2_42_00751
##       geneName
##    &lt;character&gt;
## 1         hprA
## 2         hprA
## 3         hprA
## 4         hprA
## 5         hprA
## 6         hprA
## 7         hprA
## 8         hprA
## 9         hprA
## 10        hprA
##                                                                                                                                                                                                                                   product
##                                                                                                                                                                                                                               &lt;character&gt;
## 1  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 2  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 3  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 4  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 5  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 6  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 7  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 8  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 9  2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
## 10 2-hydroxyacid dehydrogenase,Putative 2-hydroxyacid dehydrogenase HI_1556,2-hydroxyacid dehydrogenase,Phosphoserine aminotransferase,phosphoglycerate dehydrogenase,D-isomer specific 2-hydroxyacid dehydrogenase%2C NAD binding domain
##                       contig      from        to      strand
##                  &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1  ERS672247|SC|contig000003    274585    275517           -
## 2  ERS672259|SC|contig000003    272748    273680           -
## 3  ERS672265|SC|contig000003    274698    275630           -
## 4  ERS672272|SC|contig000003    274593    275525           -
## 5  ERS739235|SC|contig000012     11617     12549           +
## 6  ERS739242|SC|contig000001    308107    309039           -
## 7  ERS739246|SC|contig000005    114555    115487           -
## 8  ERS686652|SC|contig000003    273866    274798           -
## 9  ERS739230|SC|contig000003     15410     16342           +
## 10 ERS739261|SC|contig000003     15409     16341           +
## 
## ...
## &lt;2597 more elements&gt;</code></pre>
<p>If you want to work with this data as a single <code>DataFrame</code>, just <code>unlist</code> it:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">p</span>$<span class="no">genes</span>, <span class="kw">use.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><code>pagoo</code> also includes predefined subsets fields to list only certain pangenome category, these are queried by adding a prefix with the desired category followed by an underscore: <code>$core_genes</code>, <code>$shell_genes</code>, and <code>$cloud_genes</code>. These kind of subsets are better explaied in the ‘4 - Subets’ tutorial, and also apply to other pangenome data described below.</p>
</div>
<div id="clusters-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#clusters-metadata" class="anchor"></a>Clusters metadata</h1>
<p>Groups of orthologues (clusters) are also stored in <code>pagoo</code> objects as a table with a cluster identifier per row, and optional metadata associated as additional columns.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">p</span>$<span class="no">clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 2599 rows and 2 columns
##          cluster                   Pfam_Arch
##      &lt;character&gt;                 &lt;character&gt;
## 1         OG0001                2-Hacid_dh_C
## 2         OG0002     2-Hacid_dh_C;2-Hacid_dh
## 3         OG0003 2-Hacid_dh_C;ACT;2-Hacid_dh
## 4         OG0004             2Fe-2S_thioredx
## 5         OG0005         4HB_MCP_1;MCPsignal
## ...          ...                         ...
## 2595      OG2595                   zf-RING_7
## 2596      OG2596                    zf-TFIIB
## 2597      OG2597                        ZinT
## 2598      OG2598                        ZnuA
## 2599      OG2599                    ZT_dimer</code></pre>
<p>Subsets also exists for this field: <code>$core_clusters</code>, <code>$shell_clusters</code>, and <code>$cloud_clusters</code>.</p>
</div>
<div id="sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#sequences" class="anchor"></a>Sequences</h1>
<p>Although is an optional field (it exists only if user provide this data as an argument when object is created), <code>$sequences</code> gives access to sequence data. Sequences are stored as a <code>List</code> of <code>DNAStringSet</code> (a.k.a <code>DNAStringSetList</code>, Biostrings package), grouped by cluster.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">p</span>$<span class="no">sequences</span>             <span class="co"># List all sequences grouped by cluster</span></pre></body></html></div>
<pre><code>## DNAStringSetList of length 2599
## [["OG0001"]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTGCGA...
## [["OG0002"]] 16244_6_18__16244_6_18_01307=ATGAAAATAGTATGCTTAGATGCCGACACGCTTGG...
## [["OG0003"]] 16244_6_18__16244_6_18_00052=ATGAAAACAGTTATAGTTTGCGATGCAATACATCC...
## [["OG0004"]] 16244_6_18__16244_6_18_01700=ATGAAATTCGAATTTACTCATGAGCAATTATCGGC...
## [["OG0005"]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAAAAC...
## [["OG0006"]] 16244_6_18__16244_6_18_01083=ATGAATTATTTTGAGAATTTAAAAGTTTCAACAAA...
## [["OG0007"]] 16244_6_18__16244_6_18_01638=ATGCGAATTAGAATTTATTATGAAGATACCGATGC...
## [["OG0008"]] 16244_6_18__16244_6_18_00021=ATGATGAAAGATATGGGCGAGCCACGTATAAAAAT...
## [["OG0009"]] 16244_6_18__16244_6_18_01216=ATGGGGCTTACTACGAGTACGACAAAGTATATAAC...
## [["OG0010"]] 16244_6_18__16244_6_18_00770=ATGAAAAGAGTGGTTATAAAAGTAGGCTCTCACGT...
## ...
## &lt;2589 more elements&gt;</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">p</span>$<span class="no">sequences</span><span class="kw">[[</span><span class="st">"OG0001"</span>]]  <span class="co"># List first cluster</span></pre></body></html></div>
<pre><code>##   A DNAStringSet instance of length 10
##      width seq                                              names               
##  [1]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 16244_6_18__16244...
##  [2]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 16244_6_24__16244...
##  [3]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 16244_6_31__16244...
##  [4]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 16244_6_6__16244_...
##  [5]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 17059_2_11__17059...
##  [6]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 17059_2_16__17059...
##  [7]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 17059_2_23__17059...
##  [8]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 17059_2_27__17059...
##  [9]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 17059_2_42__17059...
## [10]  1023 ATGGCGATAACAGTTTATTACGA...AGTTAATAAAGACAAAAATTAA 17150_1_73__17150...</code></pre>
<p>Note that sequence names are created by pasteing organism names and gene names, separated by a string that by default is <code>sep = '__'</code> (two underscores). This are the same as the <code>gid</code> column in the <code>$genes</code> field, and are initially set when <code>pagoo</code> object is created. If you think your dataset contain names with this separator, then you should set this parameter to other string to avoid conflicts. <code>$sequences</code> field also has predefied subsets: <code>$core_sequences</code>, <code>$shell_sequences</code>, and <code>$cloud_sequences</code>.</p>
</div>
<div id="organism-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#organism-metadata" class="anchor"></a>Organism metadata</h1>
<p>The <code>$organisms</code> field contain a table with organisms and metadata as additional columns if provided.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">p</span>$<span class="no">organisms</span></pre></body></html></div>
<pre><code>## DataFrame with 10 rows and 8 columns
##           org          id      strain      year     country        host
##      &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1   16244_6_6        FR15   2008/170h      2008      France       Human
## 2  16244_6_18        FR27   2012/185h      2012      France       Human
## 3  16244_6_24        FR33    2014/52h      2014      France       Human
## 4  16244_6_31        FR40    2009/56h      2009      France       Human
## 5  17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 6  17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 7  17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 8  17150_1_73         CA1   001A-0374      2005      Canada       Human
## 9  17059_2_11        CA22    ID136207      2014      Canada       Human
## 10 17059_2_42         TW6        1830      2008      Taiwan       Human
##         source   accession
##    &lt;character&gt; &lt;character&gt;
## 1        Feces   ERS672247
## 2        Blood   ERS672259
## 3          CSF   ERS672265
## 4          CSF   ERS672272
## 5      Prepuce   ERS739235
## 6        Fetus   ERS739242
## 7           VM   ERS739246
## 8        Blood   ERS686652
## 9        Blood   ERS739230
## 10       Blood   ERS739261</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres, Gregorio Iraola, Institut Pasteur de Montevideo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
