<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>4 - Subseting • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="4 - Subseting">
<meta property="og:description" content="pagoo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1-Quick_Start.html">1 - Quick Start</a>
    </li>
    <li>
      <a href="../articles/2-Input.html">2 - Input</a>
    </li>
    <li>
      <a href="../articles/3-Querying_Data.html">3 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/4-Subseting.html">4 - Subseting</a>
    </li>
    <li>
      <a href="../articles/5-Methods_Plots.html">5 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/6-Recipes.html">6 - Recipes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iferres/pagoo/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>4 - Subseting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/iferres/pagoo/blob/master/vignettes/4-Subseting.Rmd"><code>vignettes/4-Subseting.Rmd</code></a></small>
      <div class="hidden name"><code>4-Subseting.Rmd</code></div>

    </div>

    
    
<p>One fundamental operation when working with pangenomic data is to subset it in order to better analyze it. <code>pagoo</code> provides three convinent ways that can help in different scenarios. First we will see predefined subsets, then we will understand how to use the classic R’s <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> operator with <code>pagoo</code> objects, and finally we will see how to temporarily remove one or more organisms from the dataset, and then to recover them.</p>
<p>Start the tutorial by loading the included <em>Campylobacter</em> pangenome.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">pagoo</span>) <span class="co"># Load package</span>
<span class="no">toy_rds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'campylobacter.RDS'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'pagoo'</span>)
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fu">load_pangenomeRDS</span>(<span class="no">toy_rds</span>)</pre></body></html></div>
<div id="predefined-subsets" class="section level1">
<h1 class="hasAnchor">
<a href="#predefined-subsets" class="anchor"></a>Predefined Subsets</h1>
<p>Predefined subsets were already superficially seen in <strong>3 - Querying Data</strong> tutorial. Here we will review them and understand its notation.</p>
<table class="table">
<thead><tr class="header">
<th>-</th>
<th><code>$core_*</code></th>
<th><code>$shell_*</code></th>
<th><code>$cloud_*</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong><code>$*_genes</code></strong></td>
<td><code>$core_genes</code></td>
<td><code>$shell_genes</code></td>
<td><code>$cloud_genes</code></td>
</tr>
<tr class="even">
<td><strong><code>$*_clusters</code></strong></td>
<td><code>$core_clusters</code></td>
<td><code>$shell_clusters</code></td>
<td><code>$cloud_clusters</code></td>
</tr>
<tr class="odd">
<td><strong><code>$*_sequences</code></strong></td>
<td><code>$core_sequences</code></td>
<td><code>$shell_sequences</code></td>
<td><code>$cloud_sequences</code></td>
</tr>
</tbody>
</table>
<p>As seen in the above table, the notation is quite straightforward. Lets take <code>clusters</code> for instance to illustrate it better.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">p</span>$<span class="no">clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 2599 rows and 2 columns
##          cluster                   Pfam_Arch
##      &lt;character&gt;                 &lt;character&gt;
## 1         OG0001                2-Hacid_dh_C
## 2         OG0002     2-Hacid_dh_C;2-Hacid_dh
## 3         OG0003 2-Hacid_dh_C;ACT;2-Hacid_dh
## 4         OG0004             2Fe-2S_thioredx
## 5         OG0005         4HB_MCP_1;MCPsignal
## ...          ...                         ...
## 2595      OG2595                   zf-RING_7
## 2596      OG2596                    zf-TFIIB
## 2597      OG2597                        ZinT
## 2598      OG2598                        ZnuA
## 2599      OG2599                    ZT_dimer</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">p</span>$<span class="no">core_clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 1623 rows and 2 columns
##          cluster                   Pfam_Arch
##      &lt;character&gt;                 &lt;character&gt;
## 1         OG0001                2-Hacid_dh_C
## 2         OG0002     2-Hacid_dh_C;2-Hacid_dh
## 3         OG0003 2-Hacid_dh_C;ACT;2-Hacid_dh
## 4         OG0004             2Fe-2S_thioredx
## 5         OG0006         4HB_MCP_1;MCPsignal
## ...          ...                         ...
## 1619      OG2595                   zf-RING_7
## 1620      OG2596                    zf-TFIIB
## 1621      OG2597                        ZinT
## 1622      OG2598                        ZnuA
## 1623      OG2599                    ZT_dimer</code></pre>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">p</span>$<span class="no">shell_clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 441 rows and 2 columns
##         cluster            Pfam_Arch
##     &lt;character&gt;          &lt;character&gt;
## 1        OG0005  4HB_MCP_1;MCPsignal
## 2        OG0009        5_nucleotid_C
## 3        OG0015                  AAA
## 4        OG0019               AAA_22
## 5        OG0021               AAA_22
## ...         ...                  ...
## 437      OG2548 VirB3;CagE_TrbE_VirB
## 438      OG2552                VirB8
## 439      OG2554                VirB8
## 440      OG2565               Y1_Tnp
## 441      OG2568           YafQ_toxin</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">p</span>$<span class="no">cloud_clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 535 rows and 2 columns
##         cluster   Pfam_Arch
##     &lt;character&gt; &lt;character&gt;
## 1        OG0018      AAA_22
## 2        OG0020      AAA_22
## 3        OG0024      AAA_25
## 4        OG0025      AAA_25
## 5        OG0028      AAA_28
## ...         ...         ...
## 531      OG2566  YafQ_toxin
## 532      OG2577      YhcG_C
## 533      OG2584        YopX
## 534      OG2589  Zeta_toxin
## 535      OG2590  Zeta_toxin</code></pre>
</div>
<div id="the-operator" class="section level1">
<h1 class="hasAnchor">
<a href="#the-operator" class="anchor"></a>The <code>[</code> operator</h1>
<p>Using <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> for subset vectors, lists, or matrices in R is one of the most common operations we, as R users, do every day. Methods using <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> for subsetting can be divided into 2 types:</p>
<ol style="list-style-type: decimal">
<li>Subsetting vector or list like objects.</li>
<li>Subsetting matrix-like objects.</li>
</ol>
<p>In the following subsections we will see how to use them to subset <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, or<code>$organisms</code> fields.</p>
<div id="vectorlist-notation" class="section level2">
<h2 class="hasAnchor">
<a href="#vectorlist-notation" class="anchor"></a>Vector/List notation</h2>
<p>When we subset a vector or list in R we use a single number or a vector of numbers, lets say from <code>i</code> to <code>j</code>, to pick those elements by indexes: <code>x[i:j]</code>. In <code>pagoo</code> we implement a method for this generic function to subset data fields, where indexes represent <strong>clusters</strong>. So instead of subseting directly the final object (which you can of course) you apply this method directly to the object, and then select any of the data fields.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># From clusters 1 to 3</span>
<span class="no">p</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">pan_matrix</span></pre></body></html></div>
<pre><code>##            OG0001 OG0005 OG0018
## 16244_6_6       1      1      0
## 16244_6_18      1      1      0
## 16244_6_24      1      1      0
## 16244_6_31      1      1      0
## 17059_2_16      1      1      0
## 17059_2_23      1      0      0
## 17059_2_27      1      1      0
## 17150_1_73      1      1      0
## 17059_2_11      1      0      0
## 17059_2_42      1      0      1</code></pre>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">p</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">genes</span></pre></body></html></div>
<pre><code>## SplitDataFrameList of length 3
## $OG0001
## DataFrame with 10 rows and 10 columns
##     cluster        org             gene                          gid
##    &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt;
## 1    OG0001  16244_6_6  16244_6_6_00150   16244_6_6__16244_6_6_00150
## 2    OG0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_18_00172
## 3    OG0001 16244_6_24 16244_6_24_01512 16244_6_24__16244_6_24_01512
## 4    OG0001 16244_6_31 16244_6_31_00173 16244_6_31__16244_6_31_00173
## 5    OG0001 17059_2_16 17059_2_16_01012 17059_2_16__17059_2_16_01012
## 6    OG0001 17059_2_23 17059_2_23_01040 17059_2_23__17059_2_23_01040
## 7    OG0001 17059_2_27 17059_2_27_00492 17059_2_27__17059_2_27_00492
## 8    OG0001 17150_1_73 17150_1_73_00221 17150_1_73__17150_1_73_00221
## 9    OG0001 17059_2_11 17059_2_11_00174 17059_2_11__17059_2_11_00174
## 10   OG0001 17059_2_42 17059_2_42_00176 17059_2_42__17059_2_42_00176
##       geneName
##    &lt;character&gt;
## 1         ilvC
## 2         ilvC
## 3         ilvC
## 4         ilvC
## 5         ilvC
## 6         ilvC
## 7         ilvC
## 8         ilvC
## 9         ilvC
## 10        ilvC
##                                                                                                                                                                   product
##                                                                                                                                                               &lt;character&gt;
## 1  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 2  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 3  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 4  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 5  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 6  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 7  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 8  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 9  ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 10 ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
##                       contig      from        to      strand
##                  &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1  ERS672247|SC|contig000001    137333    138355           +
## 2  ERS672259|SC|contig000001    173541    174563           +
## 3  ERS672265|SC|contig000005    173654    174676           +
## 4  ERS672272|SC|contig000001    173541    174563           +
## 5  ERS739235|SC|contig000005     43645     44667           -
## 6  ERS739242|SC|contig000004    173578    174600           +
## 7  ERS739246|SC|contig000002    184035    185057           -
## 8  ERS686652|SC|contig000001    207206    208228           +
## 9  ERS739230|SC|contig000001    173577    174599           +
## 10 ERS739261|SC|contig000001    173569    174591           +
## 
## $OG0005
## DataFrame with 7 rows and 10 columns
##    cluster        org             gene                          gid    geneName
##   &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt; &lt;character&gt;
## 1   OG0005  16244_6_6  16244_6_6_00049   16244_6_6__16244_6_6_00049      mcpB_1
## 2   OG0005 16244_6_18 16244_6_18_00072 16244_6_18__16244_6_18_00072      mcpB_1
## 3   OG0005 16244_6_24 16244_6_24_01411 16244_6_24__16244_6_24_01411      mcpB_2
## 4   OG0005 16244_6_31 16244_6_31_00072 16244_6_31__16244_6_31_00072      mcpB_1
## 5   OG0005 17059_2_16 17059_2_16_01270 17059_2_16__17059_2_16_01270      mcpB_3
## 6   OG0005 17059_2_27 17059_2_27_00592 17059_2_27__17059_2_27_00592      mcpB_1
## 7   OG0005 17150_1_73 17150_1_73_00073 17150_1_73__17150_1_73_00073      mcpB_1
##                                                                                                                                                                            product
##                                                                                                                                                                        &lt;character&gt;
## 1 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 2 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 3 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 4 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 5 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 6 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 7 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
##                      contig      from        to      strand
##                 &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 ERS672247|SC|contig000001     46407     48068           +
## 2 ERS672259|SC|contig000001     82617     84278           +
## 3 ERS672265|SC|contig000005     82728     84389           +
## 4 ERS672272|SC|contig000001     82615     84276           +
## 5 ERS739235|SC|contig000007     82775     84436           +
## 6 ERS739246|SC|contig000002    274321    275982           -
## 7 ERS686652|SC|contig000001     82648     84309           +
## 
## $OG0018
## DataFrame with 1 row and 10 columns
##    cluster        org             gene                          gid    geneName
##   &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt; &lt;character&gt;
## 1   OG0018 17059_2_42 17059_2_42_00333 17059_2_42__17059_2_42_00333            
##                                                              product
##                                                          &lt;character&gt;
## 1 bacteriocin,putative secretion ATPase%2C PEP-CTERM locus subfamily
##                      contig      from        to      strand
##                 &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 ERS739261|SC|contig000001    303996    304949           -</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">p</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 3 rows and 2 columns
##       cluster           Pfam_Arch
##   &lt;character&gt;         &lt;character&gt;
## 1      OG0001        2-Hacid_dh_C
## 2      OG0005 4HB_MCP_1;MCPsignal
## 3      OG0018              AAA_22</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">p</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">sequences</span></pre></body></html></div>
<pre><code>## DNAStringSetList of length 3
## [["OG0001"]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTGCGA...
## [["OG0005"]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAAAAC...
## [["OG0018"]] 17059_2_42__17059_2_42_00333=ATGAGTAAATTTGAAGATATTAGGAATGAGCTTAG...</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">p</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">organisms</span></pre></body></html></div>
<pre><code>## DataFrame with 10 rows and 8 columns
##           org          id      strain      year     country        host
##      &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1   16244_6_6        FR15   2008/170h      2008      France       Human
## 2  16244_6_18        FR27   2012/185h      2012      France       Human
## 3  16244_6_24        FR33    2014/52h      2014      France       Human
## 4  16244_6_31        FR40    2009/56h      2009      France       Human
## 5  17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 6  17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 7  17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 8  17150_1_73         CA1   001A-0374      2005      Canada       Human
## 9  17059_2_11        CA22    ID136207      2014      Canada       Human
## 10 17059_2_42         TW6        1830      2008      Taiwan       Human
##         source   accession
##    &lt;character&gt; &lt;character&gt;
## 1        Feces   ERS672247
## 2        Blood   ERS672259
## 3          CSF   ERS672265
## 4          CSF   ERS672272
## 5      Prepuce   ERS739235
## 6        Fetus   ERS739242
## 7           VM   ERS739246
## 8        Blood   ERS686652
## 9        Blood   ERS739230
## 10       Blood   ERS739261</code></pre>
<p>Note that in each case, indexes can be interpreted as clusters. In <code>pan_matrix</code> they refer to columns, in <code>genes</code> and <code>sequences</code> to elements in a list, in <code>clusters</code> rows in the dataframe, and in the case of <code>$organisms</code> it returns the organisms (rows) where those clusters are present. See what happen for this last case when we query for organisms present for a shell cluster:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">shell_clust</span> <span class="kw">&lt;-</span> <span class="no">p</span>$<span class="no">shell_clusters</span>$<span class="no">cluster</span>[<span class="fl">1</span>] <span class="co"># [1] "OG0005"</span>
<span class="no">p</span>[<span class="no">shell_clust</span>]$<span class="no">organisms</span></pre></body></html></div>
<pre><code>## DataFrame with 7 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1  16244_6_6        FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 16244_6_24        FR33    2014/52h      2014      France       Human
## 4 16244_6_31        FR40    2009/56h      2009      France       Human
## 5 17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 7 17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 8 17150_1_73         CA1   001A-0374      2005      Canada       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3         CSF   ERS672265
## 4         CSF   ERS672272
## 5     Prepuce   ERS739235
## 7          VM   ERS739246
## 8       Blood   ERS686652</code></pre>
<p>Only (7) organisms which contain that shell cluster will be listed.</p>
</div>
<div id="matrix-notation" class="section level2">
<h2 class="hasAnchor">
<a href="#matrix-notation" class="anchor"></a>Matrix notation</h2>
<p>The other use of <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> notation is when we subset a matrix-like object. In this case we provide 2 sets of indexes: the first for rows, and the second for columns, separated by a coma. <code>pagoo</code> interprets these indexes as organisms (rows) and clusters (columns), so you are referencing a set of genes (cell value). Basically you are referencing cells in the pan_matrix.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">p</span>$<span class="no">pan_matrix</span>[<span class="fl">1</span>:<span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]</pre></body></html></div>
<pre><code>##            OG0001 OG0005 OG0018
## 16244_6_6       1      1      0
## 16244_6_18      1      1      0
## 16244_6_24      1      1      0</code></pre>
<p>Here we are selecting the first 3 organisms, and the clusters indexed as 1, 5, and 18. Note that we see (sum columns) 3 genes in the first cluster, 3 in the second, and 0 in the third. Now we flip the notation and subset directly the <code>pagoo</code> object, and then ask for a data field:</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">p</span>[<span class="fl">1</span>:<span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">pan_matrix</span> <span class="co"># The same as above</span></pre></body></html></div>
<pre><code>##            OG0001 OG0005 OG0018
## 16244_6_6       1      1      0
## 16244_6_18      1      1      0
## 16244_6_24      1      1      0</code></pre>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">p</span>[<span class="fl">1</span>:<span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">genes</span></pre></body></html></div>
<pre><code>## SplitDataFrameList of length 2
## $OG0001
## DataFrame with 3 rows and 10 columns
##    cluster        org             gene                          gid    geneName
##   &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt; &lt;character&gt;
## 1   OG0001  16244_6_6  16244_6_6_00150   16244_6_6__16244_6_6_00150        ilvC
## 2   OG0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_18_00172        ilvC
## 3   OG0001 16244_6_24 16244_6_24_01512 16244_6_24__16244_6_24_01512        ilvC
##                                                                                                                                                                  product
##                                                                                                                                                              &lt;character&gt;
## 1 ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 2 ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
## 3 ketol-acid reductoisomerase,Ketol-acid reductoisomerase,ketol-acid reductoisomerase,ketol-acid reductoisomerase,Acetohydroxy acid isomeroreductase%2C catalytic domain
##                      contig      from        to      strand
##                 &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 ERS672247|SC|contig000001    137333    138355           +
## 2 ERS672259|SC|contig000001    173541    174563           +
## 3 ERS672265|SC|contig000005    173654    174676           +
## 
## $OG0005
## DataFrame with 3 rows and 10 columns
##    cluster        org             gene                          gid    geneName
##   &lt;factor&gt;   &lt;factor&gt;      &lt;character&gt;                  &lt;character&gt; &lt;character&gt;
## 1   OG0005  16244_6_6  16244_6_6_00049   16244_6_6__16244_6_6_00049      mcpB_1
## 2   OG0005 16244_6_18 16244_6_18_00072 16244_6_18__16244_6_18_00072      mcpB_1
## 3   OG0005 16244_6_24 16244_6_24_01411 16244_6_24__16244_6_24_01411      mcpB_2
##                                                                                                                                                                            product
##                                                                                                                                                                        &lt;character&gt;
## 1 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 2 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
## 3 methyl-accepting chemotaxis protein,H3,methyl-accepting protein IV,Predicted periplasmic ligand-binding sensor domain,Methyl-accepting chemotaxis protein (MCP) signaling domain
##                      contig      from        to      strand
##                 &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
## 1 ERS672247|SC|contig000001     46407     48068           +
## 2 ERS672259|SC|contig000001     82617     84278           +
## 3 ERS672265|SC|contig000005     82728     84389           +</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">p</span>[<span class="fl">1</span>:<span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">clusters</span></pre></body></html></div>
<pre><code>## DataFrame with 2 rows and 2 columns
##       cluster           Pfam_Arch
##   &lt;character&gt;         &lt;character&gt;
## 1      OG0001        2-Hacid_dh_C
## 2      OG0005 4HB_MCP_1;MCPsignal</code></pre>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">p</span>[<span class="fl">1</span>:<span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">sequences</span></pre></body></html></div>
<pre><code>## DNAStringSetList of length 2
## [["OG0001"]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTGCGA...
## [["OG0005"]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAAAAC...</code></pre>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">p</span>[<span class="fl">1</span>:<span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span>)]$<span class="no">organisms</span></pre></body></html></div>
<pre><code>## DataFrame with 3 rows and 8 columns
##          org          id      strain      year     country        host
##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1  16244_6_6        FR15   2008/170h      2008      France       Human
## 2 16244_6_18        FR27   2012/185h      2012      France       Human
## 3 16244_6_24        FR33    2014/52h      2014      France       Human
##        source   accession
##   &lt;character&gt; &lt;character&gt;
## 1       Feces   ERS672247
## 2       Blood   ERS672259
## 3         CSF   ERS672265</code></pre>
<p>So this selection returns a list of length 2, with 3 elements each for <code>$genes</code> and <code>$sequences</code>, a dataframe showing only the selected clusters in the <code>$clusters</code> field, and a dataframe showing only the selected organisms in the <code>$organisms</code> field.</p>
<p>We found this implemetation quite useful for both data exploration and fine grained analysis.</p>
</div>
</div>
<div id="dropping-and-recovering-organisms" class="section level1">
<h1 class="hasAnchor">
<a href="#dropping-and-recovering-organisms" class="anchor"></a>Dropping and Recovering Organisms</h1>
<p>One useful feature implemented in <code>pagoo</code> is the possibility of easily removing (hiding) organisms from the dataset. This is useful if during the analysis we identify some genome with weird charactersitics (i.e. potentially contaminated), or if we want to focus just in a subset of genomes of interest given any metadata value, or if we included an outgroup for phylogenetic purposes but we want to remove it from downstream analyses. Let’s see how it works..</p>
<div id="dropping-organisms" class="section level2">
<h2 class="hasAnchor">
<a href="#dropping-organisms" class="anchor"></a>Dropping organisms</h2>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">p</span>$<span class="no">organisms</span></pre></body></html></div>
<pre><code>## DataFrame with 10 rows and 8 columns
##           org          id      strain      year     country        host
##      &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1   16244_6_6        FR15   2008/170h      2008      France       Human
## 2  16244_6_18        FR27   2012/185h      2012      France       Human
## 3  16244_6_24        FR33    2014/52h      2014      France       Human
## 4  16244_6_31        FR40    2009/56h      2009      France       Human
## 5  17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 6  17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 7  17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 8  17150_1_73         CA1   001A-0374      2005      Canada       Human
## 9  17059_2_11        CA22    ID136207      2014      Canada       Human
## 10 17059_2_42         TW6        1830      2008      Taiwan       Human
##         source   accession
##    &lt;character&gt; &lt;character&gt;
## 1        Feces   ERS672247
## 2        Blood   ERS672259
## 3          CSF   ERS672265
## 4          CSF   ERS672272
## 5      Prepuce   ERS739235
## 6        Fetus   ERS739242
## 7           VM   ERS739246
## 8        Blood   ERS686652
## 9        Blood   ERS739230
## 10       Blood   ERS739261</code></pre>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">p</span>$<span class="no">summary_stats</span></pre></body></html></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total      2599
## 2        Core      1623
## 3       Shell       441
## 4       Cloud       535</code></pre>
<p>We can see that we have 10 organisms, and we have the pangenome summary statistics. Let’s say, for instance, that we want to exclude french isolates (because any reason) from the dataset, which are indexed from 1 to 4 (rows) in the above dataframe.</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">p</span>$<span class="fu"><a href="https://rdrr.io/r/base/drop.html">drop</a></span>(<span class="fl">1</span>:<span class="fl">4</span>)
<span class="no">p</span>$<span class="no">organisms</span> <span class="co"># Updated organims !!</span></pre></body></html></div>
<pre><code>## DataFrame with 6 rows and 8 columns
##           org          id      strain      year     country        host
##      &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 5  17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 6  17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 7  17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 8  17150_1_73         CA1   001A-0374      2005      Canada       Human
## 9  17059_2_11        CA22    ID136207      2014      Canada       Human
## 10 17059_2_42         TW6        1830      2008      Taiwan       Human
##         source   accession
##    &lt;character&gt; &lt;character&gt;
## 5      Prepuce   ERS739235
## 6        Fetus   ERS739242
## 7           VM   ERS739246
## 8        Blood   ERS686652
## 9        Blood   ERS739230
## 10       Blood   ERS739261</code></pre>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">p</span>$<span class="no">summary_stats</span> <span class="co"># Updated stats !!</span></pre></body></html></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##      Category    Number
##   &lt;character&gt; &lt;integer&gt;
## 1       Total      2532
## 2        Core      1635
## 3       Shell       346
## 4       Cloud       551</code></pre>
<p>As you see, not only the french organisms have been removed from the dataset, but also summary statistics have been updated! This is also true for any other data field: <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, and<code>$organisms</code> fields are automatically updated. Also any embebbed statistical or visualization methods will now only consider available organisms/clusters/genes. When we drop an organims, we are hiding all features associated to that genomes including genes, sequences, gene clusters and metadata.</p>
<p><strong>It’s important to note that you don’t have to reassign the object to a new one, it is self modified (in place modification). That’s R6 reference semantics (see R6 documentation for details), use with caution.</strong></p>
</div>
<div id="recovering-dropped-organisms" class="section level2">
<h2 class="hasAnchor">
<a href="#recovering-dropped-organisms" class="anchor"></a>Recovering dropped organisms</h2>
<p>You can recover dropped organisms. To see if we have any hidden organism, we use the <code>$dropped</code> field, and then recover it using its index or its name.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">p</span>$<span class="no">dropped</span></pre></body></html></div>
<pre><code>##            1            2            3            4 
##  "16244_6_6" "16244_6_18" "16244_6_24" "16244_6_31"</code></pre>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">p</span>$<span class="fu"><a href="https://rdrr.io/r/utils/recover.html">recover</a></span>( <span class="no">p</span>$<span class="no">dropped</span> )
<span class="no">p</span>$<span class="no">organisms</span></pre></body></html></div>
<pre><code>## DataFrame with 10 rows and 8 columns
##           org          id      strain      year     country        host
##      &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
## 1   16244_6_6        FR15   2008/170h      2008      France       Human
## 2  16244_6_18        FR27   2012/185h      2012      France       Human
## 3  16244_6_24        FR33    2014/52h      2014      France       Human
## 4  16244_6_31        FR40    2009/56h      2009      France       Human
## 5  17059_2_16         AR1      99/801      1999   Argentina      Bovine
## 6  17059_2_23         AR8      04/875      2004   Argentina      Bovine
## 7  17059_2_27        AR12      06/195      2006   Argentina      Bovine
## 8  17150_1_73         CA1   001A-0374      2005      Canada       Human
## 9  17059_2_11        CA22    ID136207      2014      Canada       Human
## 10 17059_2_42         TW6        1830      2008      Taiwan       Human
##         source   accession
##    &lt;character&gt; &lt;character&gt;
## 1        Feces   ERS672247
## 2        Blood   ERS672259
## 3          CSF   ERS672265
## 4          CSF   ERS672272
## 5      Prepuce   ERS739235
## 6        Fetus   ERS739242
## 7           VM   ERS739246
## 8        Blood   ERS686652
## 9        Blood   ERS739230
## 10       Blood   ERS739261</code></pre>
<p>Now french isolates and all its features are again available.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres, Gregorio Iraola, Institut Pasteur de Montevideo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
