<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pagoo">
<title>Input • pagoo</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Input">
<meta property="og:description" content="pagoo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pagoo</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.18</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/pagoo.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Input.html">Input</a>
    <a class="dropdown-item" href="../articles/Methods_Plots.html">Methods and Plots</a>
    <a class="dropdown-item" href="../articles/Querying_Data.html">Querying Data</a>
    <a class="dropdown-item" href="../articles/Recipes.html">Recipes</a>
    <a class="dropdown-item" href="../articles/Subseting.html">Subseting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/iferres/pagoo/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Input</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/iferres/pagoo/blob/HEAD/vignettes/Input.Rmd" class="external-link"><code>vignettes/Input.Rmd</code></a></small>
      <div class="d-none name"><code>Input.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="input-from-scratch">Input From Scratch<a class="anchor" aria-label="anchor" href="#input-from-scratch"></a>
</h2>
<p>In order to generate a <code>pagoo</code> object, the only mandatory
data structure is a <code>data.frame</code> which has to contain 3 basic
columns. Let’s illustrate it with an toy dataset.</p>
<p>First load <code>pagoo</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://iferres.github.io/pagoo/">pagoo</a></span><span class="op">)</span> <span class="co"># Load package</span></span>
<span><span class="va">tgz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'toy_data.tar.gz'</span>, package <span class="op">=</span> <span class="st">'pagoo'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="va">tgz</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># Decompress example dataset</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">'tsv$|fasta$'</span><span class="op">)</span> <span class="co"># List files</span></span>
<span><span class="va">files</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/tmp/RtmpdkgAUX/case_clusters_meta.tsv"</span></span>
<span><span class="co">## [2] "/tmp/RtmpdkgAUX/case_df.tsv"           </span></span>
<span><span class="co">## [3] "/tmp/RtmpdkgAUX/case_orgs_meta.tsv"    </span></span>
<span><span class="co">## [4] "/tmp/RtmpdkgAUX/organismA.fasta"       </span></span>
<span><span class="co">## [5] "/tmp/RtmpdkgAUX/organismB.fasta"       </span></span>
<span><span class="co">## [6] "/tmp/RtmpdkgAUX/organismC.fasta"       </span></span>
<span><span class="co">## [7] "/tmp/RtmpdkgAUX/organismD.fasta"       </span></span>
<span><span class="co">## [8] "/tmp/RtmpdkgAUX/organismE.fasta"</span></span></code></pre>
<p>The file we need now is <code>case_df.tsv</code>. Lets load it and
see what’s its structure:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"case_df.tsv"</span>, <span class="va">files</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">data_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      gene       org cluster                             annot</span></span>
<span><span class="co">## 1 gene081 organismA   OG001  Thioesterase superfamily protein</span></span>
<span><span class="co">## 2 gene122 organismB   OG001          Thioesterase superfamily</span></span>
<span><span class="co">## 3 gene299 organismC   OG001  Thioesterase superfamily protein</span></span>
<span><span class="co">## 4 gene186 organismD   OG001  Thioesterase superfamily protein</span></span>
<span><span class="co">## 5 gene076 organismE   OG001          Thioesterase superfamily</span></span>
<span><span class="co">## 6 gene352 organismA   OG002 Inherit from proNOG: Thioesterase</span></span></code></pre>
<p>So it is a <code>data.frame</code> with 4 columns. The first one with
the name of each gene, the second one with the organism to which each
gene belongs, the third one with the cluster to which each gene was
assigned in the pangenome reconstruction, and the last one with
annotation metadata for each gene. Of the 4 columns, the former 3 are
required, and <code>pagoo</code> will look for columns named “gene”,
“org”, and “cluster”. More columns are optional, and you can add as many
as you want (or none) to add metadata of each gene.</p>
<p>With only this data (even ignoring the fourth column, which is
metadata), you can start working with pagoo:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pagoo.html">pagoo</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>but let’s continue adding more.</p>
<div class="section level3">
<h3 id="including-organism-metadata">Including organism metadata<a class="anchor" aria-label="anchor" href="#including-organism-metadata"></a>
</h3>
<p>The next 2 <code>.tsv</code> files contains metadata for each cluster
and for each organism, respectively, and are optional arguments.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Organism metadata</span></span>
<span><span class="va">orgs_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"case_orgs_meta.tsv"</span>, <span class="va">files</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">orgs_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">orgs_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">orgs_meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         org sero  country</span></span>
<span><span class="co">## 1 organismA    a Westeros</span></span>
<span><span class="co">## 2 organismB    b Westeros</span></span>
<span><span class="co">## 3 organismC    c Westeros</span></span>
<span><span class="co">## 4 organismD    a    Essos</span></span>
<span><span class="co">## 5 organismE    b    Essos</span></span></code></pre>
<p>On this <code>data.frame</code> we have a column named
<code>org</code> which is mandatory in case you provide this argument.
Other columns are metadata associated to each organism. Beware that
organisms provided in this table (<code>orgs_meta$org</code>) must
coincide with the names provided in the <code>data$org</code> field, in
order to correctly map each variables.</p>
</div>
<div class="section level3">
<h3 id="including-cluster-metadata">Including cluster metadata<a class="anchor" aria-label="anchor" href="#including-cluster-metadata"></a>
</h3>
<p>Last file contains metadata associated with each cluster of
orthologous:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster metadata</span></span>
<span><span class="va">clust_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"case_clusters_meta.tsv"</span>, <span class="va">files</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clust_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">clust_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">'\t'</span>, quote <span class="op">=</span> <span class="st">''</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clust_meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   cluster   kegg  cog</span></span>
<span><span class="co">## 1   OG001   &lt;NA&gt;    S</span></span>
<span><span class="co">## 2   OG002   &lt;NA&gt;    S</span></span>
<span><span class="co">## 3   OG003   &lt;NA&gt; &lt;NA&gt;</span></span>
<span><span class="co">## 4   OG004   &lt;NA&gt;    D</span></span>
<span><span class="co">## 5   OG005 K01990    V</span></span>
<span><span class="co">## 6   OG006   &lt;NA&gt;    V</span></span></code></pre>
<p>Again, the column <code>clust_meta$cluster</code> must contain the
same elements as <code>data$cluster</code> column to be able to map one
into the other.</p>
<p>With all this data the <code>pagoo</code> object will look much more
complete. But you can still add sequence information to the pangenome,
which makes it much more useful and interesting to work with.</p>
</div>
<div class="section level3">
<h3 id="including-sequences">Including sequences<a class="anchor" aria-label="anchor" href="#including-sequences"></a>
</h3>
<p>In this made up dataset we have 5 organisms, so if you decide to add
sequences to the pangenome you must provide them for all 5 organisms.
The type of data needed is a DNA multifasta file for each organism, in
which each sequence is a gene whose name can be mapped to the
<code>data$gene</code> column. You must first load the sequences into a
<code>list</code>, and name each list element as the organism provided
in <code>data$org</code> (as well as <code>org_meta$org</code>). The
<code>list</code> would look something like:</p>
<ol style="list-style-type: decimal">
<li>organism1
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneN</li>
</ul>
</li>
<li>organism2
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneM</li>
</ul>
</li>
<li>organism3
<ul>
<li>gene1</li>
<li>gene2</li>
<li>…</li>
<li>geneP</li>
</ul>
</li>
<li>…</li>
</ol>
<p>In the case of the example we are working on:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sequences</span></span>
<span><span class="va">fasta_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"[.]fasta"</span>, <span class="va">files</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># List fasta files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fasta_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">'[.]fasta'</span>, <span class="st">''</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fasta_files</span><span class="op">)</span><span class="op">)</span> <span class="co"># Name them </span></span>
<span><span class="co"># Read fasta files with Biostrings:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="va">sq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fasta_files</span>, <span class="va">readDNAStringSet</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="co"># Is list?</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="co"># One list element per organism</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="co"># Names are the same as in data$org</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "organismA" "organismB" "organismC" "organismD" "organismE"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sq</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># Class of each element of the list</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "DNAStringSet"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Biostrings"</span></span></code></pre>
<p>And we have a <code>list</code> of <code>DNAStringSet</code>
(Biostrings package). Now we can load a quite complete
<code>pagoo</code> object (you could still add more metadata to genes,
clusters, or organisms):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pagoo.html">pagoo</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span>, <span class="co"># Required data</span></span>
<span>           org_meta <span class="op">=</span> <span class="va">orgs_meta</span>, <span class="co"># Organism's metadata</span></span>
<span>           cluster_meta <span class="op">=</span> <span class="va">clust_meta</span>, <span class="co"># Cluster's metadata</span></span>
<span>           sequences <span class="op">=</span> <span class="va">sq</span><span class="op">)</span> <span class="co"># Sequences</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="input-from-pangenome-reconstruction-software">Input From Pangenome Reconstruction Software<a class="anchor" aria-label="anchor" href="#input-from-pangenome-reconstruction-software"></a>
</h2>
<p>All the above stuff with preparing data and loading classes seems
difficult and time-consuming, but in real life working datasets this
will be rarely needed. We are explaining it here to provide full details
about how the software works, but this package also provides functions
to automatically read-in output files from pangenome reconstruction
software into <code>pagoo</code>, avoiding any formatting or
manipulation of data.</p>
<p>Currently <code>pagoo</code> supports input from <a href="https://sanger-pathogens.github.io/Roary/" class="external-link">roary</a> (Page et al.,
2015), which has been the standard and most cited software for pangenome
reconstruction, and <a href="https://github.com/gtonkinhill/panaroo" class="external-link">panaroo</a> (Tonkin-Hill
et al., 2020). It is worth noticing that as roary become the most used
software in this field, other tools as <a href="https://github.com/zheminzhou/PEPPAN" class="external-link">PEPPAN</a>, <a href="https://github.com/SionBayliss/PIRATE" class="external-link">PRIATE</a> and panaroo
include scripts to convert their output to roary’s format. To work with
<code>roary</code>’s output, please refer to <code><a href="../reference/roary_2_pagoo.html">?roary_2_pagoo</a></code>
documentation. Although panaroo also includes a tool for this, we
provide a function to load their native output format, see
<code><a href="../reference/panaroo_2_pagoo.html">?panaroo_2_pagoo</a></code>. For both functions you will only need the
<code>.gff</code> files used as input for roary of panaroo, and the
<code>gene_presence_absence.csv</code> file.</p>
<p>Also, we have created our own pangenome reconstruction software
called <a href="https://github.com/iferres/pewit" class="external-link">pewit</a> (Ferrés et
al., still unpublished), which automatically generates a
<code>pagoo</code>-like object to perform downstream analyses. This
object contain all the methods and fields <code>pagoo</code> provides,
plus a set of methods and fields exclusive to this software.</p>
<p>Other good pangenome reconstruction software already exists like <a href="https://pangenome.org/" class="external-link">PanX</a> (Ding et al., 2018), <a href="https://github.com/cran/micropan" class="external-link">micropan</a> (Snipen &amp;
Liland, 2015), <a href="https://github.com/eead-csic-compbio/get_homologues" class="external-link">GET_HOMOLOGUES</a>
(Contreras-Moreira &amp; Vinuesa, 2013), among others. We plan to
provide support to some of them in the future.</p>
</div>
<div class="section level2">
<h2 id="adding-metadata-after-object-creation">Adding Metadata After Object Creation<a class="anchor" aria-label="anchor" href="#adding-metadata-after-object-creation"></a>
</h2>
<p>After object creation, you may want to add new metadata given new
information or as result of posterior analyses. <code>pagoo</code>
objects include a function to add columns of metadata to each gene, each
cluster, or each organism. To illustrate it, we will add a new column to
the <code>$organisms</code> field named <code>host</code> to add made up
information about the host where each genome was isolated from.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">host_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>org <span class="op">=</span> <span class="va">p</span><span class="op">$</span><span class="va">organisms</span><span class="op">$</span><span class="va">org</span>, host <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cow"</span>, <span class="st">"Dog"</span>, <span class="st">"Cat"</span>, <span class="st">"Cow"</span>, <span class="st">"Sheep"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="fu">add_metadata</span><span class="op">(</span>map <span class="op">=</span> <span class="st">"org"</span>, <span class="va">host_df</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">organisms</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 5 rows and 4 columns</span></span>
<span><span class="co">##         org        sero     country        host</span></span>
<span><span class="co">##    &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1 organismA           a    Westeros         Cow</span></span>
<span><span class="co">## 2 organismB           b    Westeros         Dog</span></span>
<span><span class="co">## 3 organismC           c    Westeros         Cat</span></span>
<span><span class="co">## 4 organismD           a       Essos         Cow</span></span>
<span><span class="co">## 5 organismE           b       Essos       Sheep</span></span></code></pre>
<p>In order to allow <code>pagoo</code> correct data mapping, the values
in the first column of the metadata table should be available at
<code>p$organisms$org</code>, and its column header must also be named
<code>org</code>.</p>
<p>As said, you can add <code>gene</code> or <code>cluster</code>
metadata following the same idea.</p>
</div>
<div class="section level2">
<h2 id="saving-and-loading-pagoo-objects">Saving and Loading <code>pagoo</code> objects<a class="anchor" aria-label="anchor" href="#saving-and-loading-pagoo-objects"></a>
</h2>
<p>Once loaded, it has two methods for saving and reloading to a new R
session. The first one is by saving them as flat (text) files:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pangenome"</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="fu">write_pangenome</span><span class="op">(</span>dir <span class="op">=</span> <span class="va">tmp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">tmp</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/tmp/RtmpdkgAUX/pangenome/clusters.tsv" </span></span>
<span><span class="co">## [2] "/tmp/RtmpdkgAUX/pangenome/data.tsv"     </span></span>
<span><span class="co">## [3] "/tmp/RtmpdkgAUX/pangenome/organisms.tsv"</span></span></code></pre>
<p>This creates a directory with 3 text files. The advantage of this
approach is that you can analyze it outside R, the disadvantage is that
from a reproducibility point of view reading text could be less stable
since class or number precision can be lost, and also you can’t save the
state of the object in any given time. Only available
organisms/genes/clusters are saved, and if you reload the class using
the tsv files, any previously dropped organisms/gene/cluster won’t be
available any more. (For information about dropping/recovering
organisms, see “Subset” tutorial).</p>
<p>If you want to save the object and continue working with it in other
R session, we recommend to save them as R objects with the RDS methods
provided:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pangenome.RDS"</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="fu">save_pangenomeRDS</span><span class="op">(</span>file <span class="op">=</span> <span class="va">rds</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_pangenomeRDS.html">load_pangenomeRDS</a></span><span class="op">(</span><span class="va">rds</span><span class="op">)</span></span></code></pre></div>
<p>This method is more stable (compatible between <code>pagoo</code>
versions), secure (uses the same metadata classes, and precision isn’t
lost), and convenient (the exact state of the object saved is restored,
keeping dropped organisms/genes/clusters hidden, available to be
recovered, and other object state configuration,
e.g. <code>core_level</code>, is also saved).</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Ignacio Ferres, Gregorio Iraola, Institut Pasteur de Montevideo.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
