<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Recipes â€¢ pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Recipes">
<meta property="og:description" content="pagoo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-164668935-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-164668935-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.4.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/pagoo.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Input.html">1 - Input</a>
    </li>
    <li>
      <a href="../articles/Querying_Data.html">2 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/Subseting.html">3 - Subseting</a>
    </li>
    <li>
      <a href="../articles/Methods_Plots.html">4 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/Recipes.html">5 - Recipes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iferres/pagoo/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Recipes_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Recipes</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/iferres/pagoo/blob/master/vignettes/Recipes.Rmd"><code>vignettes/Recipes.Rmd</code></a></small>
      <div class="hidden name"><code>Recipes.Rmd</code></div>

    </div>

    
    
<p>Here we introduce our cookbook which is made of recipes. Each recipe is a customized piece of code to complete different advanced tasks. By using these recipes (or creating new ones) the user can take full profit of functionalities provided by <code>pagoo</code> and easily interact with other <code>R</code> packages to perform a variety of analyses including phylogenetics, pangenome-wide association studies, sequence comparisons, ecological measures, prepare publication-quality figures, among others. If you want to standarize any analysis from your pangenome data that is not covered in this tutorial, we can help you and write a recipe. Contact us!</p>
<p>Start by loading the <code>pagoo</code> object, as in previous tutorials:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://iferres.github.io/pagoo">pagoo</a></span>) <span class="co"># Load package</span>
<span class="kw">toy_rds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'campylobacter.RDS'</span>, package = <span class="st">'pagoo'</span>)
<span class="kw">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_pangenomeRDS.html">load_pangenomeRDS</a></span>(<span class="kw">toy_rds</span>)
</pre></div>
<div id="publication-quality-figures" class="section level1">
<h1 class="hasAnchor">
<a href="#publication-quality-figures" class="anchor"></a>Publication-quality figures</h1>
<p>In the section <strong>5 - Methods and Plots</strong> we introduced a set of several plots that <code>pagoo</code> can generate for the basic exploration of pangenome features. These plots are generated using <code>ggplot2</code>, so their aesthetic features can be easily improved. Here we show how <code>pagoo</code> interacts with <code>ggplot2</code> and some of its extensions to produce publication-quality figures from the previous standard pangenome plots, allowing flexible styling and reproducible generation of figures. Some customization has already been shown in previous tutorials.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://patchwork.data-imaginist.com">patchwork</a></span>)

<span class="co"># 1. Pangenome curves</span>
<span class="kw">curves</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">gg_curves</span>() <span class="op">+</span>                                     <span class="co"># Plot core- and pan-genome curves</span>
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(values = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'black'</span>, <span class="st">'black'</span>)) <span class="op">+</span>  <span class="co"># Customize line colors</span>
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(alpha = <span class="fl">.05</span>, size = <span class="fl">4</span>, color = <span class="st">'grey'</span>) <span class="op">+</span> <span class="co"># Add semi-transparent data points</span>
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>(base_size = <span class="fl">15</span>) <span class="op">+</span>                          <span class="co"># Customize background theme</span>
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(legend.position = <span class="st">'none'</span>,                     <span class="co"># Remove legend</span>
        axis.title = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">12</span>),                     <span class="co"># Customize axis title</span>
        axis.text = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">12</span>))                      <span class="co"># Customize axis text size</span>

<span class="co"># 2. Gene frequency bar plots</span>
<span class="kw">bars</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">gg_barplot</span>() <span class="op">+</span>                                       <span class="co"># Plot gene frequency distribution</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>(base_size = <span class="fl">15</span>) <span class="op">+</span>                                   <span class="co"># Customize background color</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(axis.title = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">12</span>),                  <span class="co"># Customize axis label size</span>
        axis.text=<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">12</span>)) <span class="op">+</span>                   <span class="co"># Customize axis text size</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(stat = <span class="st">'identity'</span>, color = <span class="st">'black'</span>, fill = <span class="st">'black'</span>) <span class="co"># Customize bar color and borders</span>

<span class="co"># 3. PCA of accessory genes colored by host</span>
<span class="kw">pca</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">gg_pca</span>(colour = <span class="st">'host'</span>, size = <span class="fl">4</span>) <span class="op">+</span>                <span class="co"># Plot PCA, color by host</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>(base_size = <span class="fl">15</span>) <span class="op">+</span>                                <span class="co"># Customize background theme</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(legend.position = <span class="st">'bottom'</span>) <span class="op">+</span>                       <span class="co"># Customize legend position</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(axis.title = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">12</span>),               <span class="co"># Customize axis title</span>
        axis.text = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">12</span>))                <span class="co"># Customize axis text size</span>

<span class="co"># 4. Pie chart of core and accessory genes</span>
<span class="kw">pie</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">gg_pie</span>() <span class="op">+</span>                                         <span class="co"># Plot pie chart</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>(base_size = <span class="fl">15</span>) <span class="op">+</span>                                <span class="co"># Customize background theme</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html">scale_fill_discrete</a></span>(guide = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(keywidth = <span class="fl">.75</span>,
                                           keyheight = <span class="fl">.75</span>)) <span class="op">+</span> <span class="co"># Customize fill</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(palette = <span class="st">"Blues"</span>) <span class="op">+</span>                    <span class="co"># Customize fill color</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span>(breaks = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">25</span>, <span class="fl">50</span>, <span class="fl">75</span>)) <span class="op">+</span>             <span class="co"># Customize axis scales</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(legend.position = <span class="st">'bottom'</span>,                         <span class="co"># Customize legend position</span>
        legend.title = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),                     <span class="co"># Remove legend title</span>
        legend.text = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(size = <span class="fl">10</span>),              <span class="co"># Change legend text size</span>
        legend.margin = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span>(<span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">13</span>, <span class="fl">0</span>),                <span class="co"># Change legend margins</span>
        legend.box.margin = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span>(<span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">0</span>),             <span class="co"># Change box margins</span>
        axis.title.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),                     <span class="co"># Remove X-axis title</span>
        axis.title.y = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),                     <span class="co"># Remove Y-axis title</span>
        axis.ticks = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),                       <span class="co"># Remove axis ticks</span>
        axis.text.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>())                      <span class="co"># Remove X-axis text</span>


<span class="co"># 5. Use patchwork to arrange plots using math operators</span>
(<span class="kw">curves</span> <span class="op">+</span> <span class="kw">bars</span>) <span class="op">/</span> (<span class="kw">pca</span> <span class="op">+</span> <span class="kw">pie</span>)
</pre></div>
<p><img src="Recipes_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div id="using-micropan-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#using-micropan-methods" class="anchor"></a>Using <code>micropan</code> Methods</h2>
<div id="estimation-of-core-and-pangenome-sizes" class="section level3">
<h3 class="hasAnchor">
<a href="#estimation-of-core-and-pangenome-sizes" class="anchor"></a>Estimation of Core and Pangenome Sizes</h3>
<p>Given a panmatrix, you could predict the number of core clusters and the total number of clusters in a pangenome. Several methods have been developed for this purpose, in this case a binomial mixture model implemented by <code>micropan</code> package is used to describe the distribution of gene clusters across genomes in a pangenome.</p>
<p>To use this method, simply pass the objectâ€™s pan-matrix to the <code><a href="https://rdrr.io/pkg/micropan/man/binomixEstimate.html">micropan::binomixEstimate</a></code> function:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">micropan</span>)

<span class="kw">micropan</span>::<span class="fu"><a href="https://rdrr.io/pkg/micropan/man/binomixEstimate.html">binomixEstimate</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">pan_matrix</span>)
</pre></div>
<p>Given that we are using a very small toy dataset, results here are not really meaningful.</p>
</div>
<div id="fluidity" class="section level3">
<h3 class="hasAnchor">
<a href="#fluidity" class="anchor"></a>Fluidity</h3>
<p>Genomic fluidity is a measure of population diversity. Itâ€™s somehow similar to computing jaccard distances, but genomic fluidity describes the whole population whereas jaccard distances are computed pairwise. As above, we are passing the objectâ€™s pan-matrix to the <code><a href="https://rdrr.io/pkg/micropan/man/fluidity.html">micropan::fluidity</a></code> function. In this case, lets use <code>magrittr</code>â€™s <code><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></code> operator to pipe it, and compute fluidity using 100 random samples:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">micropan</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)

<span class="kw">p</span><span class="op">$</span><span class="kw">pan_matrix</span> <span class="op">%&gt;%</span> <span class="kw">micropan</span>::<span class="fu"><a href="https://rdrr.io/pkg/micropan/man/fluidity.html">fluidity</a></span>(n.sim = <span class="fl">100</span>)
</pre></div>
</div>
</div>
</div>
<div id="blastp" class="section level1">
<h1 class="hasAnchor">
<a href="#blastp" class="anchor"></a>Blastp</h1>
<p>After you have reconstructed the pangenome, itâ€™s a common practice to use a single representative gene from each cluster to annotate them using a curated database such as COG, EggNOG or KEGG. After doing that, you probably want to add any new information to the <code>pagoo</code> object as clusterâ€™s metadata. In this recipe we show how to extract one representative sequence from each cluster, translate each of them, and use <a href="https://github.com/mhahsler/rBLAST">rBLAST</a> package to query a hypothetical blastp database. <strong>NOTE:</strong> we didnâ€™t include any blast database in the package, the code here is just to illustrate the general idea. Running the following chunk will throw an error.</p>
<p><strong>DEVELOPER NOTE: ADDING METADATA STEP MISSING</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># Load required packages</span>
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="kw">rBlast</span>)) <span class="kw">devtools</span>::<span class="fu">install_github</span>(<span class="st">'mhahsler/rBLAST'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">Biostrings</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">rBLAST</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)

<span class="kw">db_path</span> <span class="op">&lt;-</span> <span class="st">'path/to/custom/blastpdb'</span>        <span class="co"># Path to custom blastp db</span>
<span class="kw">db</span> <span class="op">&lt;-</span> <span class="fu">blast</span>(db = <span class="kw">db_path</span>, type = <span class="st">'blastp'</span>)  <span class="co"># Set blastdb</span>

<span class="kw">blast_result</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="kw">sequences</span> <span class="op">%&gt;%</span>             <span class="co"># Pangenome sequences</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="st">'[['</span>, <span class="fl">1L</span>) <span class="op">%&gt;%</span>                      <span class="co"># Subset 1 sequence from each cluster</span>
  <span class="kw">Biostrings</span>::<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span>() <span class="op">%&gt;%</span>            <span class="co"># Transform list to DNAStringSet</span>
  <span class="kw">Biostrings</span>::<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/translate.html">translate</a></span>() <span class="op">%&gt;%</span>               <span class="co"># Translate DNAStringSet</span>
  <span class="kw">rBLAST</span>::<span class="fu">predict.BLAST</span>(<span class="kw">db</span>, <span class="kw">.</span>)              <span class="co"># Run blastp. Returns data.frame</span>

<span class="co">### ADD METADATA STEP MISSING</span>
</pre></div>
</div>
<div id="neutral-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#neutral-genes" class="anchor"></a>Neutral genes</h1>
<p>The core genome is composed by those genes that are present in every or almost every genome in the sample. As these genes are present in all genomes, they can be used to extract important biological information like phylogenetic relationships, study recombination or selective pressures. For doing these, core genes need to be aligned. In the next few sections we explain how to do this to perform some of the above mentioned downstream analyses.</p>
<p>To reveal the vertical evolutionary history of a bacterial population we should take into account the presence of horizontal acquisition of genetic material, like by means of recombination, that is mantained by the action of natural selection. Hence, we can use the Tajimaâ€™s D test of neutrality as implemented in <code>pegas</code> ( <a href="https://academic.oup.com/bioinformatics/article/26/3/419/215731">Paradis, 2010</a>) to identify those core genes that are not subjected to strong selective pressures and are likely evolving neutrally.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># Load required packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">DECIPHER</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ape-package.ird.fr/pegas.html">pegas</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ape-package.ird.fr">ape</a></span>)

<span class="kw">tajimaD</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">core_seqs_4_phylo</span>() <span class="op">%&gt;%</span>      <span class="co"># Core genome sequences</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">DECIPHER</span>::<span class="kw"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignTranslation.html">AlignTranslation</a></span>) <span class="op">%&gt;%</span>  <span class="co"># Align translation</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">ape</span>::<span class="kw"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html">as.DNAbin</a></span>) <span class="op">%&gt;%</span>              <span class="co"># Transform class to DNAbin</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">pegas</span>::<span class="kw"><a href="https://rdrr.io/pkg/pegas/man/tajima.test.html">tajima.test</a></span>) <span class="op">%&gt;%</span>          <span class="co"># Compute Tajima's test</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="st">'[['</span>, <span class="st">'D'</span>)                       <span class="co"># Get Tajima's 'D' statistic from each</span>

<span class="co"># Which are neutral?</span>
<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw">tajimaD</span> <span class="op">&lt;=</span> <span class="fl">0.2</span> <span class="op">&amp;</span> <span class="kw">tajimaD</span> <span class="op">&gt;=</span> <span class="op">-</span><span class="fl">0.2</span>)  
</pre></div>
</div>
<div id="quick-dirty-phylogeny" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-dirty-phylogeny" class="anchor"></a>Quick dirty phylogeny</h1>
<p>The best practices on how to build a phylogeny from core genomes is a topic of debate. In this first example we provide a one-liner to align individual core genes (like in the previous case), produce a concatenated core genome alignment, calculate a core genome phylogeny using the Neighbor-Joining method as implemented by <code>phangorn</code> (Schliep, 2011), and visualize it using host metadata to colour tree tips using <code>ggtree</code> (Yu et al., 2017). Phylogenetic tree is assigned to the <code>phy</code> variable, and plotted as side effect by using magrittrâ€™s <code><a href="https://rdrr.io/pkg/magrittr/man/tee.html">%T&gt;%</a></code> operator.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co"># Load required packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">DECIPHER</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">Biostrings</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/KlausVigo/phangorn">phangorn</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">ggtree</span>)

<span class="kw">phy</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">core_seqs_4_phylo</span>() <span class="op">%&gt;%</span>                 <span class="co"># Core genome sequences</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">DECIPHER</span>::<span class="kw"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignSeqs.html">AlignSeqs</a></span>) <span class="op">%&gt;%</span>                <span class="co"># Align</span>
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="kw">Biostrings</span>::<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/xscat.html">xscat</a></span>, <span class="kw">.</span>) <span class="op">%&gt;%</span>              <span class="co"># Concatenate alignments</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span><span class="op">$</span><span class="kw">org</span>) <span class="op">%&gt;%</span>                  <span class="co"># Set sequence names</span>
  <span class="fu">as</span>(<span class="st">'matrix'</span>) <span class="op">%&gt;%</span>                               <span class="co"># Transform to matrix</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/phyDat.html">phyDat</a></span>(type = <span class="st">'DNA'</span>) <span class="op">%&gt;%</span>             <span class="co"># Transform to phangorn's phyDat</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/dist.hamming.html">dist.ml</a></span>() <span class="op">%&gt;%</span>                        <span class="co"># Compute distance</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/NJ.html">NJ</a></span>() <span class="op">%T&gt;%</span> {                          <span class="co"># Compute NJ, and assign "phy"</span>
    {
      <span class="kw">ggtree</span>::<span class="fu">ggtree</span>(<span class="kw">.</span>) <span class="op">%&lt;+%</span>                        <span class="co"># Create ggtree</span>
        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span>) <span class="op">+</span>                <span class="co"># Get organisms metadata</span>
        <span class="kw">ggtree</span>::<span class="fu">geom_tippoint</span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour = <span class="kw">host</span>)) <span class="op">+</span> <span class="co"># Add coloured tip points</span>
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(palette = <span class="st">'Set1'</span>)        <span class="co"># Set color palette</span>
    } <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>()
}
</pre></div>
</div>
<div id="maximum-likelihood-phylogeny" class="section level1">
<h1 class="hasAnchor">
<a href="#maximum-likelihood-phylogeny" class="anchor"></a>Maximum Likelihood phylogeny</h1>
<p>The following method is simmilar to the previous, but in this case we are optimizing the topology and branch lengths by maximum likelihood method implemented in <code>phangorn</code> package. Although this is a fully working example, its purpose is just to illustrate the idea. You should consider tuning the parameters to better fit your dataset (e.g.: just 4 discrete gamma distributions for a whole coregenome alignment may be too low).</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># Load required packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">DECIPHER</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">Biostrings</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/KlausVigo/phangorn">phangorn</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">ggtree</span>)

<span class="kw">phy</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">core_seqs_4_phylo</span>() <span class="op">%&gt;%</span>                 <span class="co"># Core genome sequences</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">DECIPHER</span>::<span class="kw"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignSeqs.html">AlignSeqs</a></span>) <span class="op">%&gt;%</span>                <span class="co"># Align</span>
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="kw">Biostrings</span>::<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/xscat.html">xscat</a></span>, <span class="kw">.</span>) <span class="op">%&gt;%</span>              <span class="co"># Concatenate alignments</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span><span class="op">$</span><span class="kw">org</span>) <span class="op">%&gt;%</span>                  <span class="co"># Set sequence names</span>
  <span class="fu">as</span>(<span class="st">'matrix'</span>) <span class="op">%&gt;%</span>                               <span class="co"># Transform to matrix</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/phyDat.html">phyDat</a></span>(type = <span class="st">'DNA'</span>) <span class="op">%T&gt;%</span>            <span class="co"># Transform to phangorn's phyDat</span>
  <span class="fu"><a href="https://rdrr.io/r/base/assign.html">assign</a></span>(<span class="st">'dat'</span>, <span class="kw">.</span>, <span class="kw">.GlobalEnv</span>) <span class="op">%&gt;%</span>               <span class="co"># Assign to "dat" in .GlobalEnv</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/dist.hamming.html">dist.ml</a></span>() <span class="op">%&gt;%</span>                        <span class="co"># Compute distance</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/NJ.html">NJ</a></span>() <span class="op">%&gt;%</span>                             <span class="co"># Compute NJ (initial tree)</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/pml.html">pml</a></span>(data = <span class="kw">dat</span>, k = <span class="fl">4</span>) <span class="op">%&gt;%</span>           <span class="co"># Compute likelihood with 4 discrete</span>
                                                 <span class="co">#  gamma distributions. </span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/pml.html">optim.pml</a></span>(rearrangement = <span class="st">"stochastic"</span>, <span class="co"># Optimize likelihood with</span>
                                                 <span class="co"># stochastic rearrangements,</span>
                      optGamma = <span class="fl">TRUE</span>,           <span class="co"># optimize gamma rate parameter,</span>
                      optInv = <span class="fl">TRUE</span>,             <span class="co"># optimize prop of variable size,</span>
                      model =<span class="st">"GTR"</span>) <span class="op">%&gt;%</span>          <span class="co"># and use "GTR" model.</span>
  <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">extract2</a></span>(<span class="st">"tree"</span>) <span class="op">%T&gt;%</span> {              <span class="co"># Extract the tree only, and pass</span>
    {                                            <span class="co">#  it to ggtree.</span>
      <span class="kw">ggtree</span>::<span class="fu">ggtree</span>(<span class="kw">.</span>) <span class="op">%&lt;+%</span>                        <span class="co"># Create ggtree</span>
        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span>) <span class="op">+</span>                <span class="co"># Get organisms metadata</span>
        <span class="kw">ggtree</span>::<span class="fu">geom_tippoint</span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour = <span class="kw">host</span>)) <span class="op">+</span> <span class="co"># Add coloured tip points</span>
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(palette = <span class="st">'Set1'</span>)        <span class="co"># Set color palette</span>
    } <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>()
  }
</pre></div>
</div>
<div id="population-structure" class="section level1">
<h1 class="hasAnchor">
<a href="#population-structure" class="anchor"></a>Population structure</h1>
<p>Identifying population structure from genomic information is a common problem in microbial ecology. This aims to identify discrete sub-populations within a more heterogeneous population, which is helpful to detect associations with particular phenotypes, geographic origin, host-association, etc. There are various methods for doing this, but probably the most used in microbial genomics is <a href="https://academic.oup.com/mbe/article/30/5/1224/998212">hierBAPS</a> (Cheng et al., 2013) which has been re-implemented in <code>R</code> as <a href="https://wellcomeopenresearch.org/articles/3-93/v1">rhierBAPS</a> (Tonkin-Hill et al., 2018). In the following recipe we will combine previous examples to 1) align core clusters; 2) compute Tajimaâ€™s D statistic over aligned core clusters to identify the ones that are likely to be evolving neutrally; 3) concatenate selected neutral clusters; 4) Run the herBAPS algorithm; 5) extract lineage information and add it to the <code>pagoo</code> object as organismâ€™s metadata; and 6) compute and plot a tree with lineage information as colour tips.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">DECIPHER</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/gtonkinhill/rhierbaps">rhierbaps</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ape-package.ird.fr">ape</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/KlausVigo/phangorn">phangorn</a></span>)

<span class="co"># 1. Align translation of core genes</span>
<span class="kw">ali</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">core_seqs_4_phylo</span>() <span class="op">%&gt;%</span>           <span class="co"># Core genome sequences</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">DECIPHER</span>::<span class="kw"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignTranslation.html">AlignTranslation</a></span>)       <span class="co"># Align translation</span>

<span class="co"># 2. Identify neutral core clusters</span>
<span class="kw">tajD</span> <span class="op">&lt;-</span> <span class="kw">ali</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">ape</span>::<span class="kw"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html">as.DNAbin</a></span>) <span class="op">%&gt;%</span>              <span class="co"># Transform class to DNAbin</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">pegas</span>::<span class="kw"><a href="https://rdrr.io/pkg/pegas/man/tajima.test.html">tajima.test</a></span>) <span class="op">%&gt;%</span>          <span class="co"># Compute Tajima's test</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="st">'[['</span>, <span class="st">'D'</span>)                       <span class="co"># Subset D statistic </span>
<span class="kw">neutral</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw">tajD</span> <span class="op">&lt;=</span> <span class="fl">0.2</span> <span class="op">&amp;</span> <span class="kw">tajD</span> <span class="op">&gt;=</span> <span class="op">-</span><span class="fl">0.2</span>)

<span class="co"># 3. Concatenate neutral core clusters</span>
<span class="kw">concat_neu</span> <span class="op">&lt;-</span> <span class="kw">ali</span>[<span class="kw">neutral</span>] <span class="op">%&gt;%</span>            <span class="co"># Select neutral clusters</span>
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="kw">Biostrings</span>::<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/xscat.html">xscat</a></span>, <span class="kw">.</span>) <span class="op">%&gt;%</span>       <span class="co"># Concatenate alignments</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span><span class="op">$</span><span class="kw">org</span>) <span class="op">%&gt;%</span>           <span class="co"># Set sequence names</span>
  <span class="fu">as</span>(<span class="st">'matrix'</span>) <span class="op">%&gt;%</span>                        <span class="co"># Transform to matrix</span>
  <span class="fu"><a href="https://rdrr.io/r/base/chartr.html">tolower</a></span>()                               <span class="co"># Translate to lower case</span>

<span class="co"># 4. Compute structure</span>
<span class="kw">rhb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhierbaps/man/hierBAPS.html">hierBAPS</a></span>(snp.matrix = <span class="kw">concat_neu</span>, <span class="co"># Input matrix alignment</span>
                n.pops = <span class="fl">10</span>,             <span class="co"># Max number of subpopulations</span>
                max.depth = <span class="fl">1</span>,           <span class="co"># Max depth for hierarchical clustering</span>
                n.extra.rounds = <span class="fl">5</span>)      <span class="co"># Extra rounds to ensure convergence</span>

<span class="co"># 5. Add lineage as metadata to organisms in pagoo object</span>
<span class="kw">res</span> <span class="op">&lt;-</span> <span class="kw">rhb</span><span class="op">$</span><span class="kw">partition.df</span>
<span class="kw">lin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(org = <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw">res</span>[, <span class="fl">1</span>]), 
                  lineage = <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="kw">res</span>[, <span class="fl">2</span>]))
<span class="kw">p</span><span class="op">$</span><span class="fu">add_metadata</span>(map = <span class="st">'org'</span>, data = <span class="kw">lin</span>)

<span class="co"># 6. Compute tree and plot it with lineage information</span>
<span class="kw">concat_neu</span> <span class="op">%&gt;%</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/phyDat.html">phyDat</a></span>(type = <span class="st">'DNA'</span>) <span class="op">%&gt;%</span>                <span class="co"># Transform to phangorn's phyDat</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/dist.hamming.html">dist.ml</a></span>() <span class="op">%&gt;%</span>                           <span class="co"># Compute distance</span>
  <span class="kw">phangorn</span>::<span class="fu"><a href="https://rdrr.io/pkg/phangorn/man/NJ.html">NJ</a></span>() <span class="op">%&gt;%</span>                                <span class="co"># Compute NJ</span>
  <span class="kw">ggtree</span>::<span class="fu">ggtree</span>() <span class="op">%&lt;+%</span>                             <span class="co"># Create ggtree</span>
     <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span>) <span class="op">+</span>                   <span class="co"># Get organisms metadata</span>
     <span class="kw">ggtree</span>::<span class="fu">geom_tippoint</span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour = <span class="kw">lineage</span>))   <span class="co"># Colour tips with lineage info</span>
</pre></div>
</div>
<div id="pairwise-distance-vs-pairwise-nucleotide-diversity" class="section level1">
<h1 class="hasAnchor">
<a href="#pairwise-distance-vs-pairwise-nucleotide-diversity" class="anchor"></a>Pairwise distance vs pairwise nucleotide diversity</h1>
<p>The following recipe computes accessory genome jaccardâ€™s distance and nucleotide diversity of synonym sites for each pair of genomes, and then plot it with a linear regression curve. This method provides insights of accessory genome adaptative evolution. If both magnitudes are not correlated, it could mean that there are some selective pressures shaping cladeâ€™s evolution. It first sets the <code>core_level</code> threshold to 100%, then creates a matrix to store pairwise distances and synonym nucleotide diversities. Accessory genome Jaccard distance between each pair of organisms is computed by a <code>pagoo</code> method which is basically a wrapper of <code><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegan::vegdist</a></code> function. Then there is a long code block with the method to align each core gene and select their polymorphic synonym sites. With this input, pairwise nucleotide diversity is computed with <code><a href="https://rdrr.io/pkg/pegas/man/nuc.div.html">pegas::nuc.div</a></code> function. At the end, a correlation between both magnitudes is plotted.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">IRanges</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">Biostrings</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">DECIPHER</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ape-package.ird.fr">ape</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ape-package.ird.fr/pegas.html">pegas</a></span>)

<span class="co"># Set core level to 100%. This recipe only works if this is set to 100%.</span>
<span class="kw">p</span><span class="op">$</span><span class="kw">core_level</span> <span class="op">&lt;-</span> <span class="fl">100</span>

<span class="co"># Create pairs matrix</span>
<span class="kw">pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/combn.html">combn</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span>), <span class="fl">2</span>)))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">pairs</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'org1'</span>, <span class="st">'org2'</span>)

<span class="co"># Compute paired jaccard similarity, transform to matrix</span>
<span class="kw">jaccard_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw">p</span><span class="op">$</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(method = <span class="st">"jaccard"</span>, binary = <span class="fl">TRUE</span>))
<span class="co"># Fill results matrix</span>
<span class="kw">pairs</span><span class="op">$</span><span class="kw">jaccard_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="kw">pairs</span>, <span class="fl">1</span>, <span class="fu">function</span>(<span class="kw">i</span>){
  <span class="kw">ii</span> <span class="op">&lt;-</span> <span class="kw">i</span>[<span class="fl">1</span>]
  <span class="kw">jj</span> <span class="op">&lt;-</span> <span class="kw">i</span>[<span class="fl">2</span>]
  <span class="kw">jaccard_sim</span>[<span class="kw">ii</span>, <span class="kw">jj</span>]
})

<span class="co"># Return only synonymous polymorphic sites.</span>
<span class="co"># First, it removes non-synonymous codons, and then retains only</span>
<span class="co"># polymorphyc sites.</span>
<span class="kw">syn_poly_sites</span> <span class="op">&lt;-</span> <span class="kw">p</span><span class="op">$</span><span class="fu">core_seqs_4_phylo</span>() <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu">function</span>(<span class="kw">x</span>){
    <span class="kw">lns</span> <span class="op">&lt;-</span> <span class="fu">elementNROWS</span>(<span class="kw">x</span>)                              <span class="co"># Align translatation filtering</span>
    <span class="kw">tali</span> <span class="op">&lt;-</span> <span class="kw">x</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw">lns</span> != <span class="fl">0</span>)] <span class="op">%&gt;%</span>                      <span class="co">#  truncated codons and returning</span>
      <span class="kw">Biostrings</span>::<span class="fu">subseq</span>(<span class="fl">1L</span>, <span class="kw">lns</span> <span class="op">%/%</span> <span class="fl">3</span> <span class="op">*</span> <span class="fl">3</span>) <span class="op">%&gt;%</span>         <span class="co">#  both DNA and AA alignments.</span>
      <span class="kw">DECIPHER</span>::<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignTranslation.html">AlignTranslation</a></span>(type = <span class="st">"both"</span>)
    <span class="kw">syno</span> <span class="op">&lt;-</span>  <span class="kw">tali</span>[[<span class="fl">2</span>]] <span class="op">%&gt;%</span>                              <span class="co"># Identify non-synonymous</span>
      <span class="kw">Biostrings</span>::<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">consensusMatrix</a></span>() <span class="op">%&gt;%</span>                 <span class="co">#  codons.</span>
      <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">equals</a></span>(<span class="fl">0</span>) <span class="op">%&gt;%</span>
      <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">not</a></span>() <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>() <span class="op">%&gt;%</span>
      <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">equals</a></span>(<span class="fl">1</span>) <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>()
    <span class="kw">neut</span> <span class="op">&lt;-</span> <span class="kw">tali</span>[[<span class="fl">1</span>]] <span class="op">%&gt;%</span>                               <span class="co"># Remove non-synonymous codons.</span>
      <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu">function</span>(<span class="kw">x</span>){
        <span class="kw">IRanges</span>::<span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/Views-class.html">successiveViews</a></span>(
          <span class="kw">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep.int</a></span>(<span class="fl">3L</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw">x</span>) <span class="op">%/%</span> <span class="fl">3L</span>))
      }) <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="st">'['</span>, <span class="kw">syno</span>) <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">unlist</span>) <span class="op">%&gt;%</span>
      <span class="kw">Biostrings</span>::<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span>()
    <span class="kw">poly</span> <span class="op">&lt;-</span> <span class="kw">neut</span> <span class="op">%&gt;%</span>                                   <span class="co"># Identify polymorphic sites.</span>
      <span class="kw">Biostrings</span>::<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">consensusMatrix</a></span>() <span class="op">%&gt;%</span>
      <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">equals</a></span>(<span class="fl">0</span>) <span class="op">%&gt;%</span>
      <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">not</a></span>() <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>() <span class="op">%&gt;%</span>
      <span class="kw">magrittr</span>::<span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">is_greater_than</a></span>(<span class="fl">1</span>) <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>()
    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">neut</span>, <span class="st">'['</span>, <span class="kw">poly</span>) <span class="op">%&gt;%</span>                        <span class="co"># Retain only polymorphic</span>
      <span class="kw">Biostrings</span>::<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span>()                       <span class="co">#  sites</span>
  }) <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="kw">Biostrings</span>::<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/xscat.html">xscat</a></span>, <span class="kw">.</span>) <span class="op">%&gt;%</span>                    <span class="co"># Concatenate.</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw">p</span><span class="op">$</span><span class="kw">organisms</span><span class="op">$</span><span class="kw">org</span>) <span class="op">%&gt;%</span>                        <span class="co"># Set names.</span>
  <span class="kw">ape</span>::<span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html">as.DNAbin</a></span>()                                     <span class="co"># Convert to DNAbin class.</span>

<span class="co"># Compute paired nucleotide diversity, and fill results matrix</span>
<span class="kw">pairs</span><span class="op">$</span><span class="kw">nuc_div</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="kw">pairs</span>, <span class="fl">1</span>, <span class="fu">function</span>(<span class="kw">i</span>){
  <span class="kw">ii</span> <span class="op">&lt;-</span> <span class="kw">i</span>[<span class="fl">1</span>]
  <span class="kw">jj</span> <span class="op">&lt;-</span> <span class="kw">i</span>[<span class="fl">2</span>]
  <span class="kw">pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">syn_poly_sites</span>[<span class="kw">ii</span>], <span class="kw">syn_poly_sites</span>[<span class="kw">jj</span>])
  <span class="kw">pegas</span>::<span class="fu"><a href="https://rdrr.io/pkg/pegas/man/nuc.div.html">nuc.div</a></span>(<span class="kw">pair</span>)
})

<span class="co"># Plot correlation with R-base graphics</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span>(<span class="kw">jaccard_sim</span> <span class="op">~</span> <span class="kw">nuc_div</span>, <span class="kw">pairs</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="kw">jaccard_sim</span> <span class="op">~</span> <span class="kw">nuc_div</span>, <span class="kw">pairs</span>))
</pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ignacio Ferres, Gregorio Iraola, Institut Pasteur de Montevideo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
