<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Subseting • pagoo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Subseting">
<meta property="og:description" content="pagoo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-164668935-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-164668935-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pagoo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/pagoo.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Input.html">1 - Input</a>
    </li>
    <li>
      <a href="../articles/Querying_Data.html">2 - Querying Data</a>
    </li>
    <li>
      <a href="../articles/Subseting.html">3 - Subseting</a>
    </li>
    <li>
      <a href="../articles/Methods_Plots.html">4 - Methods and Plots</a>
    </li>
    <li>
      <a href="../articles/Recipes.html">5 - Recipes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iferres/pagoo/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Subseting_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Subseting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/iferres/pagoo/blob/HEAD/vignettes/Subseting.Rmd" class="external-link"><code>vignettes/Subseting.Rmd</code></a></small>
      <div class="hidden name"><code>Subseting.Rmd</code></div>

    </div>

    
    
<p>One fundamental operation when working with pangenomic data is to subset it in order to better analyze it. <code>pagoo</code> provides three convenient ways that can help in different scenarios. First we will see predefined subsets, then we will understand how to use the classic R’s <code>[</code> operator with <code>pagoo</code> objects, and finally we will see how to temporarily remove one or more organisms from the dataset, and then to recover them.</p>
<p>Start the tutorial by loading the included <em>Campylobacter</em> pangenome.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://iferres.github.io/pagoo/" class="external-link">pagoo</a></span><span class="op">)</span> <span class="co"># Load package</span>
<span class="va">toy_rds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'campylobacter.RDS'</span>, package <span class="op">=</span> <span class="st">'pagoo'</span><span class="op">)</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_pangenomeRDS.html">load_pangenomeRDS</a></span><span class="op">(</span><span class="va">toy_rds</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="predefined-subsets">Predefined Subsets<a class="anchor" aria-label="anchor" href="#predefined-subsets"></a>
</h2>
<p>Predefined subsets were already superficially seen in <strong>3 - Querying Data</strong> tutorial. Here we will review them and understand its notation.</p>
<table class="table">
<thead><tr class="header">
<th>-</th>
<th><code>$core_*</code></th>
<th><code>$shell_*</code></th>
<th><code>$cloud_*</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong><code>$*_genes</code></strong></td>
<td><code>$core_genes</code></td>
<td><code>$shell_genes</code></td>
<td><code>$cloud_genes</code></td>
</tr>
<tr class="even">
<td><strong><code>$*_clusters</code></strong></td>
<td><code>$core_clusters</code></td>
<td><code>$shell_clusters</code></td>
<td><code>$cloud_clusters</code></td>
</tr>
<tr class="odd">
<td><strong><code>$*_sequences</code></strong></td>
<td><code>$core_sequences</code></td>
<td><code>$shell_sequences</code></td>
<td><code>$cloud_sequences</code></td>
</tr>
</tbody>
</table>
<p>As seen in the above table, the notation is quite straightforward. Lets take <code>clusters</code> for instance to illustrate it better.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">clusters</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 2588 rows and 2 columns</span>
<span class="co">##        cluster              Pfam_Arch</span>
<span class="co">##       &lt;factor&gt;            &lt;character&gt;</span>
<span class="co">## 1    group0001           2-Hacid_dh_C</span>
<span class="co">## 2    group0002 2-Hacid_dh_C;2-Hacid..</span>
<span class="co">## 3    group0003 2-Hacid_dh_C;ACT;2-H..</span>
<span class="co">## 4    group0004        2Fe-2S_thioredx</span>
<span class="co">## 5    group0005    4HB_MCP_1;MCPsignal</span>
<span class="co">## ...        ...                    ...</span>
<span class="co">## 2584 group2584              zf-RING_7</span>
<span class="co">## 2585 group2585               zf-TFIIB</span>
<span class="co">## 2586 group2586                   ZinT</span>
<span class="co">## 2587 group2587                   ZnuA</span>
<span class="co">## 2588 group2588               ZT_dimer</span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">core_clusters</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 1627 rows and 2 columns</span>
<span class="co">##        cluster              Pfam_Arch</span>
<span class="co">##       &lt;factor&gt;            &lt;character&gt;</span>
<span class="co">## 1    group0001           2-Hacid_dh_C</span>
<span class="co">## 2    group0002 2-Hacid_dh_C;2-Hacid..</span>
<span class="co">## 3    group0003 2-Hacid_dh_C;ACT;2-H..</span>
<span class="co">## 4    group0004        2Fe-2S_thioredx</span>
<span class="co">## 5    group0006    4HB_MCP_1;MCPsignal</span>
<span class="co">## ...        ...                    ...</span>
<span class="co">## 1623 group2584              zf-RING_7</span>
<span class="co">## 1624 group2585               zf-TFIIB</span>
<span class="co">## 1625 group2586                   ZinT</span>
<span class="co">## 1626 group2587                   ZnuA</span>
<span class="co">## 1627 group2588               ZT_dimer</span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">shell_clusters</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 413 rows and 2 columns</span>
<span class="co">##       cluster            Pfam_Arch</span>
<span class="co">##      &lt;factor&gt;          &lt;character&gt;</span>
<span class="co">## 1   group0005  4HB_MCP_1;MCPsignal</span>
<span class="co">## 2   group0009        5_nucleotid_C</span>
<span class="co">## 3   group0015                  AAA</span>
<span class="co">## 4   group0019               AAA_22</span>
<span class="co">## 5   group0021               AAA_22</span>
<span class="co">## ...       ...                  ...</span>
<span class="co">## 409 group2537 VirB3;CagE_TrbE_VirB</span>
<span class="co">## 410 group2541                VirB8</span>
<span class="co">## 411 group2543                VirB8</span>
<span class="co">## 412 group2554               Y1_Tnp</span>
<span class="co">## 413 group2557           YafQ_toxin</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">cloud_clusters</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 548 rows and 2 columns</span>
<span class="co">##       cluster   Pfam_Arch</span>
<span class="co">##      &lt;factor&gt; &lt;character&gt;</span>
<span class="co">## 1   group0018      AAA_22</span>
<span class="co">## 2   group0020      AAA_22</span>
<span class="co">## 3   group0024      AAA_25</span>
<span class="co">## 4   group0025      AAA_25</span>
<span class="co">## 5   group0028      AAA_28</span>
<span class="co">## ...       ...         ...</span>
<span class="co">## 544 group2555  YafQ_toxin</span>
<span class="co">## 545 group2566      YhcG_C</span>
<span class="co">## 546 group2573        YopX</span>
<span class="co">## 547 group2578  Zeta_toxin</span>
<span class="co">## 548 group2579  Zeta_toxin</span></code></pre>
</div>
<div class="section level2">
<h2 id="the-operator">The <code>[</code> operator<a class="anchor" aria-label="anchor" href="#the-operator"></a>
</h2>
<p>Using <code>[</code> for subset vectors, lists, or matrices in R is one of the most common operations we, as R users, do every day. Methods using <code>[</code> for subsetting can be divided into 2 types:</p>
<ol style="list-style-type: decimal">
<li>Subsetting vector or list like objects.</li>
<li>Subsetting matrix-like objects.</li>
</ol>
<p>In the following subsections we will see how to use them to subset <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, or<code>$organisms</code> fields.</p>
<div class="section level3">
<h3 id="vectorlist-notation">Vector/List notation<a class="anchor" aria-label="anchor" href="#vectorlist-notation"></a>
</h3>
<p>When we subset a vector or list in R we use a single number or a vector of numbers, lets say from <code>i</code> to <code>j</code>, to pick those elements by indexes: <code>x[i:j]</code>. In <code>pagoo</code> we implement a method for this generic function to subset data fields, where indexes represent <strong>clusters</strong>. So instead of subsetting directly the final object (which you can of course) you apply this method directly to the object, and then select any of the data fields.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># From clusters 1 to 3</span>
<span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">pan_matrix</span></code></pre></div>
<pre><code><span class="co">##            group0001 group0005 group0018</span>
<span class="co">## 16244_6_6          1         1         0</span>
<span class="co">## 16244_6_18         1         1         0</span>
<span class="co">## 17059_2_16         1         1         0</span>
<span class="co">## 17059_2_23         1         0         0</span>
<span class="co">## 17059_2_27         1         1         0</span>
<span class="co">## 17150_1_73         1         1         0</span>
<span class="co">## 17059_2_42         1         0         1</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span></code></pre></div>
<pre><code><span class="co">## SplitDataFrameList of length 3</span>
<span class="co">## $group0001</span>
<span class="co">## DataFrame with 7 rows and 10 columns</span>
<span class="co">##     cluster        org             gene                    gid    geneName</span>
<span class="co">##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 group0001 16244_6_6  16244_6_6_00150  16244_6_6__16244_6_6..        ilvC</span>
<span class="co">## 2 group0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_..        ilvC</span>
<span class="co">## 3 group0001 17059_2_16 17059_2_16_01012 17059_2_16__17059_2_..        ilvC</span>
<span class="co">## 4 group0001 17059_2_23 17059_2_23_01040 17059_2_23__17059_2_..        ilvC</span>
<span class="co">## 5 group0001 17059_2_27 17059_2_27_00492 17059_2_27__17059_2_..        ilvC</span>
<span class="co">## 6 group0001 17150_1_73 17150_1_73_00221 17150_1_73__17150_1_..        ilvC</span>
<span class="co">## 7 group0001 17059_2_42 17059_2_42_00176 17059_2_42__17059_2_..        ilvC</span>
<span class="co">##                  product                 contig      from        to      strand</span>
<span class="co">##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## 1 ketol-acid reductois.. ERS672247|SC|contig0..    137333    138355           +</span>
<span class="co">## 2 ketol-acid reductois.. ERS672259|SC|contig0..    173541    174563           +</span>
<span class="co">## 3 ketol-acid reductois.. ERS739235|SC|contig0..     43645     44667           -</span>
<span class="co">## 4 ketol-acid reductois.. ERS739242|SC|contig0..    173578    174600           +</span>
<span class="co">## 5 ketol-acid reductois.. ERS739246|SC|contig0..    184035    185057           -</span>
<span class="co">## 6 ketol-acid reductois.. ERS686652|SC|contig0..    207206    208228           +</span>
<span class="co">## 7 ketol-acid reductois.. ERS739261|SC|contig0..    173569    174591           +</span>
<span class="co">## </span>
<span class="co">## $group0005</span>
<span class="co">## DataFrame with 5 rows and 10 columns</span>
<span class="co">##     cluster        org             gene                    gid    geneName</span>
<span class="co">##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 group0005 16244_6_6  16244_6_6_00049  16244_6_6__16244_6_6..      mcpB_1</span>
<span class="co">## 2 group0005 16244_6_18 16244_6_18_00072 16244_6_18__16244_6_..      mcpB_1</span>
<span class="co">## 3 group0005 17059_2_16 17059_2_16_01270 17059_2_16__17059_2_..      mcpB_3</span>
<span class="co">## 4 group0005 17059_2_27 17059_2_27_00592 17059_2_27__17059_2_..      mcpB_1</span>
<span class="co">## 5 group0005 17150_1_73 17150_1_73_00073 17150_1_73__17150_1_..      mcpB_1</span>
<span class="co">##                  product                 contig      from        to      strand</span>
<span class="co">##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## 1 methyl-accepting che.. ERS672247|SC|contig0..     46407     48068           +</span>
<span class="co">## 2 methyl-accepting che.. ERS672259|SC|contig0..     82617     84278           +</span>
<span class="co">## 3 methyl-accepting che.. ERS739235|SC|contig0..     82775     84436           +</span>
<span class="co">## 4 methyl-accepting che.. ERS739246|SC|contig0..    274321    275982           -</span>
<span class="co">## 5 methyl-accepting che.. ERS686652|SC|contig0..     82648     84309           +</span>
<span class="co">## </span>
<span class="co">## $group0018</span>
<span class="co">## DataFrame with 1 row and 10 columns</span>
<span class="co">##     cluster        org             gene                    gid    geneName</span>
<span class="co">##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 group0018 17059_2_42 17059_2_42_00333 17059_2_42__17059_2_..            </span>
<span class="co">##                  product                 contig      from        to      strand</span>
<span class="co">##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## 1 bacteriocin,putative.. ERS739261|SC|contig0..    303996    304949           -</span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">clusters</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 3 rows and 2 columns</span>
<span class="co">##     cluster           Pfam_Arch</span>
<span class="co">##    &lt;factor&gt;         &lt;character&gt;</span>
<span class="co">## 1 group0001        2-Hacid_dh_C</span>
<span class="co">## 2 group0005 4HB_MCP_1;MCPsignal</span>
<span class="co">## 3 group0018              AAA_22</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">sequences</span></code></pre></div>
<pre><code><span class="co">## DNAStringSetList of length 3</span>
<span class="co">## [["group0001"]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTG...</span>
<span class="co">## [["group0005"]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAA...</span>
<span class="co">## [["group0018"]] 17059_2_42__17059_2_42_00333=ATGAGTAAATTTGAAGATATTAGGAATGAGCT...</span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">organisms</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 7 rows and 8 columns</span>
<span class="co">##          org          id      strain      year     country        host</span>
<span class="co">##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 16244_6_6         FR15   2008/170h      2008      France       Human</span>
<span class="co">## 2 16244_6_18        FR27   2012/185h      2012      France       Human</span>
<span class="co">## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine</span>
<span class="co">## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine</span>
<span class="co">## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine</span>
<span class="co">## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human</span>
<span class="co">## 7 17059_2_42         TW6        1830      2008      Taiwan       Human</span>
<span class="co">##        source   accession</span>
<span class="co">##   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1       Feces   ERS672247</span>
<span class="co">## 2       Blood   ERS672259</span>
<span class="co">## 3     Prepuce   ERS739235</span>
<span class="co">## 4       Fetus   ERS739242</span>
<span class="co">## 5          VM   ERS739246</span>
<span class="co">## 6       Blood   ERS686652</span>
<span class="co">## 7       Blood   ERS739261</span></code></pre>
<p>Note that in each case, indexes can be interpreted as clusters. In <code>pan_matrix</code> they refer to columns, in <code>genes</code> and <code>sequences</code> to elements in a list, in <code>clusters</code> rows in the dataframe, and in the case of <code>$organisms</code> it returns the organisms (rows) where those clusters are present. See what happen for this last case when we query for organisms present for a shell cluster:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">shell_clust</span> <span class="op">&lt;-</span> <span class="va">p</span><span class="op">$</span><span class="va">shell_clusters</span><span class="op">$</span><span class="va">cluster</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="co"># [1] "OG0005"</span>
<span class="va">p</span><span class="op">[</span><span class="va">shell_clust</span><span class="op">]</span><span class="op">$</span><span class="va">organisms</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 5 rows and 8 columns</span>
<span class="co">##          org          id      strain      year     country        host</span>
<span class="co">##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 16244_6_6         FR15   2008/170h      2008      France       Human</span>
<span class="co">## 2 16244_6_18        FR27   2012/185h      2012      France       Human</span>
<span class="co">## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine</span>
<span class="co">## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine</span>
<span class="co">## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human</span>
<span class="co">##        source   accession</span>
<span class="co">##   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1       Feces   ERS672247</span>
<span class="co">## 2       Blood   ERS672259</span>
<span class="co">## 3     Prepuce   ERS739235</span>
<span class="co">## 5          VM   ERS739246</span>
<span class="co">## 6       Blood   ERS686652</span></code></pre>
<p>Only (6) organisms which contain that shell cluster will be listed.</p>
</div>
<div class="section level3">
<h3 id="matrix-notation">Matrix notation<a class="anchor" aria-label="anchor" href="#matrix-notation"></a>
</h3>
<p>The other use of <code>[</code> notation is when we subset a matrix-like object. In this case we provide 2 sets of indexes: the first for rows, and the second for columns, separated by a coma. <code>pagoo</code> interprets these indexes as organisms (rows) and clusters (columns), so you are referencing a set of genes (cell value). Basically you are referencing cells in the pan_matrix.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">pan_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##            group0001 group0005 group0018</span>
<span class="co">## 16244_6_6          1         1         0</span>
<span class="co">## 16244_6_18         1         1         0</span>
<span class="co">## 17059_2_16         1         1         0</span></code></pre>
<p>Here we are selecting the first 3 organisms, and the clusters indexed as 1, 5, and 18. Note that we see (sum columns) 3 genes in the first cluster, 3 in the second, and 0 in the third. Now we flip the notation and subset directly the <code>pagoo</code> object, and then ask for a data field:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">pan_matrix</span> <span class="co"># The same as above</span></code></pre></div>
<pre><code><span class="co">##            group0001 group0005 group0018</span>
<span class="co">## 16244_6_6          1         1         0</span>
<span class="co">## 16244_6_18         1         1         0</span>
<span class="co">## 17059_2_16         1         1         0</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span></code></pre></div>
<pre><code><span class="co">## SplitDataFrameList of length 2</span>
<span class="co">## $group0001</span>
<span class="co">## DataFrame with 3 rows and 10 columns</span>
<span class="co">##     cluster        org             gene                    gid    geneName</span>
<span class="co">##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 group0001 16244_6_6  16244_6_6_00150  16244_6_6__16244_6_6..        ilvC</span>
<span class="co">## 2 group0001 16244_6_18 16244_6_18_00172 16244_6_18__16244_6_..        ilvC</span>
<span class="co">## 3 group0001 17059_2_16 17059_2_16_01012 17059_2_16__17059_2_..        ilvC</span>
<span class="co">##                  product                 contig      from        to      strand</span>
<span class="co">##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## 1 ketol-acid reductois.. ERS672247|SC|contig0..    137333    138355           +</span>
<span class="co">## 2 ketol-acid reductois.. ERS672259|SC|contig0..    173541    174563           +</span>
<span class="co">## 3 ketol-acid reductois.. ERS739235|SC|contig0..     43645     44667           -</span>
<span class="co">## </span>
<span class="co">## $group0005</span>
<span class="co">## DataFrame with 3 rows and 10 columns</span>
<span class="co">##     cluster        org             gene                    gid    geneName</span>
<span class="co">##    &lt;factor&gt;   &lt;factor&gt;         &lt;factor&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 group0005 16244_6_6  16244_6_6_00049  16244_6_6__16244_6_6..      mcpB_1</span>
<span class="co">## 2 group0005 16244_6_18 16244_6_18_00072 16244_6_18__16244_6_..      mcpB_1</span>
<span class="co">## 3 group0005 17059_2_16 17059_2_16_01270 17059_2_16__17059_2_..      mcpB_3</span>
<span class="co">##                  product                 contig      from        to      strand</span>
<span class="co">##              &lt;character&gt;            &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## 1 methyl-accepting che.. ERS672247|SC|contig0..     46407     48068           +</span>
<span class="co">## 2 methyl-accepting che.. ERS672259|SC|contig0..     82617     84278           +</span>
<span class="co">## 3 methyl-accepting che.. ERS739235|SC|contig0..     82775     84436           +</span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">clusters</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 2 rows and 2 columns</span>
<span class="co">##     cluster           Pfam_Arch</span>
<span class="co">##    &lt;factor&gt;         &lt;character&gt;</span>
<span class="co">## 1 group0001        2-Hacid_dh_C</span>
<span class="co">## 2 group0005 4HB_MCP_1;MCPsignal</span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">sequences</span></code></pre></div>
<pre><code><span class="co">## DNAStringSetList of length 2</span>
<span class="co">## [["group0001"]] 16244_6_18__16244_6_18_00172=ATGGCGATAACAGTTTATTACGACAAAGATTG...</span>
<span class="co">## [["group0005"]] 16244_6_18__16244_6_18_00072=ATGTCAAATTTAACTACTAATTTAACTACCAA...</span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">organisms</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 3 rows and 8 columns</span>
<span class="co">##          org          id      strain      year     country        host</span>
<span class="co">##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 16244_6_6         FR15   2008/170h      2008      France       Human</span>
<span class="co">## 2 16244_6_18        FR27   2012/185h      2012      France       Human</span>
<span class="co">## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine</span>
<span class="co">##        source   accession</span>
<span class="co">##   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1       Feces   ERS672247</span>
<span class="co">## 2       Blood   ERS672259</span>
<span class="co">## 3     Prepuce   ERS739235</span></code></pre>
<p>So this selection returns a list of length 2, with 3 elements each for <code>$genes</code> and <code>$sequences</code>, a dataframe showing only the selected clusters in the <code>$clusters</code> field, and a dataframe showing only the selected organisms in the <code>$organisms</code> field.</p>
<p>We found this implementation quite useful for both data exploration and fine grained analysis.</p>
</div>
</div>
<div class="section level2">
<h2 id="dropping-and-recovering-organisms">Dropping and Recovering Organisms<a class="anchor" aria-label="anchor" href="#dropping-and-recovering-organisms"></a>
</h2>
<p>One useful feature implemented in <code>pagoo</code> is the possibility of easily removing (hiding) organisms from the dataset. This is useful if during the analysis we identify some genome with weird characteristics (i.e. potentially contaminated), or if we want to focus just in a subset of genomes of interest given any metadata value, or if we included an outgroup for phylogenetic purposes but we want to remove it from downstream analyses. Let’s see how it works..</p>
<div class="section level3">
<h3 id="dropping-organisms">Dropping organisms<a class="anchor" aria-label="anchor" href="#dropping-organisms"></a>
</h3>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">organisms</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 7 rows and 8 columns</span>
<span class="co">##          org          id      strain      year     country        host</span>
<span class="co">##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 16244_6_6         FR15   2008/170h      2008      France       Human</span>
<span class="co">## 2 16244_6_18        FR27   2012/185h      2012      France       Human</span>
<span class="co">## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine</span>
<span class="co">## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine</span>
<span class="co">## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine</span>
<span class="co">## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human</span>
<span class="co">## 7 17059_2_42         TW6        1830      2008      Taiwan       Human</span>
<span class="co">##        source   accession</span>
<span class="co">##   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1       Feces   ERS672247</span>
<span class="co">## 2       Blood   ERS672259</span>
<span class="co">## 3     Prepuce   ERS739235</span>
<span class="co">## 4       Fetus   ERS739242</span>
<span class="co">## 5          VM   ERS739246</span>
<span class="co">## 6       Blood   ERS686652</span>
<span class="co">## 7       Blood   ERS739261</span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">summary_stats</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 4 rows and 2 columns</span>
<span class="co">##      Category    Number</span>
<span class="co">##   &lt;character&gt; &lt;integer&gt;</span>
<span class="co">## 1       Total      2588</span>
<span class="co">## 2        Core      1627</span>
<span class="co">## 3       Shell       413</span>
<span class="co">## 4       Cloud       548</span></code></pre>
<p>We can see that we have 7 organisms, and we have the pangenome summary statistics. Let’s say, for instance, that we want to exclude french isolates (because any reason) from the dataset, which are indexed 1 and 2 (rows) in the above dataframe.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="fu">drop</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> 
<span class="va">p</span><span class="op">$</span><span class="va">organisms</span> <span class="co"># Updated organisms !!</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 5 rows and 8 columns</span>
<span class="co">##          org          id      strain      year     country        host</span>
<span class="co">##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine</span>
<span class="co">## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine</span>
<span class="co">## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine</span>
<span class="co">## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human</span>
<span class="co">## 7 17059_2_42         TW6        1830      2008      Taiwan       Human</span>
<span class="co">##        source   accession</span>
<span class="co">##   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 3     Prepuce   ERS739235</span>
<span class="co">## 4       Fetus   ERS739242</span>
<span class="co">## 5          VM   ERS739246</span>
<span class="co">## 6       Blood   ERS686652</span>
<span class="co">## 7       Blood   ERS739261</span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">summary_stats</span> <span class="co"># Updated stats !!</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 4 rows and 2 columns</span>
<span class="co">##      Category    Number</span>
<span class="co">##   &lt;character&gt; &lt;integer&gt;</span>
<span class="co">## 1       Total      2504</span>
<span class="co">## 2        Core      1648</span>
<span class="co">## 3       Shell       309</span>
<span class="co">## 4       Cloud       547</span></code></pre>
<p>As you see, not only the french organisms have been removed from the dataset, but also summary statistics have been updated! This is also true for any other data field: <code>$pan_matrix</code>, <code>$genes</code>, <code>$clusters</code>, <code>$sequences</code>, and<code>$organisms</code> fields are automatically updated. Also any embedded statistical or visualization methods will now only consider available organisms/clusters/genes. When we drop an organisms, we are hiding all features associated to that genomes including genes, sequences, gene clusters and metadata.</p>
<p><strong>It’s important to note that you don’t have to reassign the object to a new one, it is self modified (in place modification). That’s R6 reference semantics (see R6 documentation for details), use with caution.</strong></p>
</div>
<div class="section level3">
<h3 id="recovering-dropped-organisms">Recovering dropped organisms<a class="anchor" aria-label="anchor" href="#recovering-dropped-organisms"></a>
</h3>
<p>You can recover dropped organisms. To see if we have any hidden organism, we use the <code>$dropped</code> field, and then recover it using its index or its name.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="va">dropped</span></code></pre></div>
<pre><code><span class="co">##            1            2 </span>
<span class="co">##  "16244_6_6" "16244_6_18"</span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span><span class="op">$</span><span class="fu">recover</span><span class="op">(</span> <span class="va">p</span><span class="op">$</span><span class="va">dropped</span> <span class="op">)</span>
<span class="va">p</span><span class="op">$</span><span class="va">organisms</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 7 rows and 8 columns</span>
<span class="co">##          org          id      strain      year     country        host</span>
<span class="co">##     &lt;factor&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1 16244_6_6         FR15   2008/170h      2008      France       Human</span>
<span class="co">## 2 16244_6_18        FR27   2012/185h      2012      France       Human</span>
<span class="co">## 3 17059_2_16         AR1      99/801      1999   Argentina      Bovine</span>
<span class="co">## 4 17059_2_23         AR8      04/875      2004   Argentina      Bovine</span>
<span class="co">## 5 17059_2_27        AR12      06/195      2006   Argentina      Bovine</span>
<span class="co">## 6 17150_1_73         CA1   001A-0374      2005      Canada       Human</span>
<span class="co">## 7 17059_2_42         TW6        1830      2008      Taiwan       Human</span>
<span class="co">##        source   accession</span>
<span class="co">##   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1       Feces   ERS672247</span>
<span class="co">## 2       Blood   ERS672259</span>
<span class="co">## 3     Prepuce   ERS739235</span>
<span class="co">## 4       Fetus   ERS739242</span>
<span class="co">## 5          VM   ERS739246</span>
<span class="co">## 6       Blood   ERS686652</span>
<span class="co">## 7       Blood   ERS739261</span></code></pre>
<p>Now french isolates and all its features are again available.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ignacio Ferres, Gregorio Iraola, Institut Pasteur de Montevideo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
